<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">estimating an allele frequency | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="estimating an allele frequency | WHG training resources"><meta data-rh="true" name="description" content="The frequency of O blood group"><meta data-rh="true" property="og:description" content="The frequency of O blood group"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.a63558cc.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.81209970.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/whg-training-resources/programming/introduction_to_R/">programming</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Hidden markov models/">README</a><button aria-label="Toggle the collapsible sidebar category &#x27;README&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/">GMS Statistical Modelling I - introduction to statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;GMS Statistical Modelling I - introduction to statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/notes/Probability solutions/">notes</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">practicals</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">a_tale_of_two_2x2_tables</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/">bayesian_meta_analysis</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical/">linear_regression</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/">malaria_exposure</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison/">o_bld_grp</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/distributions warmup/">warmup</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/distributions warmup/">distributions warmup</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency/">estimating an allele frequency</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/probability warmup/">probability warmup</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-5 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/solutions/probability warmup/">solutions</a></div></li></ul></li></ul></li></ul></li></ul></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/"><span itemprop="name">Statistical modelling</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/Introduction/"><span itemprop="name">GMS Statistical Modelling I - introduction to statistical modelling</span></a><meta itemprop="position" content="2"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">practicals</span><meta itemprop="position" content="3"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">warmup</span><meta itemprop="position" content="4"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">estimating an allele frequency</span><meta itemprop="position" content="5"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>estimating an allele frequency</h1></header><h2 class="anchor anchorWithStickyNavbar_mojV" id="the-frequency-of-o-blood-group">The frequency of O blood group<a class="hash-link" href="#the-frequency-of-o-blood-group" title="Direct link to heading">​</a></h2><p>In this section we will estimate the frequency of O blood group (as encoded by the <a href="https://www.ensembl.org/Homo_sapiens/Variation/Explore?v=rs8176719" target="_blank" rel="noopener noreferrer">common
loss-of-function deletion in the ABO
gene</a>, in multiple populations.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="an-aside-on-the-abo-blood-group-system">An aside on the ABO blood group system.<a class="hash-link" href="#an-aside-on-the-abo-blood-group-system" title="Direct link to heading">​</a></h3><p>The ABO blood group system was the first genetic polymorphism discovered in humans - <a href="https://www.ncbi.nlm.nih.gov/books/NBK2267/" target="_blank" rel="noopener noreferrer">long before
the structure of DNA was solved</a>. It was discovered by
studying agglutination patterns of red cells in serum from other individuals and is of course of
extreme relevance to blood transfusion. Beyond this, however, <em>ABO</em> is an interesting gene; the A/B
split seems to have been preserved <a href="https://www.pnas.org/content/109/45/18493" target="_blank" rel="noopener noreferrer">under balancing
selection</a> across primates, while the O mutation itself
may be a recurrent polymorphism. ABO is also <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5207801/" target="_blank" rel="noopener noreferrer">one of the most pleiotropic loci in the human
genome</a> and there is evidence it is
<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7594382/" target="_blank" rel="noopener noreferrer">associated with risk of SARS-CoV-2
infection</a>. </p><p>But I was interested because it is also <a href="https://www.nature.com/articles/s41467-019-13480-z" target="_blank" rel="noopener noreferrer">protective against
malaria</a>. (The data here comes from that paper).</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="loading-the-data">Loading the data<a class="hash-link" href="#loading-the-data" title="Direct link to heading">​</a></h3><p>First, let&#x27;s load the data.</p><p><strong>Note.</strong> I am using the <a href="https://www.tidyverse.org" target="_blank" rel="noopener noreferrer">tidyverse</a> for these examples. If you don&#x27;t
have it installed, either install it or use base R (e.g. <code>read.csv()</code>) instead.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">library( tidyverse )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data = read_csv( &quot;o_bld_grp.csv&quot; )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><strong>Question</strong>.  What countries are in the data?  What ethnic groups?</p><p>The O blood group data is in the column called <code>O.bld.grp</code>. A <code>1</code> in this column means the
individual has O blood group (which happens if they have two copies of the above rs8176719
deletion). A <code>0</code> generally means they will have either A, AB, or B blood group depending on the
alleles they carry. (There are actually a few other mutations that cause loss of function of the <em>ABO</em> gene,
but we&#x27;re ignoring them here.)</p><p><strong>Question</strong>. Table this data for each population or ethnicity. Can you estimate the frequency in
each population? In each ethnic group?</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="building-a-model">Building a model<a class="hash-link" href="#building-a-model" title="Direct link to heading">​</a></h3><p>You probably estimated the allele frequency as: the number of O individuals divided by the total
number of individuals.  If so congratulations - that&#x27;s a sensible estimate!</p><p>But hang on... some of them vary quite a bit, don&#x27;t they?</p><p><strong>Question.</strong> How much do you trust these estimates? Do you trust them all as much as each other?  Why?</p><p>This course is all about handling uncertainty. We want to know <strong>both</strong> a good estimate, and how
uncertain that estimate is.  This is what statistical models are for!</p><p>Let&#x27;s focus on a single population first - say Tanzania:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">w = which( data$country == &quot;Tanzania&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&gt; table( data$country[w], data$O.bld.grp[w] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">          </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            0  1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  Tanzania 51 47</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Counts like this can be modelled well using a <a target="_blank" href="/whg-training-resources/assets/files/Distributions cheatsheet-cc4e21d39d4e4e0af58815c6c5bc670b.pdf/">binomial distribution</a>.  Let&#x27;s do that now.</p><p><strong>Question</strong>. If we use a binomial distribution to model these counts, what assumptions are we making?</p><p>The binomial distribution takes two parameters: <code>n</code>, the number of &#x27;trials&#x27; (i.e. samples), and <em>θ</em>, the
frequency. (<!-- -->θ<!-- --> is called <em>p</em> on the <a target="_blank" href="/whg-training-resources/assets/files/Probability cheatsheet-2344e889eac5732144ee5396d655e561.pdf/">probability%20cheatsheet</a>, but we&#x27;ll
call it <!-- -->θ<!-- --> here.)</p><p>The data is <em>k</em> - the number of O blood group alleles observed. Our conceptual model is that the population &#x27;emits&#x27; O
blood group alleles at frequency <!-- -->θ<!-- -->. It is this frequency that we want to infer.</p><p>The basic inference formula (Bayes rule) is:</p><p>$$ P(\theta=x|\text{data}) = \frac{P(\text{data}|\theta=x) \cdot P(\theta=x)}{P(\text{data})} $$</p><p>For the moment let us ignore the prior term <em>P(<!-- -->θ<!-- -->=x)</em>. Or to put it another way, we will assume it is uniform.</p><p>The term P(data|<!-- -->θ<!-- -->=x) is our <em>likelihood function</em>. This is what we will model with a binomial distribution. That
is, for the Tanzania data above we will assume:</p><p>$$ P\left(\text{data}|\theta=x\right) = \text{binom}\left( k=47 | n=98, \theta=x \right) $$</p><p><strong>Note.</strong> As in the <a target="_blank" href="/whg-training-resources/assets/files/Probability cheatsheet-2344e889eac5732144ee5396d655e561.pdf/">probability cheatsheet</a> you must remember that all
probability is conditional. In all probabilities above we are assuming this <em>highly unrealistic</em> but <em>hopefully useful</em>
model in which a single frequency parameter governs O blood group in an entire human population. The probabilities
don&#x27;t exist in the real world, but only in the model.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="plotting-the-unnormalised-posterior">Plotting the unnormalised posterior<a class="hash-link" href="#plotting-the-unnormalised-posterior" title="Direct link to heading">​</a></h3><p><strong>Question.</strong> Given the counts for Tanzania above, plot the unnormalised posterior of the parameter <!-- -->θ<!-- -->. Also add
on a line showing your point estimate.</p><p><strong>Note.</strong> Because we are using a flat prior and the normalising constant here, this is the same as plotting the
likelihood function - the only R function you really need for this is <code>dbinom()</code>. For readability it can be nice to
make a friendlier function matching the notation above:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">binomial.likelihood &lt;- function( k, n, theta ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return( dbinom( x = k, size = n, prob = theta ) )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>When you plot this you should see something like this:</p><p><img loading="lazy" alt="img" src="/whg-training-resources/assets/images/Tanzania_o_blood_group_likelihood-76977c5b769a3216541fb08227562812.svg" width="384" height="192" class="img_E7b_">
<strong>Note.</strong> You did label your axes, right?  Hey, you must always label your axes!</p><p><strong>Question.</strong> Can you make a grid of these plots, one per population?  One per ethnicity?</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="plotting-the-fully-normalised-posterior">Plotting the (fully normalised) posterior<a class="hash-link" href="#plotting-the-fully-normalised-posterior" title="Direct link to heading">​</a></h3><p>In principle computing the posterior is not hard. We just need to compute the denominator of Bayes
theorem above, which (using the <em>law of total probability</em> from the <a target="_blank" href="/whg-training-resources/assets/files/Probability cheatsheet-2344e889eac5732144ee5396d655e561.pdf/">probability cheatsheet</a>) is:</p><p>$$ P(\text{data}) = \int_y P(\text{data}|\theta=y) P(\theta=y) $$</p><p>You could for example numerically compute this using the <code>integrate()</code> function - e.g. using the Tanzania counts above:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">f &lt;- function( y ) { return( binomial.likelihood( 47, 98, y ) ) ; }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">denominator = integrate( f, 0, 1 )$value</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>It we now plot the (normalised) posterior, it looks the same as the likelihood but the y axis scale is different:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">plot_normalised_posterior &lt;- function( k, n ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    f &lt;- function( y ) { return( binomial.likelihood( k, n, y ) ) ; }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    denominator = integrate( f, 0, 1 )$value</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    x = seq( from = 0, to = 1, by = 0.01 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    plot(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        x, binomial.likelihood( k = k, n = n, theta = x ) / denominator,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        type = &#x27;l&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        xlab = &quot;O blood group frequency (θ)&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        ylab = &quot;Normalised posterior&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        yaxt = &#x27;n&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        bty = &#x27;n&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    grid()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    abline( v = k/n, col = &#x27;red&#x27; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    axis( 2, las = 1 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot_normalised_posterior( 47, 98 )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="img" src="/whg-training-resources/assets/images/Tanzania_o_blood_group_posterior-5882b1f9c8de9b7da4e86dcdff7b748e.svg" width="384" height="192" class="img_E7b_">
(If you stare at this a bit you&#x27;ll see it looks about right in terms of the total mass under the function - which
should sum to 1).</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="a-beta-way-to-do-it">A Beta way to do it<a class="hash-link" href="#a-beta-way-to-do-it" title="Direct link to heading">​</a></h3><p>In the general case solving the denominator can indeed be hard and require numerical integration like the above.
However, for this binomial model it turns out to be much easier. This is because the binomial has a &#x27;conjugate&#x27; family of prior
distributions - the <a target="_blank" href="/whg-training-resources/assets/files/Distributions cheatsheet-cc4e21d39d4e4e0af58815c6c5bc670b.pdf/">Beta distribution</a>.</p><p>If you stare at the Beta and binomial distributions <a target="_blank" href="/whg-training-resources/assets/files/Distributions cheatsheet-cc4e21d39d4e4e0af58815c6c5bc670b.pdf/">on the cheatsheet</a>
you&#x27;ll see why this works. Using our notation here, the parameter <!-- -->θ<!-- --> (called <em>p</em> on the cheatsheet) enters the
binomial distribution pdf in the term:</p><p>$$ \theta^k (1-\theta)^n-k $$</p><p>on the other hand, the frequency parameter of the Beta distribution (called <em>x</em>) on the cheatsheet) enters its pdf in the
similar term:</p><p>$$ \theta^{\alpha-1} (1-\theta)^{\beta-1} $$</p><p>All the other terms are constant as far as the frequency parameter is concerned. What all this means is that <strong>for a uniform (or
Beta) prior, and a binomial likelihood, the posterior distribution is a Beta distribution</strong>:</p><p>$$ \text{Beta prior} \rightarrow \text{Binomial likelihood} \Rightarrow \text{Beta posterior} $$</p><p>This relationship between prior and likelihood is known as &#x27;conjugacy&#x27; - we say the Beta prior is conjugate to the Binomial
likelihood.</p><p>We can prove this for our data by adding it to our plot:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x = seq( from = 0, to = 1, by = 0.01 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">points( x, dbeta( x, shape1 = 47+1, shape2 = 51+1 ), type = &#x27;l&#x27;, lty = 2, col = &#x27;blue&#x27; )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="img" src="/whg-training-resources/assets/images/Tanzania_o_blood_group_posterior+beta-64925d62c2588a674962821752fa0847.svg" width="432" height="192" class="img_E7b_">
<strong>Note.</strong> The <code>+1s</code> are there in the above because of the way Beta is defined.  (It has exponents with <code>-1s</code>.)</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="comparing-estimates">Comparing estimates<a class="hash-link" href="#comparing-estimates" title="Direct link to heading">​</a></h3><p>Having this analytical version of the posterior is good news. Here are two ways it helps. </p><h4 class="anchor anchorWithStickyNavbar_mojV" id="easily-summarising-our-inference-about-the-parameter">Easily summarising our inference about the parameter.<a class="hash-link" href="#easily-summarising-our-inference-about-the-parameter" title="Direct link to heading">​</a></h4><p>For example, we could summarise our inference by computing a <strong>credible interval</strong> for our parameter. A good interval
to take is a 95% posterior mass interavls, which you can form by chopping of 2.5% of the mass from both tails of the
distribution. If you want to see how to do this, look at the diagram <a target="_blank" href="/whg-training-resources/assets/files/Distributions cheatsheet-cc4e21d39d4e4e0af58815c6c5bc670b.pdf/">on the probability cheatsheet</a>. In R, the beta cdf is given by <code>pbeta()</code>, and its inverse (the
quantile function) by <code>qbeta()</code>. So we can do something like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">compute.credible.interval &lt;- function( k, n, mass = 0.95 ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    tail = 1-mass</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return( qbeta(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        c(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            lower = tail/2,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            median = 0.5,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            upper = 1.0-(tail/2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        shape1 = k+1,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        shape2 = n-k+1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>And let&#x27;s use it to make a function to easily summarise any given subset of the data:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">compute.counts &lt;- function( data ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return( c(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        nonO = length( which( data$O.bld.grp == 0 )),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        O = length( which( data$O.bld.grp == 1 ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">summarise &lt;- function( data.subset, name ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    counts = compute.counts( data.subset )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    credible = compute.credible.interval( counts[2], sum(counts ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return( c(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        list( name = name, nonO = counts[1], O = counts[2], estimate = counts[2]/sum(counts) ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        credible</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><strong>Question</strong> Compute the 95% credible interval for all the populations. Compare to the sample size. (For example - can
you plot sample size vs size of credible interval?). What is the pattern?</p><p><strong>Question.</strong> Do all the 95% credible intervals overlap? What about for individual ethnicities?</p><p><strong>Hint.</strong> A neat way to do this is to use <code>map_dfr()</code> from the <a href="https://github.com/rstudio/cheatsheets/blob/main/purrr.pdf" target="_blank" rel="noopener noreferrer">purr package</a>:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">countries = unique( data$country )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">map_dfr( countries, function(country) { summarise( data[ data$country == country, ], country ) })</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data$country_ethnicity = sprintf( &quot;%s:%s&quot;, data$country, data$ethnicity )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ethnicities = unique( data$country_ethnicity )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">map_dfr( ethnicities, function(ethnicity) { summarise( data[ data$country_ethnicity == ethnicity, ], ethnicity ) })</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We could also add these to our plots - see the code in <a target="_blank" href="/whg-training-resources/assets/files/solutions-e956342871da265a3e7f29cd31b729df.R/"><code>solutions/solutions.R</code></a> for a working
version:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">source( &quot;solutions/solutions.R&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot.bygroup( data, &quot;country_ethnicity&quot; )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="img" src="/whg-training-resources/assets/images/all_ethnicities_o_blood_group_beta_posterior-8f6e1249c4499b4c5fa8f4fff7fa0cda.svg" width="768" height="768" class="img_E7b_"></p><h4 class="anchor anchorWithStickyNavbar_mojV" id="dealing-with-very-uncertain-estimates">Dealing with very uncertain estimates<a class="hash-link" href="#dealing-with-very-uncertain-estimates" title="Direct link to heading">​</a></h4><p>Another way the conjugate prior / analytical formulation helps is in dealing with very uncertain estimates. For
example, the estimate for the <code>AKANS[ASHANTI_EASTERN]</code> has a point estimate of &gt; 0.6 even though most of the others are
&lt; 0.8. But this is based on only 21 observations. And what about the <code>Ghana:NORTHENER</code> or other groups where the
estimate is one or zero, but the counts are tiny?</p><p>One way we can deal with this is give up on the point estimate - just satisfy ourselves that there isn&#x27;t much evidence.
For example, across all samples the joint estimate is <code>0.425</code>. How likely is it that the <code>AKANS[ASHANTI_EASTERN]</code>
estimate is larger than this?</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt; 1 - pbeta( 0.425, shape1 = 9, shape2 = 13 ) </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">[1] 0.4301758</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>There&#x27;s only 43% chance - that&#x27;s a 57% chance the estimate is at or lower than the joint estimate.  </p><p>However, it can nevertheless be useful in lots of contexts to get a sensible point estimate, even when there is little
data. To do this let&#x27;s replace our uniform prior with a <em>weakly informative prior</em>. This prior will say that we believe
O blood group is not at 0 or 100% frequency. If you followed the above, you&#x27;ll see that we can just replace our uniform
prior with a <code>Beta(2,2)</code> prior.  Equivalently, we add 1 to boht the non-O and O counts in the data.  Try this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">plot.bygroup( data, &quot;country_ethnicity&quot;, prior.counts = c(1,1) )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Now none of the estimates are 1 or zero, and all the credible intervals look at least somewhat sensible. Our prior has
<strong>regularised</strong> the estimation.</p><p><strong>Question</strong>. Try adjusting the prior counts and seeing what happens to the estimates. Does it affect all populations
equally? How much prior data do you need to get all the estimates within 0.1 of each other?</p><p><strong>Note.</strong> This technique is often used in practice. For example, the famous <a href="https://www.nature.com/articles/ng1847" target="_blank" rel="noopener noreferrer">EIGENSTRAT
software</a> for computing principal components analysis estimates
allele frequencies in just this way - with a Beta(2,2) prior.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/distributions warmup/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">distributions warmup</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/probability warmup/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">probability warmup</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#the-frequency-of-o-blood-group" class="table-of-contents__link toc-highlight">The frequency of O blood group</a><ul><li><a href="#an-aside-on-the-abo-blood-group-system" class="table-of-contents__link toc-highlight">An aside on the ABO blood group system.</a></li><li><a href="#loading-the-data" class="table-of-contents__link toc-highlight">Loading the data</a></li><li><a href="#building-a-model" class="table-of-contents__link toc-highlight">Building a model</a></li><li><a href="#plotting-the-unnormalised-posterior" class="table-of-contents__link toc-highlight">Plotting the unnormalised posterior</a></li><li><a href="#plotting-the-fully-normalised-posterior" class="table-of-contents__link toc-highlight">Plotting the (fully normalised) posterior</a></li><li><a href="#a-beta-way-to-do-it" class="table-of-contents__link toc-highlight">A Beta way to do it</a></li><li><a href="#comparing-estimates" class="table-of-contents__link toc-highlight">Comparing estimates</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.a63558cc.js"></script>
<script src="/whg-training-resources/assets/js/main.81209970.js"></script>
</body>
</html>