"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[4336],{3905:function(e,t,n){n.d(t,{Zo:function(){return c},kt:function(){return d}});var i=n(7294);function a(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function o(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);t&&(i=i.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,i)}return n}function r(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?o(Object(n),!0).forEach((function(t){a(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):o(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,i,a=function(e,t){if(null==e)return{};var n,i,a={},o=Object.keys(e);for(i=0;i<o.length;i++)n=o[i],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(i=0;i<o.length;i++)n=o[i],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}var l=i.createContext({}),p=function(e){var t=i.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):r(r({},t),e)),n},c=function(e){var t=p(e.components);return i.createElement(l.Provider,{value:t},e.children)},u={inlineCode:"code",wrapper:function(e){var t=e.children;return i.createElement(i.Fragment,{},t)}},m=i.forwardRef((function(e,t){var n=e.components,a=e.mdxType,o=e.originalType,l=e.parentName,c=s(e,["components","mdxType","originalType","parentName"]),m=p(n),d=a,h=m["".concat(l,".").concat(d)]||m[d]||u[d]||o;return n?i.createElement(h,r(r({ref:t},c),{},{components:n})):i.createElement(h,r({ref:t},c))}));function d(e,t){var n=arguments,a=t&&t.mdxType;if("string"==typeof e||a){var o=n.length,r=new Array(o);r[0]=m;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"==typeof e?e:a,r[1]=s;for(var p=2;p<o;p++)r[p]=n[p];return i.createElement.apply(null,r)}return i.createElement.apply(null,n)}m.displayName="MDXCreateElement"},5842:function(e,t,n){n.r(t),n.d(t,{assets:function(){return c},contentTitle:function(){return l},default:function(){return d},frontMatter:function(){return s},metadata:function(){return p},toc:function(){return u}});var i=n(7462),a=n(3366),o=(n(7294),n(3905)),r=["components"],s={},l=void 0,p={unversionedId:"genome_wide_association_studies/Principal components analysis/relatedness_pruning",id:"genome_wide_association_studies/Principal components analysis/relatedness_pruning",title:"relatedness_pruning",description:"Up to the table of contents - Back to the page on LD pruning - Forward to compute the PCs",source:"@site/docs/genome_wide_association_studies/Principal components analysis/relatedness_pruning.md",sourceDirName:"genome_wide_association_studies/Principal components analysis",slug:"/genome_wide_association_studies/Principal components analysis/relatedness_pruning",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/relatedness_pruning",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/genome_wide_association_studies/Principal components analysis/relatedness_pruning.md",tags:[],version:"current",frontMatter:{},sidebar:"tutorialSidebar",previous:{title:"overview",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/overview"},next:{title:"the_maths",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/the_maths"}},c={},u=[{value:"IBD pruning of samples and identification of close relationships",id:"ibd-pruning-of-samples-and-identification-of-close-relationships",level:3},{value:"Picking samples to exclude",id:"picking-samples-to-exclude",level:4},{value:"Computing principal components",id:"computing-principal-components",level:4}],m={toc:u};function d(e){var t=e.components,n=(0,a.Z)(e,r);return(0,o.kt)("wrapper",(0,i.Z)({},m,n,{components:t,mdxType:"MDXLayout"}),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/Introduction"},"Up to the table of contents")," - ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/ld_pruning"},"Back to the page on LD pruning")," - ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/computing_PCs"},"Forward to compute the PCs")),(0,o.kt)("h3",{id:"ibd-pruning-of-samples-and-identification-of-close-relationships"},"IBD pruning of samples and identification of close relationships"),(0,o.kt)("p",null,"We want our top PCs to reflect the relatedness structure across the majority of samples in our\nGWAS dataset. Although all samples in our dataset are nominally unrelated, a few duplicated or\nrelated samples may have slipped in through the sampling process or through sample handling. As\ndescribed in the lectures, this can seriously affect principal components. We'll therefore first\nidentify and remove any close relationships before computing PCs."),(0,o.kt)("p",null,"Let's use plink to compute the relatedness between samples. A simple way to do this would be to use\nthe relatedness matrix that we will construct to compute PCA. However, here we'll use genome-wide\nestimates of identity by descent (IBD) instead. To do this we use the following command:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"$ plink --vcf chr19-clean.vcf.gz --genome gz --out chr19-clean --extract chr19-clean.prune.in\n")),(0,o.kt)("p",null,"Here, the ",(0,o.kt)("inlineCode",{parentName:"p"},"--genome")," option tells plink to compute genome-wide IBD estimates (as this file is rather large, we ask plink to store it in a gzipped file).  We've used the ",(0,o.kt)("inlineCode",{parentName:"p"},"--extract")," option to tell plink to only use the LD-pruned set of SNPs we computed above.  Further details of the options for computing relatedness and file formats are described ",(0,o.kt)("a",{parentName:"p",href:"https://www.cog-genomics.org/plink2/ibd"},"on the plink documentation page"),"."),(0,o.kt)("p",null,"The output file from the command above is in ",(0,o.kt)("inlineCode",{parentName:"p"},"chr19-clean.genome.gz"),". Let's load that into ",(0,o.kt)("inlineCode",{parentName:"p"},"R"),"\nand have a look at it."),(0,o.kt)("p",null,"In R/RStudio:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},'ibd <- read.table("chr19-clean.genome.gz", hea=T, as.is=T)\nView(ibd)\nhist( ibd$PI_HAT, breaks = 100 )\n')),(0,o.kt)("p",null,"You can also zoom in along the y-axis to see any close relationships:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"hist( ibd$PI_HAT, breaks = 100, ylim = c(0,1000) )\n")),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Question"),". What is ",(0,o.kt)("inlineCode",{parentName:"p"},"PI_HAT"),"? Can you figure it out from ",(0,o.kt)("a",{parentName:"p",href:"https://www.cog-genomics.org/plink2/ibd"},"the documentation\npage"),"?"),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Question"),". Which samples have the closest relationship in the dataset?  What is the average relationship between samples in this dataset?"),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Question"),". In the histogram, there's a tail or 'bump' of relationships up to about 0.2.  What could this represent?"),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Note.")," The above is one way to compute relatedness estimates - many other methods are available.\nFor example, ",(0,o.kt)("a",{parentName:"p",href:"https://www.kingrelatedness.com"},"KING")," is a popular choice.  More computationally demanding methods also exist that can ",(0,o.kt)("a",{parentName:"p",href:"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7553009/"},"identify actual segments inherited IBD"),"."),(0,o.kt)("h4",{id:"picking-samples-to-exclude"},"Picking samples to exclude"),(0,o.kt)("p",null,"Before we compute PCs, we'll pick one sample from each closely-related pair and exclude it. For\nnow we'll just do this simply by picking the second sample; a more refined approach might look at\ngenotyping performance across the genome and exclude the sample with greater missingness. (Better\nyet, you could try to pick the fewest number of samples that remove all close relationships, as\nthis can sometimes substantially improve the resulting sample size). We'll store the results in a\nfile that we can tell plink about in later steps."),(0,o.kt)("p",null,"In RStudio:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"# put a line on our plot so we can see what we'll exclude\nabline( v = 0.2, col = 'red' )\n\n# Write a list of excluded samples\nexclusions = ibd[ ibd$PI_HAT > 0.2, c('FID2','IID2') ]\nwrite.table( exclusions, file=\"related_samples.txt\", col.names = F, row.names = F, quote = F )\n")),(0,o.kt)("p",null,"Q. How many samples will be excluded?"),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Note"),": the same sample might appear twice in the exclusions data frame.  This would happen if they are closely related to more than one other sample.  Are there any samples like that?  You can use ",(0,o.kt)("inlineCode",{parentName:"p"},"unique(exclusions)")," to get a list of unique samples to exclude."),(0,o.kt)("h4",{id:"computing-principal-components"},"Computing principal components"),(0,o.kt)("p",null,"When you're ready, go ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/computing_PCs"},"here")," to start the principal component analysis proper."))}d.isMDXComponent=!0}}]);