"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[6164],{3905:function(e,t,n){n.d(t,{Zo:function(){return c},kt:function(){return d}});var a=n(7294);function o(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function r(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function s(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?r(Object(n),!0).forEach((function(t){o(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):r(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function i(e,t){if(null==e)return{};var n,a,o=function(e,t){if(null==e)return{};var n,a,o={},r=Object.keys(e);for(a=0;a<r.length;a++)n=r[a],t.indexOf(n)>=0||(o[n]=e[n]);return o}(e,t);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);for(a=0;a<r.length;a++)n=r[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(o[n]=e[n])}return o}var l=a.createContext({}),p=function(e){var t=a.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):s(s({},t),e)),n},c=function(e){var t=p(e.components);return a.createElement(l.Provider,{value:t},e.children)},u={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},m=a.forwardRef((function(e,t){var n=e.components,o=e.mdxType,r=e.originalType,l=e.parentName,c=i(e,["components","mdxType","originalType","parentName"]),m=p(n),d=o,g=m["".concat(l,".").concat(d)]||m[d]||u[d]||r;return n?a.createElement(g,s(s({ref:t},c),{},{components:n})):a.createElement(g,s({ref:t},c))}));function d(e,t){var n=arguments,o=t&&t.mdxType;if("string"==typeof e||o){var r=n.length,s=new Array(r);s[0]=m;var i={};for(var l in t)hasOwnProperty.call(t,l)&&(i[l]=t[l]);i.originalType=e,i.mdxType="string"==typeof e?e:o,s[1]=i;for(var p=2;p<r;p++)s[p]=n[p];return a.createElement.apply(null,s)}return a.createElement.apply(null,n)}m.displayName="MDXCreateElement"},4585:function(e,t,n){n.r(t),n.d(t,{assets:function(){return c},contentTitle:function(){return l},default:function(){return d},frontMatter:function(){return i},metadata:function(){return p},toc:function(){return u}});var a=n(7462),o=n(3366),r=(n(7294),n(3905)),s=["components"],i={},l=void 0,p={unversionedId:"genome_wide_association_studies/Principal components analysis/global_analysis",id:"genome_wide_association_studies/Principal components analysis/global_analysis",title:"global_analysis",description:"Up to the table of contents - Back to the PC computation step",source:"@site/docs/genome_wide_association_studies/Principal components analysis/global_analysis.md",sourceDirName:"genome_wide_association_studies/Principal components analysis",slug:"/genome_wide_association_studies/Principal components analysis/global_analysis",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/global_analysis",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/genome_wide_association_studies/Principal components analysis/global_analysis.md",tags:[],version:"current",frontMatter:{},sidebar:"tutorialSidebar",previous:{title:"getting_setup",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/getting_setup"},next:{title:"ld_pruning",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/ld_pruning"}},c={},u=[{value:"Plotting samples against a global dataset",id:"plotting-samples-against-a-global-dataset",level:3},{value:"Summary",id:"summary",level:4}],m={toc:u};function d(e){var t=e.components,n=(0,o.Z)(e,s);return(0,r.kt)("wrapper",(0,a.Z)({},m,n,{components:t,mdxType:"MDXLayout"}),(0,r.kt)("p",null,(0,r.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/Introduction"},"Up to the table of contents")," - ",(0,r.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/computing_PCs"},"Back to the PC computation step")),(0,r.kt)("h3",{id:"plotting-samples-against-a-global-dataset"},"Plotting samples against a global dataset"),(0,r.kt)("p",null,"A common strategy for identifying samples with unusual ancestry is to plot samples against a global\nreference dataset, such as that produced by the 1000 Genomes project (1000G). A common way to do\nthis is to compute a PCA of the external samples and project the GWAS dataset onto it. In plink,\nthis can be done in plink using the ",(0,r.kt)("inlineCode",{parentName:"p"},"--pca")," and ",(0,r.kt)("inlineCode",{parentName:"p"},"--pca-cluster-names")," options - see the ",(0,r.kt)("a",{parentName:"p",href:"https://www.cog-genomics.org/plink/"},"plink\nwebsite")," for details). For this practical, however, we'll take\na simpler approach and compute a PCA of all the data together."),(0,r.kt)("p",null,"To do this, we've created a merged dataset ",(0,r.kt)("inlineCode",{parentName:"p"},"merged.with.1000G.vcf.gz")," including our GWAS data with the African (AFR), European (EUR) and East Asian (EAS) samples from the 1000G reference panel.  To compute PCs, we use a similar command as before, changing the dataset and exclusions files:"),(0,r.kt)("pre",null,(0,r.kt)("code",{parentName:"pre"},"plink --vcf merged.with.1000G.vcf.gz --extract chr19-clean.prune.in --pca var-wts --out merged.with.1000G\n")),(0,r.kt)("p",null,"As before this creates files names ",(0,r.kt)("inlineCode",{parentName:"p"},"merged.with.1000G.eigenvec"),", etc."),(0,r.kt)("p",null,"Let's load the merged PCs and plot them with coloured ethnicities as before.  It's worth being careful with colours here, so we'll plot points with a particular colour scheme and use different shapes to distinguish the GWAS and reference panel samples:"),(0,r.kt)("p",null,"In R/RStudio:"),(0,r.kt)("pre",null,(0,r.kt)("code",{parentName:"pre"},'pcs = read.table( "merged.with.1000G.eigenvec" )\ncolnames(pcs)[1:7] = c( "group", "ID", "PC1", "PC2", "PC3", "PC4", "PC5" )\npcs$group = as.character( pcs$group ) # make sure to treat group as strings\n\npalette = c(\n  CAN = "red2", FAN = "green2", JAN = "blue2", RAN = "yellow3",\n  AFR = "grey40", EUR = "grey70", EAS = "purple"\n)\nshapes = c(\n  CAN = 4, FAN = 4, JAN = 4, RAN = 4,\n  AFR = 20, EUR = 20, EAS = 20\n)\n\npcs$colour = palette[pcs$group]\npcs$shape = shapes[pcs$group]\n\npairs( pcs[,3:7], col = pcs$colour, pch = pcs$shape, lower.panel = NULL )\nlegend(\n  0.1, 0.5,\n  col = palette,\n  legend = names(palette),\n  pch = shapes,\n  xpd=NA\n)\n')),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Question"),". Which reference panel group do most of the GWAS dataset samples cluster with?  Are there any that don't?"),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Question"),". Use R to identify the samples that seem to cluster in the wrong place.  Do you recognise these samples?  Which reference panel group do they cluster near?  What do you conclude about these samples?"),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Question"),". Some of the 1000G samples labelled 'AFR' also cluster nearer to the Europeans than others.  Why is this?"),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Hint"),": Population codes can be found in the file [",(0,r.kt)("inlineCode",{parentName:"p"},"resources/1000GP_Phase3.sample"),".  Look up the codes for a couple of samples, and look at the population definitions on the 1000 Genomes website at ",(0,r.kt)("inlineCode",{parentName:"p"},"http://www.1000genomes.org/category/frequently-asked-questions/population"),"."),(0,r.kt)("h4",{id:"summary"},"Summary"),(0,r.kt)("p",null,"Congratulations!  You have completed a basic principal components analysis.  We will use these PCs later when we run the ",(0,r.kt)("a",{parentName:"p",href:"%22../GWAS_analysis_practical%22"},"genome-wide association study practical"),"."))}d.isMDXComponent=!0}}]);