"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[9008],{3905:function(e,n,t){t.d(n,{Zo:function(){return c},kt:function(){return m}});var r=t(7294);function o(e,n,t){return n in e?Object.defineProperty(e,n,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[n]=t,e}function i(e,n){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);n&&(r=r.filter((function(n){return Object.getOwnPropertyDescriptor(e,n).enumerable}))),t.push.apply(t,r)}return t}function a(e){for(var n=1;n<arguments.length;n++){var t=null!=arguments[n]?arguments[n]:{};n%2?i(Object(t),!0).forEach((function(n){o(e,n,t[n])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):i(Object(t)).forEach((function(n){Object.defineProperty(e,n,Object.getOwnPropertyDescriptor(t,n))}))}return e}function s(e,n){if(null==e)return{};var t,r,o=function(e,n){if(null==e)return{};var t,r,o={},i=Object.keys(e);for(r=0;r<i.length;r++)t=i[r],n.indexOf(t)>=0||(o[t]=e[t]);return o}(e,n);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(r=0;r<i.length;r++)t=i[r],n.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(o[t]=e[t])}return o}var l=r.createContext({}),f=function(e){var n=r.useContext(l),t=n;return e&&(t="function"==typeof e?e(n):a(a({},n),e)),t},c=function(e){var n=f(e.components);return r.createElement(l.Provider,{value:n},e.children)},g={inlineCode:"code",wrapper:function(e){var n=e.children;return r.createElement(r.Fragment,{},n)}},u=r.forwardRef((function(e,n){var t=e.components,o=e.mdxType,i=e.originalType,l=e.parentName,c=s(e,["components","mdxType","originalType","parentName"]),u=f(t),m=o,p=u["".concat(l,".").concat(m)]||u[m]||g[m]||i;return t?r.createElement(p,a(a({ref:n},c),{},{components:t})):r.createElement(p,a({ref:n},c))}));function m(e,n){var t=arguments,o=n&&n.mdxType;if("string"==typeof e||o){var i=t.length,a=new Array(i);a[0]=u;var s={};for(var l in n)hasOwnProperty.call(n,l)&&(s[l]=n[l]);s.originalType=e,s.mdxType="string"==typeof e?e:o,a[1]=s;for(var f=2;f<i;f++)a[f]=t[f];return r.createElement.apply(null,a)}return r.createElement.apply(null,t)}u.displayName="MDXCreateElement"},8916:function(e,n,t){t.r(n),t.d(n,{assets:function(){return c},contentTitle:function(){return l},default:function(){return m},frontMatter:function(){return s},metadata:function(){return f},toc:function(){return g}});var r=t(7462),o=t(3366),i=(t(7294),t(3905)),a=["components"],s={},l=void 0,f={unversionedId:"programming/programming_with_gene_annotations/List_of_ensembl_files",id:"programming/programming_with_gene_annotations/List_of_ensembl_files",title:"List_of_ensembl_files",description:"A list of relevant GFF files.",source:"@site/docs/programming/programming_with_gene_annotations/List_of_ensembl_files.md",sourceDirName:"programming/programming_with_gene_annotations",slug:"/programming/programming_with_gene_annotations/List_of_ensembl_files",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/List_of_ensembl_files",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/programming/programming_with_gene_annotations/List_of_ensembl_files.md",tags:[],version:"current",frontMatter:{},sidebar:"tutorialSidebar",previous:{title:"Getting_organised_and_scaling_up",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_organised_and_scaling_up"},next:{title:"Solutions",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Solutions"}},c={},g=[{value:"A list of relevant GFF files.",id:"a-list-of-relevant-gff-files",level:2}],u={toc:g};function m(e){var n=e.components,s=(0,o.Z)(e,a);return(0,i.kt)("wrapper",(0,r.Z)({},u,s,{components:n,mdxType:"MDXLayout"}),(0,i.kt)("h2",{id:"a-list-of-relevant-gff-files"},"A list of relevant GFF files."),(0,i.kt)("p",null,"A list of gff files from ensembl can be found in ",(0,i.kt)("a",{target:"_blank",href:t(5647).Z},(0,i.kt)("code",null,"solutions/ensembl_gff_listing.txt")),"."),(0,i.kt)("p",null,"If you look ",(0,i.kt)("a",{parentName:"p",href:"http://ftp.ensembl.org/pub/current_gff3/"},"on ensembl")," you will find lots of these gff files - I have selected some as follows:"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},"if a species has a file of the form ",(0,i.kt)("inlineCode",{parentName:"li"},"[species].[version].chr.gff3.gz")," then I used that.  "),(0,i.kt)("li",{parentName:"ul"},"otherwise, I used one of the form ",(0,i.kt)("inlineCode",{parentName:"li"},"[species].[version].gff3.gz"),".")),(0,i.kt)("p",null,"The idea is to get annotations for the whole genome.  Genome assemblies typically have a set of core 'primary' contigs, and then might have a set of additional contigs.\nI ",(0,i.kt)("em",{parentName:"p"},"think")," (but not sure) that the ",(0,i.kt)("inlineCode",{parentName:"p"},".chr.gff.gz")," are the annotations for the main reference chromosomes, so have picked these."),(0,i.kt)("p",null,"You should read the README files for the annotations for details."))}m.isMDXComponent=!0},5647:function(e,n,t){n.Z=t.p+"assets/files/ensembl_gff_listing-8ed1d53a7804ccc737923326a7fbf7be.txt"}}]);