"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[7363],{3905:function(e,t,n){n.d(t,{Zo:function(){return p},kt:function(){return m}});var r=n(7294);function i(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function o(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function a(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?o(Object(n),!0).forEach((function(t){i(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):o(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,r,i=function(e,t){if(null==e)return{};var n,r,i={},o=Object.keys(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||(i[n]=e[n]);return i}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(i[n]=e[n])}return i}var l=r.createContext({}),g=function(e){var t=r.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):a(a({},t),e)),n},p=function(e){var t=g(e.components);return r.createElement(l.Provider,{value:t},e.children)},u={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},h=r.forwardRef((function(e,t){var n=e.components,i=e.mdxType,o=e.originalType,l=e.parentName,p=s(e,["components","mdxType","originalType","parentName"]),h=g(n),m=i,c=h["".concat(l,".").concat(m)]||h[m]||u[m]||o;return n?r.createElement(c,a(a({ref:t},p),{},{components:n})):r.createElement(c,a({ref:t},p))}));function m(e,t){var n=arguments,i=t&&t.mdxType;if("string"==typeof e||i){var o=n.length,a=new Array(o);a[0]=h;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"==typeof e?e:i,a[1]=s;for(var g=2;g<o;g++)a[g]=n[g];return r.createElement.apply(null,a)}return r.createElement.apply(null,n)}h.displayName="MDXCreateElement"},9407:function(e,t,n){n.r(t),n.d(t,{assets:function(){return p},contentTitle:function(){return l},default:function(){return m},frontMatter:function(){return s},metadata:function(){return g},toc:function(){return u}});var r=n(7462),i=n(3366),o=(n(7294),n(3905)),a=["components"],s={sidebar_position:9},l=void 0,g={unversionedId:"programming/programming_with_gene_annotations/Counting_genes_3",id:"programming/programming_with_gene_annotations/Counting_genes_3",title:"Counting_genes_3",description:"Up to table of contents",source:"@site/docs/programming/programming_with_gene_annotations/Counting_genes_3.md",sourceDirName:"programming/programming_with_gene_annotations",slug:"/programming/programming_with_gene_annotations/Counting_genes_3",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/programming/programming_with_gene_annotations/Counting_genes_3.md",tags:[],version:"current",sidebarPosition:9,frontMatter:{sidebar_position:9},sidebar:"tutorialSidebar",previous:{title:"Counting_genes_2",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2"},next:{title:"Getting_sequence_lengths",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths"}},p={},u=[{value:"Extremely big, small and complex genes",id:"extremely-big-small-and-complex-genes",level:2},{value:"Single-exon genes",id:"single-exon-genes",level:3},{value:"Visualising",id:"visualising",level:3},{value:"Getting sequence lengths",id:"getting-sequence-lengths",level:3}],h={toc:u};function m(e){var t=e.components,s=(0,i.Z)(e,a);return(0,o.kt)("wrapper",(0,r.Z)({},h,s,{components:t,mdxType:"MDXLayout"}),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/"},"Up to table of contents")),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2"},"Back to the previous page")),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths"},"On to the next page")),(0,o.kt)("p",null,"Let's use our data to answer some questions on the extremes of the distribution of genes."),(0,o.kt)("h2",{id:"extremely-big-small-and-complex-genes"},"Extremely big, small and complex genes"),(0,o.kt)("p",null,"What is the largest gene, the gene with the most exons, or the gene with the most transcripts in each species?"),(0,o.kt)("p",null,"You can work this out, but here are my solutions for reference:\n",(0,o.kt)("a",{target:"_blank",href:n(5397).Z},"sql solution"),"\n",(0,o.kt)("a",{target:"_blank",href:n(2350).Z},"python solution")),(0,o.kt)("p",null,"Here are some things I noticed:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"In all the data I looked at, there's only one gene with > 100 transcripts. It is the human gene ",(0,o.kt)("a",{parentName:"p",href:"https://www.uniprot.org/uniprot/P53779"},(0,o.kt)("em",{parentName:"a"},"MAPK10"),', "Mitogen-activated\nprotein kinase 10"')," and it has 151 transcripts! It has so many that it takes a while to load in\n",(0,o.kt)("a",{parentName:"p",href:"http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000109339;r=4:85990007-86594625"},"Ensembl")," and the\n",(0,o.kt)("a",{parentName:"p",href:"https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr4%3A86016491%2D86594110&hgsid=275896231_HieWdPQTMOsgYQUFAnwTALgAECs0"},"UCSC Genome browser"),". If you explore these genome browsers you will find a wealth of\ninformation on this gene including its expression patterns and function. Is there much published literature about it?\nWhat does it look like in other species?")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"Quite a few genes have lots of exons, but one that really stands out is ",(0,o.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Titin"},(0,o.kt)("em",{parentName:"a"},"Titin")),". It has a slightly incredible 184 exons in Chimpanzees (",(0,o.kt)("em",{parentName:"p"},"Pan troglodytes"),") and ~113 in humans. Another is ",(0,o.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Nebulin"},(0,o.kt)("em",{parentName:"a"},"Nebulin")),". These genes seem\nespecially huge in Chimpanzees. Are they big in all species? In all great apes? Is there literature that might\nexplain the results for these genes, or other genes with lots of exons?")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"The genes with lots of exons aren't the largest though (by length). In humans and chimpanzees, that accolade goes\nto ",(0,o.kt)("a",{parentName:"p",href:"http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000078328;r=16:5239802-7713340"},(0,o.kt)("inlineCode",{parentName:"a"},"RBFOX1"))," with a whopping length of around 2 and a half megabases, or 0.08% of the genome. What does this look\nlike across species? Where is it expressed? (Interestingly, this gene has been previously associated ",(0,o.kt)("a",{parentName:"p",href:"https://www.nature.com/articles/s41380-018-0068-7"},"with aggressive\nbehaviour"),").")),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("p",{parentName:"li"},"There are also some pretty tiny genes around. In humans the smallest is\n",(0,o.kt)("a",{parentName:"p",href:"http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000288608;r=7:151061928-151061978;t=ENST00000674552"},(0,o.kt)("em",{parentName:"a"},"ENSG00000288608")),",\nwhich is only 50 base pairs long. (However, it seems to be contained in an exon of the much\nlonger ","[",(0,o.kt)("em",{parentName:"p"},"SLC4A2")," Anion exhange protein\n2]","(",(0,o.kt)("a",{parentName:"p",href:"https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType"},"https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType"),"\n=default&virtMode=0&nonVirtPosition=&position=chr7%3A151061877%2D151062029&hgsid=275886241_mxgWOGDr4elcf2SdW0zGz6ukwVJ\nW).  I'm not sure what this means...)"))),(0,o.kt)("p",null,"What other extreme genes are there?"),(0,o.kt)("h3",{id:"single-exon-genes"},"Single-exon genes"),(0,o.kt)("p",null,"What proportion of protein-coding genes in each species have a single exon? (My solution is also in\n",(0,o.kt)("a",{target:"_blank",href:n(5397).Z},(0,o.kt)("code",null,"solutions/part3/solutions.sql"))," and\n",(0,o.kt)("a",{target:"_blank",href:n(2350).Z},(0,o.kt)("code",null,"solutions/part3/solutions.py")),"). Almost 44% of ",(0,o.kt)("a",{parentName:"p",href:"https://en.wikipedia.org/wiki/Plasmodium_falciparum"},"malaria\nparasite")," genes have a single exon, but the number is much lower\nin other species I've looked at."),(0,o.kt)("h3",{id:"visualising"},"Visualising"),(0,o.kt)("p",null,"Visualisation (plotting) is beyond the scope of this tutorial - but you probably know how to do it already. It would be\ninteresting to visualise some of these statistics, for example"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"histogram #transcripts, gene length, # of exons in each species"),(0,o.kt)("li",{parentName:"ul"},"plot the number of exons (or number of transcripts) versus gene length by species")),(0,o.kt)("p",null,"and so on. The sky is the limit in terms of what you could investigate here - every gene has its own story. We'll talk\nthrough your findings in the discussion session."),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Note.")," Plotting is one thing you can't do in SQL - load the data into python or R for this. For reference, you can load\ndata into R using this type of code:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},'library( RSQLite )\ndb = dbConnect( dbDriver( "SQLite" ), "genes.sqlite" )\ndata = dbGetQuery( db, "SELECT * FROM gene_summary_view" )\n')),(0,o.kt)("h3",{id:"getting-sequence-lengths"},"Getting sequence lengths"),(0,o.kt)("p",null,"To go further than this and compute the amount of genome that is covered by genes, we will need to compute genome lengths.\nThe ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths"},"next page")," shows one way to do this."))}m.isMDXComponent=!0},2350:function(e,t,n){t.Z=n.p+"assets/files/solutions-ec50e4e517140870f43279535b47e66f.py"},5397:function(e,t,n){t.Z=n.p+"assets/files/solutions-3795ac42e16132bd667f971c8e746408.sql"}}]);