"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[4875],{3905:function(e,t,n){n.d(t,{Zo:function(){return c},kt:function(){return h}});var a=n(7294);function r(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function o(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){r(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,a,r=function(e,t){if(null==e)return{};var n,a,r={},i=Object.keys(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||(r[n]=e[n]);return r}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(a=0;a<i.length;a++)n=i[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(r[n]=e[n])}return r}var p=a.createContext({}),l=function(e){var t=a.useContext(p),n=t;return e&&(n="function"==typeof e?e(t):o(o({},t),e)),n},c=function(e){var t=l(e.components);return a.createElement(p.Provider,{value:t},e.children)},m={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},u=a.forwardRef((function(e,t){var n=e.components,r=e.mdxType,i=e.originalType,p=e.parentName,c=s(e,["components","mdxType","originalType","parentName"]),u=l(n),h=r,d=u["".concat(p,".").concat(h)]||u[h]||m[h]||i;return n?a.createElement(d,o(o({ref:t},c),{},{components:n})):a.createElement(d,o({ref:t},c))}));function h(e,t){var n=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var i=n.length,o=new Array(i);o[0]=u;var s={};for(var p in t)hasOwnProperty.call(t,p)&&(s[p]=t[p]);s.originalType=e,s.mdxType="string"==typeof e?e:r,o[1]=s;for(var l=2;l<i;l++)o[l]=n[l];return a.createElement.apply(null,o)}return a.createElement.apply(null,n)}u.displayName="MDXCreateElement"},7640:function(e,t,n){n.r(t),n.d(t,{assets:function(){return c},contentTitle:function(){return p},default:function(){return h},frontMatter:function(){return s},metadata:function(){return l},toc:function(){return m}});var a=n(7462),r=n(3366),i=(n(7294),n(3905)),o=["components"],s={},p=void 0,l={unversionedId:"genome_wide_association_studies/Principal components analysis/the_maths",id:"genome_wide_association_studies/Principal components analysis/the_maths",title:"the_maths",description:"Up to the table of contents - Back to the page on computing PCs",source:"@site/docs/genome_wide_association_studies/Principal components analysis/the_maths.md",sourceDirName:"genome_wide_association_studies/Principal components analysis",slug:"/genome_wide_association_studies/Principal components analysis/the_maths",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/the_maths",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/genome_wide_association_studies/Principal components analysis/the_maths.md",tags:[],version:"current",frontMatter:{},sidebar:"tutorialSidebar",previous:{title:"relatedness_pruning",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/relatedness_pruning"},next:{title:"Next-generation sequencing",permalink:"/whg-training-resources/next_generation_sequencing/"}},c={},m=[{value:"An aside on the maths",id:"an-aside-on-the-maths",level:3},{value:"Back to the computation...",id:"back-to-the-computation",level:4}],u={toc:m};function h(e){var t=e.components,n=(0,r.Z)(e,o);return(0,i.kt)("wrapper",(0,a.Z)({},u,n,{components:t,mdxType:"MDXLayout"}),(0,i.kt)("p",null,(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/Introduction"},"Up to the table of contents")," - ",(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/computing_PCs"},"Back to the page on computing PCs")),(0,i.kt)("h3",{id:"an-aside-on-the-maths"},"An aside on the maths"),(0,i.kt)("p",null,"The maths of PCA (in the way usually applied) works like this. Suppose ",(0,i.kt)("inlineCode",{parentName:"p"},"X")," is a big matrix of\ngenotypes at the ",(0,i.kt)("inlineCode",{parentName:"p"},"L")," genetic variants (rows) and ",(0,i.kt)("inlineCode",{parentName:"p"},"N")," samples (in columns)."),(0,i.kt)("p",null,"It is typical to always standardise the genotypes at by dividing by each variant (row) in ",(0,i.kt)("em",{parentName:"p"},"X")," by its standard deviation, i.e. by $$ \\sqrt{f(1-f)} $$ where ",(0,i.kt)("em",{parentName:"p"},"f")," is the\nallele frequency of the variant - and then subtracting the mean.  So do that first."),(0,i.kt)("p",null,"Now there are two possible square matrices you can make out of ",(0,i.kt)("em",{parentName:"p"},"X"),":"),(0,i.kt)("ol",null,(0,i.kt)("li",{parentName:"ol"},"You can form the ",(0,i.kt)("em",{parentName:"li"},"relatedness matrix"),":")),(0,i.kt)("p",null,"$$ R=\\frac{1}{L} X^t X $$."),(0,i.kt)("p",null,"  ",(0,i.kt)("em",{parentName:"p"},"R")," is an ",(0,i.kt)("em",{parentName:"p"},"N ","\xd7"," N")," matrix i.e. has one row and one column for each\nsample. The motivation is that each entry ",(0,i.kt)("em",{parentName:"p"},"r",(0,i.kt)("sub",null,"i,j"))," captures the degree of allele sharing\n(covariance) between individual i and j. Also, because of the frequency scaling, sharing of rarer\nvariants has greater weight ",(0,i.kt)("em",{parentName:"p"},"r",(0,i.kt)("sub",null,"i,j")),"."),(0,i.kt)("ol",{start:2},(0,i.kt)("li",{parentName:"ol"},"Or you can form the ",(0,i.kt)("em",{parentName:"li"},"LD matrix"),":")),(0,i.kt)("p",null,"$$ Z=\\frac{1}{N} X X^t $$."),(0,i.kt)("p",null,"Again because of the scaling to unit variance, entry ",(0,i.kt)("em",{parentName:"p"},"z",(0,i.kt)("sub",null,"i,j"))," is the LD (correlation) between genotypes at variants i and j, and the diagonal entries are ",(0,i.kt)("em",{parentName:"p"},"1"),"."),(0,i.kt)("p",null,"If you followed the tutorial this morning you will realise that ",(0,i.kt)("em",{parentName:"p"},"relatedness")," and ",(0,i.kt)("em",{parentName:"p"},"LD")," are in some\nsense dual to each other. This duality appears in principal components analysis too. Namely:"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},(0,i.kt)("p",{parentName:"li"},"The right eigenvectors of ",(0,i.kt)("em",{parentName:"p"},"Z")," are the ",(0,i.kt)("em",{parentName:"p"},"principal component loading vectors")," of ",(0,i.kt)("em",{parentName:"p"},"X"),".")),(0,i.kt)("li",{parentName:"ul"},(0,i.kt)("p",{parentName:"li"},"The right eigenvectors of ",(0,i.kt)("em",{parentName:"p"},"R")," are (proportional to) the ",(0,i.kt)("em",{parentName:"p"},"principal component vectors"),' (or just\n"principal components") of ',(0,i.kt)("em",{parentName:"p"},"X"),".")),(0,i.kt)("li",{parentName:"ul"},(0,i.kt)("p",{parentName:"li"},"The two are closely related, since the projection of each sample's genotypes onto the ",(0,i.kt)("em",{parentName:"p"},"i"),"th principal component loading vector, is the ",(0,i.kt)("em",{parentName:"p"},"i"),"th principal component for that sample."))),(0,i.kt)("p",null,"The key property that makes these vectors useful is this: the first loading vector picks out the\ndirection in genotype space (a linear combinations of SNPs) so that the first principal component\nhas the maximum possible variance. The second loading vector then picks out the direction in\ngenotype space ",(0,i.kt)("em",{parentName:"p"},"orthogonal to the first")," that makes the second principal component have the maximum\npossible variance - and so on. These vectors thus pick out ",(0,i.kt)("em",{parentName:"p"},"the directions of greatest variance in\nthe genotype data"),"."),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Note.")," There are two sources of LD that could turn up here. The first derives from genetic drift\nor directional selection, which (as we discussed this morning) causes LD between local SNPs on a\nchromosome. The second is LD deriving from population structure. This type of LD can occur between\nSNPs anywhere in the genome - for example, it would be seen between any SNPs that differ in\nfrequency between sub-populations. Because we have ld-thinned our data, it is primarily this type\nof population structure-driven LD that will be picked up by our principal components analysis."),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Computation in practice"),". Typically ",(0,i.kt)("em",{parentName:"p"},"Z")," is huge, while ",(0,i.kt)("em",{parentName:"p"},"R")," is somewhat smaller.  So tools like\n",(0,i.kt)("inlineCode",{parentName:"p"},"plink")," compute ",(0,i.kt)("em",{parentName:"p"},"R"),", use this to compute the principal components, and then compute loadings using\na second pass through the data."),(0,i.kt)("p",null,"You can easily do this in R as well, for example:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},'X = (load data here, convert to a matrix and standardise...)\nL = nrow(X)\nR = (1/L) * t(X) %*% X\nPCs = eigen(R)$vectors\nplot( PCs[,1], PCs[,2], xlab = "PC 1", ylab = "PC 2" )\n# etc.\n')),(0,i.kt)("p",null,"The ",(0,i.kt)("inlineCode",{parentName:"p"},"eigen()")," step will take quite a while on any real dataset, but is generally manageable up to around a few thousand samples."),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Computing in big cohorts"),". While the above works in many studies, very large cohorts such as the\n[UK Biobank[(",(0,i.kt)("a",{parentName:"p",href:"https://www.nature.com/articles/s41586-018-0579-z"},"https://www.nature.com/articles/s41586-018-0579-z"),") typically require other methods such as ",(0,i.kt)("a",{parentName:"p",href:"https://github.com/gabraham/flashpca"},"flashPCA")," that avoid computing either of the matrices directly."),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Note.")," If all this hasn't melted your brain, ",(0,i.kt)("a",{parentName:"p",href:"https://doi.org/10.1371/journal.pgen.1000686"},"try reading\nGil McVean's paper on genealogies and PCA"),"."),(0,i.kt)("h4",{id:"back-to-the-computation"},"Back to the computation..."),(0,i.kt)("p",null,"When you're ready, ",(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/computing_PCs"},"go back to the page on computing PCs"),"."))}h.isMDXComponent=!0}}]);