"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[6658],{3905:function(e,n,t){t.d(n,{Zo:function(){return c},kt:function(){return m}});var i=t(7294);function a(e,n,t){return n in e?Object.defineProperty(e,n,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[n]=t,e}function r(e,n){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);n&&(i=i.filter((function(n){return Object.getOwnPropertyDescriptor(e,n).enumerable}))),t.push.apply(t,i)}return t}function o(e){for(var n=1;n<arguments.length;n++){var t=null!=arguments[n]?arguments[n]:{};n%2?r(Object(t),!0).forEach((function(n){a(e,n,t[n])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):r(Object(t)).forEach((function(n){Object.defineProperty(e,n,Object.getOwnPropertyDescriptor(t,n))}))}return e}function s(e,n){if(null==e)return{};var t,i,a=function(e,n){if(null==e)return{};var t,i,a={},r=Object.keys(e);for(i=0;i<r.length;i++)t=r[i],n.indexOf(t)>=0||(a[t]=e[t]);return a}(e,n);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);for(i=0;i<r.length;i++)t=r[i],n.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(a[t]=e[t])}return a}var l=i.createContext({}),u=function(e){var n=i.useContext(l),t=n;return e&&(t="function"==typeof e?e(n):o(o({},n),e)),t},c=function(e){var n=u(e.components);return i.createElement(l.Provider,{value:n},e.children)},g={inlineCode:"code",wrapper:function(e){var n=e.children;return i.createElement(i.Fragment,{},n)}},p=i.forwardRef((function(e,n){var t=e.components,a=e.mdxType,r=e.originalType,l=e.parentName,c=s(e,["components","mdxType","originalType","parentName"]),p=u(t),m=a,h=p["".concat(l,".").concat(m)]||p[m]||g[m]||r;return t?i.createElement(h,o(o({ref:n},c),{},{components:t})):i.createElement(h,o({ref:n},c))}));function m(e,n){var t=arguments,a=n&&n.mdxType;if("string"==typeof e||a){var r=t.length,o=new Array(r);o[0]=p;var s={};for(var l in n)hasOwnProperty.call(n,l)&&(s[l]=n[l]);s.originalType=e,s.mdxType="string"==typeof e?e:a,o[1]=s;for(var u=2;u<r;u++)o[u]=t[u];return i.createElement.apply(null,o)}return i.createElement.apply(null,t)}p.displayName="MDXCreateElement"},7474:function(e,n,t){t.r(n),t.d(n,{assets:function(){return c},contentTitle:function(){return l},default:function(){return m},frontMatter:function(){return s},metadata:function(){return u},toc:function(){return g}});var i=t(7462),a=t(3366),r=(t(7294),t(3905)),o=["components"],s={sidebar_position:6},l="Viewing alignments",u={unversionedId:"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments",id:"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments",title:"Viewing alignments",description:"If you got this far, you should have a file called QG0033-C.bam containing aligned,",source:"@site/docs/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments.md",sourceDirName:"next_generation_sequencing/Introduction to next-generation sequencing data analysis",slug:"/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments",permalink:"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments.md",tags:[],version:"current",sidebarPosition:6,frontMatter:{sidebar_position:6},sidebar:"tutorialSidebar",previous:{title:"Aligning reads",permalink:"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Aligning_reads"},next:{title:"Challenge questions",permalink:"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Challenge_questions"}},c={},g=[{value:"Viewing alignments with <code>tview</code>",id:"viewing-alignments-with-tview",level:2},{value:"Next steps",id:"next-steps",level:3}],p={toc:g};function m(e){var n=e.components,s=(0,a.Z)(e,o);return(0,r.kt)("wrapper",(0,i.Z)({},p,s,{components:n,mdxType:"MDXLayout"}),(0,r.kt)("h1",{id:"viewing-alignments"},"Viewing alignments"),(0,r.kt)("p",null,"If you got this far, you should have a file called ",(0,r.kt)("inlineCode",{parentName:"p"},"QG0033-C.bam")," containing aligned,\nduplicate-marked reads.  And we'd like to see how they stack up on the genome."),(0,r.kt)("h2",{id:"viewing-alignments-with-tview"},"Viewing alignments with ",(0,r.kt)("inlineCode",{parentName:"h2"},"tview")),(0,r.kt)("p",null,"The simplest way to do this is using ",(0,r.kt)("inlineCode",{parentName:"p"},"samtools tview"),". This is a terminal program which shows you a\nsimple picture of how the read alignments stack up.  Let's try this now by running in the terminal:"),(0,r.kt)("pre",null,(0,r.kt)("code",{parentName:"pre"},"samtools tview -p Pf3D7_07_v3:403818 QG0033-C.bam\n")),(0,r.kt)("p",null,"You should see something like this:"),(0,r.kt)("p",null,(0,r.kt)("img",{alt:"img",src:t(707).Z,width:"1342",height:"838"})),(0,r.kt)("p",null,"What's shown are the reads as they align to the reference sequence. (Reads that align on the\nforward strand are in upper case, and reads that aligned to the reverse strand are shown in\nlower-case.) The numbers along the top are of course the reference genome position (represented at\nthe leftmost digit; so position 403,618 is at the left of the screen).  "),(0,r.kt)("p",null,"You can use arrow keys to move around, ",(0,r.kt)("inlineCode",{parentName:"p"},"g")," to jump to another genome location, or ",(0,r.kt)("inlineCode",{parentName:"p"},"m")," / ",(0,r.kt)("inlineCode",{parentName:"p"},"b")," / ",(0,r.kt)("inlineCode",{parentName:"p"},"n"),"\nto colour reads by mapping quality, base quality, or nucleotide. Press ",(0,r.kt)("inlineCode",{parentName:"p"},"?")," for more help. The ",(0,r.kt)("inlineCode",{parentName:"p"},"q"),"\nkey quits the program."),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Question.")," Can you see:"),(0,r.kt)("ul",null,(0,r.kt)("li",{parentName:"ul"},"A sequencing error?"),(0,r.kt)("li",{parentName:"ul"},"A SNP or other genetic variant? "),(0,r.kt)("li",{parentName:"ul"},"Something that looks like a misalignment?")),(0,r.kt)("p",null,"The answer to the above is 'probably not' at the moment. This becomes much easier when we include the reference sequence\nitself. Press ",(0,r.kt)("inlineCode",{parentName:"p"},"q")," to quit, and now try this command:"),(0,r.kt)("pre",null,(0,r.kt)("code",{parentName:"pre"},"samtools tview -p Pf3D7_07_v3:403818 QG0033-C.bam Pf3D7_v3.fa\n")),(0,r.kt)("p",null,(0,r.kt)("em",{parentName:"p"},"Now")," you'll see something more like this:\n",(0,r.kt)("img",{alt:"img",src:t(3268).Z,width:"1632",height:"950"})),(0,r.kt)("p",null,"Now dots and commas mean 'reads with the same base as the reference' (dots for forward strand,\ncommas for reverse strand) - anything else means a mismatching base.  "),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Question.")," scroll around a bit. Can you find a probable sequencing error? A SNP or other genetic\nvariant? A possible misalignment?"),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Note.")," I focussed on this particular region because it's within the ",(0,r.kt)("em",{parentName:"p"},"CRT")," gene on chromosome 6.\nMutations in these gene at 403,618-403,625 are confer resistance to the antimalarial chloroquine.\nDoes this parasite have these mutations?"),(0,r.kt)("p",null,(0,r.kt)("strong",{parentName:"p"},"Note.")," If you look at other variable sites - for example, Pf3D7_02_v3:631190 which is one of the\nalleles ",(0,r.kt)("a",{parentName:"p",href:"https://doi.org/10.1038/s41586-021-04288-3"},"thought to be associated with infections of individuals carrying the sickle\nhaemoglobin")," - you'll see something a bit funny. It\nlooks like this may be a ",(0,r.kt)("strong",{parentName:"p"},"mixed infection"),". (That is, it might be two or more strains of parasite\nin the same infection. This happens if someone is bitten twice by two infected mosquitos, or if the\nmosquito itself has a mixture of parasites.) I'm thinking this because around half the reads have\nan ",(0,r.kt)("inlineCode",{parentName:"p"},"A")," allele and the other half have a ",(0,r.kt)("inlineCode",{parentName:"p"},"T")," allele at the site. This is also seen at other\nsickle-associated alleles like ",(0,r.kt)("inlineCode",{parentName:"p"},"Pf3D7_02_v3:814288"),"."),(0,r.kt)("h3",{id:"next-steps"},"Next steps"),(0,r.kt)("p",null,"Now go ",(0,r.kt)("a",{parentName:"p",href:"/whg-training-resources/next_generation_sequencing/Introduction%20to%20next-generation%20sequencing%20data%20analysis/Pipeline_outline#the-practical-in-a-nutshell"},"back to the practical")," and try the ",(0,r.kt)("a",{parentName:"p",href:"/whg-training-resources/next_generation_sequencing/Introduction%20to%20next-generation%20sequencing%20data%20analysis/Challenge_questions"},"challenge\nquestions"),"."))}m.isMDXComponent=!0},707:function(e,n,t){n.Z=t.p+"assets/images/samtools_tview-38773c686c64b2ddf2fb4256062bbde1.png"},3268:function(e,n,t){n.Z=t.p+"assets/images/samtools_tview_ref-ed954a9bae98674b0cf0763ff9ad31a4.png"}}]);