"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[53],{1109:function(i){i.exports=JSON.parse('{"pluginId":"default","version":"current","label":"Next","banner":null,"badge":false,"className":"docs-version-current","isLast":true,"docsSidebars":{"tutorialSidebar":[{"type":"link","label":"Overview of the tutorials","href":"/whg-training-resources/overview","docId":"overview"},{"type":"category","label":"Getting your environment setup","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Setting up a UNIX terminal","href":"/whg-training-resources/prerequisites/UNIX","docId":"prerequisites/UNIX"},{"type":"link","label":"Setting up Conda","href":"/whg-training-resources/prerequisites/CONDA","docId":"prerequisites/CONDA"},{"type":"link","label":"R and RStudio","href":"/whg-training-resources/prerequisites/R","docId":"prerequisites/R"},{"type":"link","label":"The R `tidyverse`","href":"/whg-training-resources/prerequisites/tidyverse","docId":"prerequisites/tidyverse"},{"type":"link","label":"Python","href":"/whg-training-resources/prerequisites/python","docId":"prerequisites/python"},{"type":"link","label":"Julia","href":"/whg-training-resources/prerequisites/julia","docId":"prerequisites/julia"},{"type":"link","label":"Using R, python and julia with JupyterLab","href":"/whg-training-resources/prerequisites/Jupyterlab","docId":"prerequisites/Jupyterlab"},{"type":"category","label":"Other languages","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"C++","href":"/whg-training-resources/prerequisites/Other_languages/C++","docId":"prerequisites/Other_languages/C++"},{"type":"link","label":"Javascript","href":"/whg-training-resources/prerequisites/Other_languages/javascript","docId":"prerequisites/Other_languages/javascript"}],"href":"/whg-training-resources/prerequisites/Other_languages/"}],"href":"/whg-training-resources/prerequisites/"},{"type":"category","label":"Population genetics","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Implementing a Wright-Fisher model with a finite number of SNPs","href":"/whg-training-resources/population_genetics/implementing_a_Wright-Fisher_model","docId":"population_genetics/implementing_a_Wright-Fisher_model"}],"href":"/whg-training-resources/population_genetics/"},{"type":"category","label":"Genome-wide association analysis","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Genome-wide association study practical","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Conducting GWAS analysis using plink","href":"/whg-training-resources/genome_wide_association_studies/Genome-wide association analysis/GWAS_analysis_practical (plink version)","docId":"genome_wide_association_studies/Genome-wide association analysis/GWAS_analysis_practical (plink version)"},{"type":"link","label":"Conducting GWAS analysis using QCTOOL and SNPTEST","href":"/whg-training-resources/genome_wide_association_studies/Genome-wide association analysis/GWAS_analysis_practical (qctool+snptest version)","docId":"genome_wide_association_studies/Genome-wide association analysis/GWAS_analysis_practical (qctool+snptest version)"}],"href":"/whg-training-resources/genome_wide_association_studies/Genome-wide association analysis/"},{"type":"category","label":"Meta-analysis and fine-mapping tutorial","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Fine-mapping","href":"/whg-training-resources/genome_wide_association_studies/Meta-analysis and fine-mapping/Fine-mapping","docId":"genome_wide_association_studies/Meta-analysis and fine-mapping/Fine-mapping"},{"type":"link","label":"Introduction","href":"/whg-training-resources/genome_wide_association_studies/Meta-analysis and fine-mapping/Introduction","docId":"genome_wide_association_studies/Meta-analysis and fine-mapping/Introduction"},{"type":"link","label":"Meta-analysis","href":"/whg-training-resources/genome_wide_association_studies/Meta-analysis and fine-mapping/Meta-analysis","docId":"genome_wide_association_studies/Meta-analysis and fine-mapping/Meta-analysis"}],"href":"/whg-training-resources/genome_wide_association_studies/Meta-analysis and fine-mapping/"},{"type":"category","label":"Principal components analysis tutorial","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"PCA practical","href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/Introduction","docId":"genome_wide_association_studies/Principal components analysis/Introduction"},{"type":"link","label":"computing_PCs","href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/computing_PCs","docId":"genome_wide_association_studies/Principal components analysis/computing_PCs"},{"type":"link","label":"getting_setup","href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/getting_setup","docId":"genome_wide_association_studies/Principal components analysis/getting_setup"},{"type":"link","label":"global_analysis","href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/global_analysis","docId":"genome_wide_association_studies/Principal components analysis/global_analysis"},{"type":"link","label":"ld_pruning","href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/ld_pruning","docId":"genome_wide_association_studies/Principal components analysis/ld_pruning"},{"type":"link","label":"overview","href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/overview","docId":"genome_wide_association_studies/Principal components analysis/overview"},{"type":"link","label":"relatedness_pruning","href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/relatedness_pruning","docId":"genome_wide_association_studies/Principal components analysis/relatedness_pruning"},{"type":"link","label":"the_maths","href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/the_maths","docId":"genome_wide_association_studies/Principal components analysis/the_maths"}],"href":"/whg-training-resources/genome_wide_association_studies/Principal components analysis/"}],"href":"/whg-training-resources/genome_wide_association_studies/"},{"type":"category","label":"Next-generation sequencing","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Introduction to next-generation sequencing","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Getting the prerequisites","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites"},{"type":"link","label":"Practical outline","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Pipeline_outline","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Pipeline_outline"},{"type":"link","label":"What\'s in the fastq files?","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Inspecting_the_fastqs","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Inspecting_the_fastqs"},{"type":"link","label":"Inspecting sequence data quality","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Quality_control","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Quality_control"},{"type":"link","label":"Paired-end sequencing in theory and practice","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Short_read_theory","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Short_read_theory"},{"type":"link","label":"Aligning reads","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Aligning_reads","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Aligning_reads"},{"type":"link","label":"Viewing alignments","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments"},{"type":"link","label":"Challenge questions","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Challenge_questions","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Challenge_questions"},{"type":"link","label":"Appendix - how accurate is Illumina sequencing?","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/De_novo_error_rate_estimation","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/De_novo_error_rate_estimation"},{"type":"link","label":"Appendix - understanding sequence duplication levels","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/duplicates","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/duplicates"},{"type":"link","label":"Appendix - read trimming","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Read_trimming","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Read_trimming"},{"type":"link","label":"Useful links","href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Useful_links","docId":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Useful_links"}],"href":"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/"},{"type":"category","label":"Building an NGS data processing pipeline","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Introduction","href":"/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Introduction","docId":"next_generation_sequencing/Building an NGS processing pipeline/Introduction"},{"type":"link","label":"Required software","href":"/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Required software","docId":"next_generation_sequencing/Building an NGS processing pipeline/Required software"},{"type":"link","label":"Required_software","href":"/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Required_software","docId":"next_generation_sequencing/Building an NGS processing pipeline/Required_software"},{"type":"link","label":"pipeline","href":"/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/pipeline","docId":"next_generation_sequencing/Building an NGS processing pipeline/pipeline"}],"href":"/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/"},{"type":"category","label":"Variant calling, phasing and imputation","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Getting the data","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Prerequisites","docId":"next_generation_sequencing/variant_calling_and_imputation/Prerequisites"},{"type":"link","label":"Variant calling with bcftools","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Variant_calling","docId":"next_generation_sequencing/variant_calling_and_imputation/Variant_calling"},{"type":"link","label":"Variant quality control","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Variant_quality_control","docId":"next_generation_sequencing/variant_calling_and_imputation/Variant_quality_control"},{"type":"link","label":"Phasing the variant calls","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Phasing","docId":"next_generation_sequencing/variant_calling_and_imputation/Phasing"},{"type":"link","label":"Imputing a set of microarray data","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Imputation","docId":"next_generation_sequencing/variant_calling_and_imputation/Imputation"},{"type":"link","label":"Challenge questions on imputation","href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/Challenge_questions","docId":"next_generation_sequencing/variant_calling_and_imputation/Challenge_questions"}],"href":"/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/"},{"type":"category","label":"Visualising alignments","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Exploring Reads in IGV","href":"/whg-training-resources/next_generation_sequencing/IGV/Exploring_reads_in_IGV","docId":"next_generation_sequencing/IGV/Exploring_reads_in_IGV"}],"href":"/whg-training-resources/next_generation_sequencing/IGV/"}],"href":"/whg-training-resources/next_generation_sequencing/"},{"type":"link","label":"LICENSE","href":"/whg-training-resources/LICENSE","docId":"LICENSE"},{"type":"link","label":"contributors","href":"/whg-training-resources/contributors","docId":"contributors"},{"type":"category","label":"programming","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Introduction to R","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"R_basic_features_practical","href":"/whg-training-resources/programming/introduction_to_R/R_basic_features_practical","docId":"programming/introduction_to_R/R_basic_features_practical"},{"type":"link","label":"R: Introduction to Basic Features","href":"/whg-training-resources/programming/introduction_to_R/R_introduction_to_basic_features","docId":"programming/introduction_to_R/R_introduction_to_basic_features"},{"type":"link","label":"Introduction to R","href":"/whg-training-resources/programming/introduction_to_R/R_introduction_to_basic_features_slides","docId":"programming/introduction_to_R/R_introduction_to_basic_features_slides"}],"href":"/whg-training-resources/programming/introduction_to_R/"},{"type":"category","label":"Introduction to the Command Line","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"What is the command-line interface?","href":"/whg-training-resources/programming/introduction_to_command_line/command_line_intro","docId":"programming/introduction_to_command_line/command_line_intro"}],"href":"/whg-training-resources/programming/introduction_to_command_line/"},{"type":"category","label":"Programming with gene annotations","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Introduction","href":"/whg-training-resources/programming/programming_with_gene_annotations/Introduction","docId":"programming/programming_with_gene_annotations/Introduction"},{"type":"link","label":"What gene annotation data looks like","href":"/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like","docId":"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like"},{"type":"link","label":"Processing GFF files","href":"/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code","docId":"programming/programming_with_gene_annotations/Getting_started_writing_some_code"},{"type":"link","label":"Refactoring_makes_code_better","href":"/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better","docId":"programming/programming_with_gene_annotations/Refactoring_makes_code_better"},{"type":"link","label":"Converting_gff_to_sqlite","href":"/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite","docId":"programming/programming_with_gene_annotations/Converting_gff_to_sqlite"},{"type":"link","label":"Gathering some first statistics","href":"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1","docId":"programming/programming_with_gene_annotations/Counting_genes_1"},{"type":"link","label":"Memory_issues_and_how_to_solve_them","href":"/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them","docId":"programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them"},{"type":"link","label":"Counting_genes_2","href":"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2","docId":"programming/programming_with_gene_annotations/Counting_genes_2"},{"type":"link","label":"Counting_genes_3","href":"/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3","docId":"programming/programming_with_gene_annotations/Counting_genes_3"},{"type":"link","label":"Getting_sequence_lengths","href":"/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths","docId":"programming/programming_with_gene_annotations/Getting_sequence_lengths"},{"type":"link","label":"How_much_of_the_genome_is_in_genes","href":"/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes","docId":"programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes"},{"type":"link","label":"Scaling_up","href":"/whg-training-resources/programming/programming_with_gene_annotations/Scaling_up","docId":"programming/programming_with_gene_annotations/Scaling_up"},{"type":"link","label":"Visualisation","href":"/whg-training-resources/programming/programming_with_gene_annotations/Visualisation","docId":"programming/programming_with_gene_annotations/Visualisation"},{"type":"link","label":"Closing_thoughts","href":"/whg-training-resources/programming/programming_with_gene_annotations/Closing_thoughts","docId":"programming/programming_with_gene_annotations/Closing_thoughts"},{"type":"link","label":"Getting_organised_and_scaling_up","href":"/whg-training-resources/programming/programming_with_gene_annotations/Getting_organised_and_scaling_up","docId":"programming/programming_with_gene_annotations/Getting_organised_and_scaling_up"},{"type":"link","label":"List_of_ensembl_files","href":"/whg-training-resources/programming/programming_with_gene_annotations/List_of_ensembl_files","docId":"programming/programming_with_gene_annotations/List_of_ensembl_files"},{"type":"link","label":"Solutions","href":"/whg-training-resources/programming/programming_with_gene_annotations/Solutions","docId":"programming/programming_with_gene_annotations/Solutions"},{"type":"link","label":"Where_next","href":"/whg-training-resources/programming/programming_with_gene_annotations/Where_next","docId":"programming/programming_with_gene_annotations/Where_next"},{"type":"link","label":"parsing_gff3_the_low_level_way","href":"/whg-training-resources/programming/programming_with_gene_annotations/parsing_gff3_the_low_level_way","docId":"programming/programming_with_gene_annotations/parsing_gff3_the_low_level_way"}],"href":"/whg-training-resources/programming/programming_with_gene_annotations/"}]},{"type":"category","label":"Statistical modelling","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"README","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"practicals","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"Glycophorin CNV calling tutorial","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Hidden Markov Model CNV calling example - continued","href":"/whg-training-resources/statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/glycophorin_cnv_hmm","docId":"statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/glycophorin_cnv_hmm"},{"type":"link","label":"Hidden Markov Model example warmup","href":"/whg-training-resources/statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/glycophorin_cnv_warmup","docId":"statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/glycophorin_cnv_warmup"}],"href":"/whg-training-resources/statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/"}]}],"href":"/whg-training-resources/statistical_modelling/Hidden markov models/"},{"type":"category","label":"GMS Statistical Modelling I - introduction to statistical modelling","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"notes","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Probability solutions","href":"/whg-training-resources/statistical_modelling/Introduction/notes/Probability solutions","docId":"statistical_modelling/Introduction/notes/Probability solutions"},{"type":"link","label":"On methods of computing P-values","href":"/whg-training-resources/statistical_modelling/Introduction/notes/computing_pvalues","docId":"statistical_modelling/Introduction/notes/computing_pvalues"},{"type":"link","label":"probability questions","href":"/whg-training-resources/statistical_modelling/Introduction/notes/probability questions","docId":"statistical_modelling/Introduction/notes/probability questions"}]},{"type":"category","label":"practicals","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"a_tale_of_two_2x2_tables","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Model checking the malaria exposure finding","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison","docId":"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison"},{"type":"link","label":"O_bld_group_genome_wide_comparison","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison","docId":"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison"},{"type":"link","label":"When asymptotics go wrong","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/When asymptotics go wrong","docId":"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/When asymptotics go wrong"},{"type":"link","label":"A tale of two 2x2 tables","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction","docId":"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction"}]},{"type":"category","label":"bayesian_meta_analysis","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Bayesian meta-analysis practical","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical","docId":"statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical"}]},{"type":"category","label":"linear_regression","collapsible":true,"collapsed":true,"items":[{"type":"category","label":"atp2b4","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"atp2b4_practical","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical","docId":"statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical"},{"type":"link","label":"simulated_linear_model_practical","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/simulated_linear_model_practical","docId":"statistical_modelling/Introduction/practicals/linear_regression/atp2b4/simulated_linear_model_practical"}]}]},{"type":"category","label":"malaria_exposure","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Model checking the malaria exposure finding","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison","docId":"statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison"},{"type":"link","label":"When asymptotics go wrong","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/When asymptotics go wrong","docId":"statistical_modelling/Introduction/practicals/malaria_exposure/When asymptotics go wrong"}]},{"type":"category","label":"o_bld_grp","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"O_bld_group_genome_wide_comparison","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison","docId":"statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison"}]},{"type":"category","label":"warmup","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"distributions warmup","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/distributions warmup","docId":"statistical_modelling/Introduction/practicals/warmup/distributions warmup"},{"type":"link","label":"estimating an allele frequency","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency","docId":"statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency"},{"type":"link","label":"probability warmup","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/probability warmup","docId":"statistical_modelling/Introduction/practicals/warmup/probability warmup"},{"type":"category","label":"solutions","collapsible":true,"collapsed":true,"items":[{"type":"link","label":"Probability warm-up questions","href":"/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/solutions/probability warmup","docId":"statistical_modelling/Introduction/practicals/warmup/solutions/probability warmup"}]}]}]}],"href":"/whg-training-resources/statistical_modelling/Introduction/"}],"href":"/whg-training-resources/statistical_modelling/"}]},"docs":{"Clinical variant interpretation/README":{"id":"Clinical variant interpretation/README","title":"Clinical variant interpretation","description":"Contributors: Nicky Whiffin"},"contributors":{"id":"contributors","title":"contributors","description":"* Azim Ansari","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Genome-wide association analysis/GWAS_analysis_practical (plink version)":{"id":"genome_wide_association_studies/Genome-wide association analysis/GWAS_analysis_practical (plink version)","title":"Conducting GWAS analysis using plink","description":"Authors: Luke Jostins, Gavin Band","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Genome-wide association analysis/GWAS_analysis_practical (qctool+snptest version)":{"id":"genome_wide_association_studies/Genome-wide association analysis/GWAS_analysis_practical (qctool+snptest version)","title":"Conducting GWAS analysis using QCTOOL and SNPTEST","description":"Authors: Luke Jostins, Gavin Band","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Genome-wide association analysis/README":{"id":"genome_wide_association_studies/Genome-wide association analysis/README","title":"Genome-wide association study practical","description":"Welcome!  When you\'re ready, go here.md).","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Meta-analysis and fine-mapping/Fine-mapping":{"id":"genome_wide_association_studies/Meta-analysis and fine-mapping/Fine-mapping","title":"Fine-mapping","description":"Up to the table of contents - Back to the meta-analysis section","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Meta-analysis and fine-mapping/Introduction":{"id":"genome_wide_association_studies/Meta-analysis and fine-mapping/Introduction","title":"Introduction","description":"Up to the table of contents - Forward to the meta-analysis section","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Meta-analysis and fine-mapping/Meta-analysis":{"id":"genome_wide_association_studies/Meta-analysis and fine-mapping/Meta-analysis","title":"Meta-analysis","description":"Up to the table of contents - Back to the Introduction - Forward to the fine-mapping section","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Meta-analysis and fine-mapping/README":{"id":"genome_wide_association_studies/Meta-analysis and fine-mapping/README","title":"Meta-analysis and fine-mapping tutorial","description":"Contributors: Luke Jostins, Gavin Band","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/computing_PCs":{"id":"genome_wide_association_studies/Principal components analysis/computing_PCs","title":"computing_PCs","description":"Up to the table of contents - Back to the relatedness pruning step - Forward to the global analysis step.","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/getting_setup":{"id":"genome_wide_association_studies/Principal components analysis/getting_setup","title":"getting_setup","description":"Back to the table of contents - Forward to the overview","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/global_analysis":{"id":"genome_wide_association_studies/Principal components analysis/global_analysis","title":"global_analysis","description":"Up to the table of contents - Back to the PC computation step","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/Introduction":{"id":"genome_wide_association_studies/Principal components analysis/Introduction","title":"PCA practical","description":"Contributors: Gavin Band, Luke Jostins","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/ld_pruning":{"id":"genome_wide_association_studies/Principal components analysis/ld_pruning","title":"ld_pruning","description":"Up to the table of contents - Back to the previous page - plink --vcf chr19-clean.vcf.gz --maf 0.01 --indep-pairwise 50 5 0.2 --out chr19-clean Forward to the next page","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/overview":{"id":"genome_wide_association_studies/Principal components analysis/overview","title":"overview","description":"Up to the table of contents - Back to the setup page - Forward to the page on LD pruning","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/README":{"id":"genome_wide_association_studies/Principal components analysis/README","title":"Principal components analysis tutorial","description":"In this practical you will compute and debug a set of principal components for use in the GWAS analysis practical.","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/relatedness_pruning":{"id":"genome_wide_association_studies/Principal components analysis/relatedness_pruning","title":"relatedness_pruning","description":"Up to the table of contents - Back to the page on LD pruning - Forward to compute the PCs","sidebar":"tutorialSidebar"},"genome_wide_association_studies/Principal components analysis/the_maths":{"id":"genome_wide_association_studies/Principal components analysis/the_maths","title":"the_maths","description":"Up to the table of contents - Back to the page on computing PCs","sidebar":"tutorialSidebar"},"genome_wide_association_studies/README":{"id":"genome_wide_association_studies/README","title":"Genome-wide association analysis","description":"In this folder are:","sidebar":"tutorialSidebar"},"genome_wide_association_studies/slides/README":{"id":"genome_wide_association_studies/slides/README","title":"README","description":"Slides from these sessions can be found here:"},"LICENSE":{"id":"LICENSE","title":"LICENSE","description":"MIT License","sidebar":"tutorialSidebar"},"Machine learning and applications/README":{"id":"Machine learning and applications/README","title":"GMS Machine Learning Applications","description":"Suggested Software"},"next_generation_sequencing/Building an NGS processing pipeline/Introduction":{"id":"next_generation_sequencing/Building an NGS processing pipeline/Introduction","title":"Introduction","description":"Next generation sequence data processing tutorial","sidebar":"tutorialSidebar"},"next_generation_sequencing/Building an NGS processing pipeline/pipeline":{"id":"next_generation_sequencing/Building an NGS processing pipeline/pipeline","title":"pipeline","description":"Implementing a NGS data processing pipeline.","sidebar":"tutorialSidebar"},"next_generation_sequencing/Building an NGS processing pipeline/README":{"id":"next_generation_sequencing/Building an NGS processing pipeline/README","title":"Building an NGS data processing pipeline","description":"Welcome!  Go to the introduction to get started.","sidebar":"tutorialSidebar"},"next_generation_sequencing/Building an NGS processing pipeline/Required software":{"id":"next_generation_sequencing/Building an NGS processing pipeline/Required software","title":"Required software","description":"The required software page has moved here.","sidebar":"tutorialSidebar"},"next_generation_sequencing/Building an NGS processing pipeline/Required_software":{"id":"next_generation_sequencing/Building an NGS processing pipeline/Required_software","title":"Required_software","description":"Required software","sidebar":"tutorialSidebar"},"next_generation_sequencing/Getting started with snakemake/README":{"id":"next_generation_sequencing/Getting started with snakemake/README","title":"Getting started with snakemake","description":"This is a quick snakemake example that uses bwa mem to align a few reads."},"next_generation_sequencing/IGV/Exploring_reads_in_IGV":{"id":"next_generation_sequencing/IGV/Exploring_reads_in_IGV","title":"Exploring Reads in IGV","description":"Contributors: Annie Forster, Gavin Band","sidebar":"tutorialSidebar"},"next_generation_sequencing/IGV/README":{"id":"next_generation_sequencing/IGV/README","title":"Visualising alignments","description":"Welcome! Go the the IGV practical to get started.","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Aligning_reads":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Aligning_reads","title":"Aligning reads","description":"Most analyses of short-read sequence data are based on the basic paradigm made up of two steps:","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Challenge_questions":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Challenge_questions","title":"Challenge questions","description":"We have three challenges.:","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/De_novo_error_rate_estimation":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/De_novo_error_rate_estimation","title":"Appendix - how accurate is Illumina sequencing?","description":"Note. Only do this section if you have plenty of time left, and/or have already read the","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/duplicates":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/duplicates","title":"Appendix - understanding sequence duplication levels","description":"Duplicates arise naturally in sequencing from random fragments that just happen to have","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Inspecting_the_fastqs":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Inspecting_the_fastqs","title":"What\'s in the fastq files?","description":"Using UNIX to inspect the fastqs","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Pipeline_outline":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Pipeline_outline","title":"Practical outline","description":"Overview","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites","title":"Getting the prerequisites","description":"To run this tutorial you will need two things: some software and some data.  This page will get you set up.","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Quality_control":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Quality_control","title":"Inspecting sequence data quality","description":"There are a set of analyses that can be carried out directly on a FASTQ file to assess the quality","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Read_trimming":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Read_trimming","title":"Appendix - read trimming","description":"Read trimming refers to the process of removing trailing parts of the reads before analysis for various reasons including","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/README":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/README","title":"Introduction to next-generation sequencing","description":"In this tutorial we will demonstrate a basic pipeline for analysing paired-end short-read genomic","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Short_read_theory":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Short_read_theory","title":"Paired-end sequencing in theory and practice","description":"What paired-end sequence data should lool like","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Useful_links":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Useful_links","title":"Useful links","description":"* Illumina\'s page on base quality scores.","sidebar":"tutorialSidebar"},"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments":{"id":"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Viewing_alignments","title":"Viewing alignments","description":"If you got this far, you should have a file called QG0033-C.bam containing aligned,","sidebar":"tutorialSidebar"},"next_generation_sequencing/README":{"id":"next_generation_sequencing/README","title":"Next-generation sequencing","description":"* Go to the introduction to next-generation sequence (NGS) data analysis tutorial.","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Challenge_questions":{"id":"next_generation_sequencing/variant_calling_and_imputation/Challenge_questions","title":"Challenge questions on imputation","description":"Imputation $r^2$","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Imputation":{"id":"next_generation_sequencing/variant_calling_and_imputation/Imputation","title":"Imputing a set of microarray data","description":"Ok, so we\'ve called genetic variants, we\'ve [QCd","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Phasing":{"id":"next_generation_sequencing/variant_calling_and_imputation/Phasing","title":"Phasing the variant calls","description":"If you got this far you should have a robust, well-QCd set of variants. But we\'re not done! Humans","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Prerequisites":{"id":"next_generation_sequencing/variant_calling_and_imputation/Prerequisites","title":"Getting the data","description":"To get the data download and unzip this tarball:","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/README":{"id":"next_generation_sequencing/variant_calling_and_imputation/README","title":"Variant calling, phasing and imputation","description":"This morning we focussed on quality control, aligning, and inspecting some sequence data for a single sample.","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Variant_calling":{"id":"next_generation_sequencing/variant_calling_and_imputation/Variant_calling","title":"Variant calling with bcftools","description":"First let\'s see how to use a simple pipeline to identify genetic variants","sidebar":"tutorialSidebar"},"next_generation_sequencing/variant_calling_and_imputation/Variant_quality_control":{"id":"next_generation_sequencing/variant_calling_and_imputation/Variant_quality_control","title":"Variant quality control","description":"For variant quality control we\'ll start with the output of bcftools call and use the various INFO","sidebar":"tutorialSidebar"},"overview":{"id":"overview","title":"Overview of the tutorials","description":"This folder contains course materials (practicals, notes and datasets) that were developed for the taught statistics component of the","sidebar":"tutorialSidebar"},"population_genetics/implementing_a_Wright-Fisher_model":{"id":"population_genetics/implementing_a_Wright-Fisher_model","title":"Implementing a Wright-Fisher model with a finite number of SNPs","description":"The Wright-Fisher model is about the simplest population-genetic model there is.  It consists of:","sidebar":"tutorialSidebar"},"population_genetics/README":{"id":"population_genetics/README","title":"Population genetics","description":"In this practical we will implement a simple Wright-Fisher population genetics model, and couple","sidebar":"tutorialSidebar"},"prerequisites/CONDA":{"id":"prerequisites/CONDA","title":"Setting up Conda","description":"One of the easiest ways to set up your environment that works across platforms is to use conda.","sidebar":"tutorialSidebar"},"prerequisites/julia":{"id":"prerequisites/julia","title":"Julia","description":"Julia is an interesting language that is pretty good for many data analysis and computational","sidebar":"tutorialSidebar"},"prerequisites/Jupyterlab":{"id":"prerequisites/Jupyterlab","title":"Using R, python and julia with JupyterLab","description":"The plain command-line interpreters for R, python and julia are useful, but for data analysis work you may want","sidebar":"tutorialSidebar"},"prerequisites/Other_languages/C++":{"id":"prerequisites/Other_languages/C++","title":"C++","description":"This page is to be written.","sidebar":"tutorialSidebar"},"prerequisites/Other_languages/javascript":{"id":"prerequisites/Other_languages/javascript","title":"Javascript","description":"You don\'t need to install javascript - you have it already in your browser. To have a go at using","sidebar":"tutorialSidebar"},"prerequisites/Other_languages/README":{"id":"prerequisites/Other_languages/README","title":"Other languages","description":"There are lots of other programming languages out there - useful for different tasks. Here are some:","sidebar":"tutorialSidebar"},"prerequisites/python":{"id":"prerequisites/python","title":"Python","description":"You shouldn\'t need to install python - both your base system, and your conda environment should have a version of python installed already.  Check by typing","sidebar":"tutorialSidebar"},"prerequisites/R":{"id":"prerequisites/R","title":"R and RStudio","description":"A great deal of data analysis work is done in R - you\'ll need this","sidebar":"tutorialSidebar"},"prerequisites/README":{"id":"prerequisites/README","title":"Getting your environment setup","description":"The tutorials found here involve running bioinformatics tools and writing computer code. You will","sidebar":"tutorialSidebar"},"prerequisites/tidyverse":{"id":"prerequisites/tidyverse","title":"The R `tidyverse`","description":"Base R is ok but it can be a bit... clunky. For many of the tutorials we wil use a set of","sidebar":"tutorialSidebar"},"prerequisites/UNIX":{"id":"prerequisites/UNIX","title":"Setting up a UNIX terminal","description":"You will be doing a lot of work in the UNIX terminal - this is a command-line interface.","sidebar":"tutorialSidebar"},"programming/introduction_to_command_line/command_line_intro":{"id":"programming/introduction_to_command_line/command_line_intro","title":"What is the command-line interface?","description":"The command-line interface (sometimes also called \u201ccommand prompt\u201d or \u201cterminal\u201d) is a way of","sidebar":"tutorialSidebar"},"programming/introduction_to_command_line/README":{"id":"programming/introduction_to_command_line/README","title":"Introduction to the Command Line","description":"An introduction to using the BASH command line.","sidebar":"tutorialSidebar"},"programming/introduction_to_R/R_basic_features_practical":{"id":"programming/introduction_to_R/R_basic_features_practical","title":"R_basic_features_practical","description":"Tutors: Helen Lockstone and Ben Wright Bioinformatics Core Wellcome","sidebar":"tutorialSidebar"},"programming/introduction_to_R/R_introduction_to_basic_features":{"id":"programming/introduction_to_R/R_introduction_to_basic_features","title":"R: Introduction to Basic Features","description":"Author: Helen Lockstone","sidebar":"tutorialSidebar"},"programming/introduction_to_R/R_introduction_to_basic_features_slides":{"id":"programming/introduction_to_R/R_introduction_to_basic_features_slides","title":"Introduction to R","description":"Authors: Helen Lockstone and Matthieu Miossec Bioinformatics Core","sidebar":"tutorialSidebar"},"programming/introduction_to_R/README":{"id":"programming/introduction_to_R/README","title":"Introduction to R","description":"Welcome! To start the tutorial, go here.","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Closing_thoughts":{"id":"programming/programming_with_gene_annotations/Closing_thoughts","title":"Closing_thoughts","description":"Go up to the table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Converting_gff_to_sqlite":{"id":"programming/programming_with_gene_annotations/Converting_gff_to_sqlite","title":"Converting_gff_to_sqlite","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Counting_genes_1":{"id":"programming/programming_with_gene_annotations/Counting_genes_1","title":"Gathering some first statistics","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Counting_genes_2":{"id":"programming/programming_with_gene_annotations/Counting_genes_2","title":"Counting_genes_2","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Counting_genes_3":{"id":"programming/programming_with_gene_annotations/Counting_genes_3","title":"Counting_genes_3","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Getting_organised_and_scaling_up":{"id":"programming/programming_with_gene_annotations/Getting_organised_and_scaling_up","title":"Getting_organised_and_scaling_up","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Getting_sequence_lengths":{"id":"programming/programming_with_gene_annotations/Getting_sequence_lengths","title":"Getting_sequence_lengths","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Getting_started_writing_some_code":{"id":"programming/programming_with_gene_annotations/Getting_started_writing_some_code","title":"Processing GFF files","description":"Up to table of contents / Back to the previous page / Go to the next page","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes":{"id":"programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes","title":"How_much_of_the_genome_is_in_genes","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Introduction":{"id":"programming/programming_with_gene_annotations/Introduction","title":"Introduction","description":"Up to table of contents / Go to next page.","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/List_of_ensembl_files":{"id":"programming/programming_with_gene_annotations/List_of_ensembl_files","title":"List_of_ensembl_files","description":"A list of relevant GFF files.","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them":{"id":"programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them","title":"Memory_issues_and_how_to_solve_them","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/parsing_gff3_the_low_level_way":{"id":"programming/programming_with_gene_annotations/parsing_gff3_the_low_level_way","title":"parsing_gff3_the_low_level_way","description":"A basic python approach","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/README":{"id":"programming/programming_with_gene_annotations/README","title":"Programming with gene annotations","description":"Welcome! This tutorial has two aims. One one hand it is about programming, specifically the type of programming","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Refactoring_makes_code_better":{"id":"programming/programming_with_gene_annotations/Refactoring_makes_code_better","title":"Refactoring_makes_code_better","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Scaling_up":{"id":"programming/programming_with_gene_annotations/Scaling_up","title":"Scaling_up","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Solutions":{"id":"programming/programming_with_gene_annotations/Solutions","title":"Solutions","description":"","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Visualisation":{"id":"programming/programming_with_gene_annotations/Visualisation","title":"Visualisation","description":"Up to table of contents","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like":{"id":"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like","title":"What gene annotation data looks like","description":"Up to table of contents / Back to the introduction / Go the next page","sidebar":"tutorialSidebar"},"programming/programming_with_gene_annotations/Where_next":{"id":"programming/programming_with_gene_annotations/Where_next","title":"Where_next","description":"Up to table of contents","sidebar":"tutorialSidebar"},"statistical_modelling/Bayesian analysis/README":{"id":"statistical_modelling/Bayesian analysis/README","title":"Statistical modelling II: Bayesian analysis","description":"This session will introduce principles of Bayesian analysis and applications to genomic datasets."},"statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/glycophorin_cnv_hmm":{"id":"statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/glycophorin_cnv_hmm","title":"Hidden Markov Model CNV calling example - continued","description":"!! Note !!  To run this code, you first need to have computed a set of log-likelihoods for each sample at each site in the relevant dataset.  The code to do this was explained in the warmup tutorial, which you can also find in this directory.","sidebar":"tutorialSidebar"},"statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/glycophorin_cnv_warmup":{"id":"statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/glycophorin_cnv_warmup","title":"Hidden Markov Model example warmup","description":"Welcome to the warm-up for the Hidden Markov Models session this week.","sidebar":"tutorialSidebar"},"statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/README":{"id":"statistical_modelling/Hidden markov models/practicals/glycophorin_cnvs/README","title":"Glycophorin CNV calling tutorial","description":"Table of contents:","sidebar":"tutorialSidebar"},"statistical_modelling/Hidden markov models/README":{"id":"statistical_modelling/Hidden markov models/README","title":"README","description":"","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/notes/computing_pvalues":{"id":"statistical_modelling/Introduction/notes/computing_pvalues","title":"On methods of computing P-values","description":"In lectures we covered the \'asymptotic normality\' property of likelihoods, to wit:","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/notes/probability questions":{"id":"statistical_modelling/Introduction/notes/probability questions","title":"probability questions","description":"Probability warm-up questions","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/notes/Probability solutions":{"id":"statistical_modelling/Introduction/notes/Probability solutions","title":"Probability solutions","description":"Probability warm-up questions","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison":{"id":"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison","title":"Model checking the malaria exposure finding","description":"Consider the following table from :","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction":{"id":"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction","title":"A tale of two 2x2 tables","description":"Welcome to (possibly) the world\'s first choose your own statistical adventure!","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison":{"id":"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison","title":"O_bld_group_genome_wide_comparison","description":"Calibration / model checking for the O blood group finding","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/When asymptotics go wrong":{"id":"statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/When asymptotics go wrong","title":"When asymptotics go wrong","description":"If you\'ve been following the practical so far, you\'ll have a plot of the parameter estimates from a binomial likelihood for both rows of our table.","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical":{"id":"statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical","title":"Bayesian meta-analysis practical","description":"In this practical we will investigate a simple method to investigate sharing of genetic effects between traits.","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical":{"id":"statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical","title":"atp2b4_practical","description":"Linear regression practical for ATP2B4 expression data","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/linear_regression/atp2b4/simulated_linear_model_practical":{"id":"statistical_modelling/Introduction/practicals/linear_regression/atp2b4/simulated_linear_model_practical","title":"simulated_linear_model_practical","description":"Linear regression for simulated data","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison":{"id":"statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison","title":"Model checking the malaria exposure finding","description":"Consider the following table from :","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/malaria_exposure/When asymptotics go wrong":{"id":"statistical_modelling/Introduction/practicals/malaria_exposure/When asymptotics go wrong","title":"When asymptotics go wrong","description":"If you\'ve been following the practical so far, you\'ll have a plot of the parameter estimates from a binomial likelihood for both rows of our table.","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison":{"id":"statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison","title":"O_bld_group_genome_wide_comparison","description":"Calibration / model checking for the O blood group finding","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/warmup/distributions warmup":{"id":"statistical_modelling/Introduction/practicals/warmup/distributions warmup","title":"distributions warmup","description":"Probability distribution questions","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency":{"id":"statistical_modelling/Introduction/practicals/warmup/estimating an allele frequency","title":"estimating an allele frequency","description":"The frequency of O blood group","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/warmup/probability warmup":{"id":"statistical_modelling/Introduction/practicals/warmup/probability warmup","title":"probability warmup","description":"Probability warm-up questions","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/practicals/warmup/solutions/probability warmup":{"id":"statistical_modelling/Introduction/practicals/warmup/solutions/probability warmup","title":"Probability warm-up questions","description":"DNA inheritance example","sidebar":"tutorialSidebar"},"statistical_modelling/Introduction/README":{"id":"statistical_modelling/Introduction/README","title":"GMS Statistical Modelling I - introduction to statistical modelling","description":"Overview","sidebar":"tutorialSidebar"},"statistical_modelling/MCMC and sampling/README":{"id":"statistical_modelling/MCMC and sampling/README","title":"Statistical modelling: MCMC and sampling","description":"Overview"},"statistical_modelling/README":{"id":"statistical_modelling/README","title":"Statistical modelling","description":"This folder contains a range of practicals on statistical modelling.","sidebar":"tutorialSidebar"},"statistical_modelling/Statistical models for polygenic traits/README":{"id":"statistical_modelling/Statistical models for polygenic traits/README","title":"GMS Statistical Analysis of Genome-wide data","description":"Required Software/Data"},"statistics_for_genomics/README":{"id":"statistics_for_genomics/README","title":"GMS Introduction to Statistics","description":"Prerequisites"}}}')}}]);