"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[540],{3905:function(e,t,n){n.d(t,{Zo:function(){return c},kt:function(){return m}});var r=n(7294);function a(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function o(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){a(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},i=Object.keys(e);for(r=0;r<i.length;r++)n=i[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(r=0;r<i.length;r++)n=i[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}var l=r.createContext({}),p=function(e){var t=r.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):o(o({},t),e)),n},c=function(e){var t=p(e.components);return r.createElement(l.Provider,{value:t},e.children)},u={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},d=r.forwardRef((function(e,t){var n=e.components,a=e.mdxType,i=e.originalType,l=e.parentName,c=s(e,["components","mdxType","originalType","parentName"]),d=p(n),m=a,g=d["".concat(l,".").concat(m)]||d[m]||u[m]||i;return n?r.createElement(g,o(o({ref:t},c),{},{components:n})):r.createElement(g,o({ref:t},c))}));function m(e,t){var n=arguments,a=t&&t.mdxType;if("string"==typeof e||a){var i=n.length,o=new Array(i);o[0]=d;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"==typeof e?e:a,o[1]=s;for(var p=2;p<i;p++)o[p]=n[p];return r.createElement.apply(null,o)}return r.createElement.apply(null,n)}d.displayName="MDXCreateElement"},1992:function(e,t,n){n.r(t),n.d(t,{assets:function(){return c},contentTitle:function(){return l},default:function(){return m},frontMatter:function(){return s},metadata:function(){return p},toc:function(){return u}});var r=n(7462),a=n(3366),i=(n(7294),n(3905)),o=["components"],s={},l=void 0,p={unversionedId:"genome_wide_association_studies/Principal components analysis/overview",id:"genome_wide_association_studies/Principal components analysis/overview",title:"overview",description:"Up to the table of contents - Back to the setup page - Forward to the page on LD pruning",source:"@site/docs/genome_wide_association_studies/Principal components analysis/overview.md",sourceDirName:"genome_wide_association_studies/Principal components analysis",slug:"/genome_wide_association_studies/Principal components analysis/overview",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/overview",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/genome_wide_association_studies/Principal components analysis/overview.md",tags:[],version:"current",frontMatter:{},sidebar:"tutorialSidebar",previous:{title:"ld_pruning",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/ld_pruning"},next:{title:"relatedness_pruning",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/relatedness_pruning"}},c={},u=[{value:"Overview of the practical",id:"overview-of-the-practical",level:3},{value:"A note on quality control",id:"a-note-on-quality-control",level:3},{value:"Preparing data for PCA",id:"preparing-data-for-pca",level:3}],d={toc:u};function m(e){var t=e.components,n=(0,a.Z)(e,o);return(0,i.kt)("wrapper",(0,r.Z)({},d,n,{components:t,mdxType:"MDXLayout"}),(0,i.kt)("p",null,(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/Introduction"},"Up to the table of contents")," - ",(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/getting_setup"},"Back to the setup page")," - ",(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/ld_pruning"},"Forward to the page on LD pruning")),(0,i.kt)("h3",{id:"overview-of-the-practical"},"Overview of the practical"),(0,i.kt)("p",null,"In this practical we will use ",(0,i.kt)("inlineCode",{parentName:"p"},"plink")," to do several things to the data:"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},"to remove closely-related samples"),(0,i.kt)("li",{parentName:"ul"},"to compute principal components"),(0,i.kt)("li",{parentName:"ul"},"and to compute the SNP weights or loadings that tell us how principal components are weighted across the genome.")),(0,i.kt)("p",null,"We'll also use ",(0,i.kt)("inlineCode",{parentName:"p"},"R")," (we recommend ",(0,i.kt)("a",{parentName:"p",href:"https://www.rstudio.com"},(0,i.kt)("inlineCode",{parentName:"a"},"RStudio")),") to inspect and plot results."),(0,i.kt)("p",null,"Open a terminal window and first make sure you are in the right directory:"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"cd /path/to/PCA_practical\n")),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Note.")," This should be the folder where ",(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/getting_setup"},"you downloaded the data"),"."),(0,i.kt)("p",null,"Also, in R / RStudio please set this directory as your current directory - either using the ",(0,i.kt)("inlineCode",{parentName:"p"},"setwd()")," command like this:"),(0,i.kt)("p",null,(0,i.kt)("em",{parentName:"p"},"In R/RStudio:")),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"> setwd( '/path/to/PCA_practical' )\n")),(0,i.kt)("p",null,"or by using the menu option ",(0,i.kt)("inlineCode",{parentName:"p"},"Session"),"->",(0,i.kt)("inlineCode",{parentName:"p"},"Set working directory"),"->",(0,i.kt)("inlineCode",{parentName:"p"},"Choose Directory"),") in RStudio."),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Note.")," The ",(0,i.kt)("inlineCode",{parentName:"p"},">")," indicates the ",(0,i.kt)("inlineCode",{parentName:"p"},"R")," command prompt above - don't type that bit in!"),(0,i.kt)("h3",{id:"a-note-on-quality-control"},"A note on quality control"),(0,i.kt)("p",null,"Before carrying out a genetic analysis like PCA, it's important to have a good-quality dataset, and\nthis typically means carrying out careful quality control (QC) first. On this course we'll cover QC\nin later lectures and practicals. For this practical we'll use an already-cleaned dataset contained\nin the file ",(0,i.kt)("inlineCode",{parentName:"p"},"chr19-clean.vcf.gz"),". You can look at the data in this file by typing"),(0,i.kt)("pre",null,(0,i.kt)("code",{parentName:"pre"},"less -S chr19-clean.vcf.gz\n")),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Note.")," If you are using Mac OS X, you will need to use ",(0,i.kt)("inlineCode",{parentName:"p"},"zless")," instead of ",(0,i.kt)("inlineCode",{parentName:"p"},"less")," because the file is gzipped."),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Note.")," Press ",(0,i.kt)("inlineCode",{parentName:"p"},"q")," when you want to quit ",(0,i.kt)("inlineCode",{parentName:"p"},"less"),"."),(0,i.kt)("p",null,"The data consists of genotype calls at different sites (rows) for different samples (columns).  Feel free to look at the data by scrolling around using the arrow keys. When you've finished, press the 'q' key to quit back to the terminal prompt."),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Note.")," This is a ",(0,i.kt)("a",{parentName:"p",href:"https://samtools.github.io/hts-specs/VCFv4.2.pdf"},"Variant Call Format")," file.\nIf you followed our earlier ",(0,i.kt)("a",{parentName:"p",href:"../../../Next_Generation_Sequencing/practicals/ngs_processing_pipeline/"},"next-generation sequencing\npractical")," you will have\ncreated a VCF file of variant calls from some sequence reads.  The VCF file for this practical is much simpler, though, because we have only included genotype calls."),(0,i.kt)("h3",{id:"preparing-data-for-pca"},"Preparing data for PCA"),(0,i.kt)("p",null,"Before computing PCs we will need to do some pruning of the data.  We will:"),(0,i.kt)("ul",null,(0,i.kt)("li",{parentName:"ul"},"remove SNPs that are in high LD (to avoid confounding the analysis by local LD patterns.)"),(0,i.kt)("li",{parentName:"ul"},"remove samples that are too closely related (so that our PCs reflect the majority of our data.)")),(0,i.kt)("p",null,"When you're ready, ",(0,i.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/ld_pruning"},"go here to start pruning"),"."))}m.isMDXComponent=!0}}]);