"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[9382],{3905:function(e,t,n){n.d(t,{Zo:function(){return c},kt:function(){return d}});var r=n(7294);function a(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function o(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function i(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?o(Object(n),!0).forEach((function(t){a(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):o(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},o=Object.keys(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}var l=r.createContext({}),u=function(e){var t=r.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):i(i({},t),e)),n},c=function(e){var t=u(e.components);return r.createElement(l.Provider,{value:t},e.children)},g={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},p=r.forwardRef((function(e,t){var n=e.components,a=e.mdxType,o=e.originalType,l=e.parentName,c=s(e,["components","mdxType","originalType","parentName"]),p=u(n),d=a,h=p["".concat(l,".").concat(d)]||p[d]||g[d]||o;return n?r.createElement(h,i(i({ref:t},c),{},{components:n})):r.createElement(h,i({ref:t},c))}));function d(e,t){var n=arguments,a=t&&t.mdxType;if("string"==typeof e||a){var o=n.length,i=new Array(o);i[0]=p;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"==typeof e?e:a,i[1]=s;for(var u=2;u<o;u++)i[u]=n[u];return r.createElement.apply(null,i)}return r.createElement.apply(null,n)}p.displayName="MDXCreateElement"},2222:function(e,t,n){n.r(t),n.d(t,{assets:function(){return c},contentTitle:function(){return l},default:function(){return d},frontMatter:function(){return s},metadata:function(){return u},toc:function(){return g}});var r=n(7462),a=n(3366),o=(n(7294),n(3905)),i=["components"],s={sidebar_position:1},l="Getting the prerequisites",u={unversionedId:"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites",id:"next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites",title:"Getting the prerequisites",description:"To run this tutorial you will need two things: some software and some data.  This page will get you set up.",source:"@site/docs/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites.md",sourceDirName:"next_generation_sequencing/Introduction to next-generation sequencing data analysis",slug:"/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites",permalink:"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites.md",tags:[],version:"current",sidebarPosition:1,frontMatter:{sidebar_position:1},sidebar:"tutorialSidebar",previous:{title:"Introduction to next-generation sequencing",permalink:"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/"},next:{title:"Practical outline",permalink:"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Pipeline_outline"}},c={},g=[{value:"Getting the software",id:"getting-the-software",level:2},{value:"Getting the data",id:"getting-the-data",level:2}],p={toc:g};function d(e){var t=e.components,n=(0,a.Z)(e,i);return(0,o.kt)("wrapper",(0,r.Z)({},p,n,{components:t,mdxType:"MDXLayout"}),(0,o.kt)("h1",{id:"getting-the-prerequisites"},"Getting the prerequisites"),(0,o.kt)("p",null,"To run this tutorial you will need two things: some software and some data.  This page will get you set up."),(0,o.kt)("h2",{id:"getting-the-software"},"Getting the software"),(0,o.kt)("p",null,"You need a version of the following tools:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("a",{parentName:"li",href:"http://samtools.github.io"},"samtools"),", a program for manipulating next-generation sequencing data reads"),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("a",{parentName:"li",href:"https://www.bioinformatics.babraham.ac.uk/projects/fastqc/"},"fastqc"),", a program for performing quality control fo sequence data"),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("a",{parentName:"li",href:"https://github.com/lh3/bwa"},"bwa"),", a program for aligning reads to a reference sequence"),(0,o.kt)("li",{parentName:"ul"},(0,o.kt)("a",{parentName:"li",href:"https://github.com/zippav/Jellyfish-2"},"jellyfish 2"),", a program for counting k-mers (short sequences of fixed length k) in sequence data reads."),(0,o.kt)("li",{parentName:"ul"},"And it is also useful to install the ",(0,o.kt)("a",{parentName:"li",href:"https://igv.org"},"Integrated Genomics Viewer"),".")),(0,o.kt)("p",null,"You can get this software in a few different ways:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"If you are working on a JupyterHub instance we have setup, these tools should be installed already."),(0,o.kt)("li",{parentName:"ul"},"If you have set up a ",(0,o.kt)("a",{parentName:"li",href:"prerequiesites/README.md"},"conda installation as described here"),", the command")),(0,o.kt)("p",null,(0,o.kt)("inlineCode",{parentName:"p"},"mamba install -c bioconda -c conda-forge samtools fastqc bwa jellyfish")),(0,o.kt)("p",null,"run in a terminal should install these tools (except IGV, which should be downloaded from its website.)"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"It's also possible to install these programs from source code. Doing this is out of scope of this\ntutorial but if you know what you're doing, feel free to try this.")),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Testing it."),"  To check that you have the right software, open a bash terminal and type the program names followed\nby ","<","enter",">",". (Don't type the dollar signs, these are there to show you the bash prompt ). You\nshould see something like:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-bash"},"$ jellyfish --version\njellyfish 2.3.0\n\n$ fastqc --version\nFastQC v0.11.9\n\n$ samtools --version\nsamtools 1.8\nUsing htslib 1.8\nCopyright (C) 2018 Genome Research Ltd.\n\n$ bwa\n\nProgram: bwa (alignment via Burrows-Wheeler transformation)\nVersion: 0.7.17-r1188\nContact: Heng Li <lh3@sanger.ac.uk>\n\nUsage:   bwa <command> [options]\n\n[...]\n")),(0,o.kt)("h2",{id:"getting-the-data"},"Getting the data"),(0,o.kt)("p",null,"To get the data for this tutorial, open a terminal and run the following commands:"),(0,o.kt)("p",null,"Download the data: "),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-sh"},"wget https://www.well.ox.ac.uk/~gav/projects/oxford_statgen_summer_school/day_one_morning.tgz\n")),(0,o.kt)("p",null,"Extract it :"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"tar -xzf day_one_morning.tgz\n")),(0,o.kt)("p",null,"You should now have a folder called ",(0,o.kt)("inlineCode",{parentName:"p"},"sequence_data_analysis"),". For the tutorial, delete the tarball\nand then change into that directory:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"rm day_one_morning.tgz\ncd sequence_data_analysis\n")),(0,o.kt)("p",null,"Now you're ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/next_generation_sequencing/Introduction%20to%20next-generation%20sequencing%20data%20analysis/Pipeline_outline"},"ready to start"),"."))}d.isMDXComponent=!0}}]);