"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[1362],{3905:function(e,n,t){t.d(n,{Zo:function(){return l},kt:function(){return d}});var r=t(7294);function i(e,n,t){return n in e?Object.defineProperty(e,n,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[n]=t,e}function o(e,n){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);n&&(r=r.filter((function(n){return Object.getOwnPropertyDescriptor(e,n).enumerable}))),t.push.apply(t,r)}return t}function a(e){for(var n=1;n<arguments.length;n++){var t=null!=arguments[n]?arguments[n]:{};n%2?o(Object(t),!0).forEach((function(n){i(e,n,t[n])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):o(Object(t)).forEach((function(n){Object.defineProperty(e,n,Object.getOwnPropertyDescriptor(t,n))}))}return e}function s(e,n){if(null==e)return{};var t,r,i=function(e,n){if(null==e)return{};var t,r,i={},o=Object.keys(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||(i[t]=e[t]);return i}(e,n);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(i[t]=e[t])}return i}var u=r.createContext({}),c=function(e){var n=r.useContext(u),t=n;return e&&(t="function"==typeof e?e(n):a(a({},n),e)),t},l=function(e){var n=c(e.components);return r.createElement(u.Provider,{value:n},e.children)},g={inlineCode:"code",wrapper:function(e){var n=e.children;return r.createElement(r.Fragment,{},n)}},p=r.forwardRef((function(e,n){var t=e.components,i=e.mdxType,o=e.originalType,u=e.parentName,l=s(e,["components","mdxType","originalType","parentName"]),p=c(t),d=i,f=p["".concat(u,".").concat(d)]||p[d]||g[d]||o;return t?r.createElement(f,a(a({ref:n},l),{},{components:t})):r.createElement(f,a({ref:n},l))}));function d(e,n){var t=arguments,i=n&&n.mdxType;if("string"==typeof e||i){var o=t.length,a=new Array(o);a[0]=p;var s={};for(var u in n)hasOwnProperty.call(n,u)&&(s[u]=n[u]);s.originalType=e,s.mdxType="string"==typeof e?e:i,a[1]=s;for(var c=2;c<o;c++)a[c]=t[c];return r.createElement.apply(null,a)}return r.createElement.apply(null,t)}p.displayName="MDXCreateElement"},3431:function(e,n,t){t.r(n),t.d(n,{assets:function(){return l},contentTitle:function(){return u},default:function(){return d},frontMatter:function(){return s},metadata:function(){return c},toc:function(){return g}});var r=t(7462),i=t(3366),o=(t(7294),t(3905)),a=["components"],s={sidebar_position:1},u="Introduction to next-generation sequencing",c={unversionedId:"next_generation_sequencing/Introduction to next-generation sequencing data analysis/README",id:"next_generation_sequencing/Introduction to next-generation sequencing data analysis/README",title:"Introduction to next-generation sequencing",description:"In this tutorial we will demonstrate a basic pipeline for analysing paired-end short-read genomic",source:"@site/docs/next_generation_sequencing/Introduction to next-generation sequencing data analysis/README.md",sourceDirName:"next_generation_sequencing/Introduction to next-generation sequencing data analysis",slug:"/next_generation_sequencing/Introduction to next-generation sequencing data analysis/",permalink:"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/next_generation_sequencing/Introduction to next-generation sequencing data analysis/README.md",tags:[],version:"current",sidebarPosition:1,frontMatter:{sidebar_position:1},sidebar:"tutorialSidebar",previous:{title:"Next-generation sequencing",permalink:"/whg-training-resources/next_generation_sequencing/"},next:{title:"Getting the prerequisites",permalink:"/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/Prerequisites"}},l={},g=[{value:"Contributors",id:"contributors",level:2}],p={toc:g};function d(e){var n=e.components,t=(0,i.Z)(e,a);return(0,o.kt)("wrapper",(0,r.Z)({},p,t,{components:n,mdxType:"MDXLayout"}),(0,o.kt)("h1",{id:"introduction-to-next-generation-sequencing"},"Introduction to next-generation sequencing"),(0,o.kt)("p",null,"In this tutorial we will demonstrate a basic pipeline for analysing paired-end short-read genomic\nsequencing data. We will start with raw data in a FASTQ file, inspect quality control metrics,\nalign the data, and then use it to look for genetic variation."),(0,o.kt)("p",null,"To get started, go ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/next_generation_sequencing/Introduction%20to%20next-generation%20sequencing%20data%20analysis/Prerequisites"},"to and get the prerequisites"),"."),(0,o.kt)("h2",{id:"contributors"},"Contributors"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"Gavin Band"),(0,o.kt)("li",{parentName:"ul"},"Mattheiu Miossec"),(0,o.kt)("li",{parentName:"ul"},"Annie Forster")))}d.isMDXComponent=!0}}]);