"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[7948],{3905:function(e,n,t){t.d(n,{Zo:function(){return p},kt:function(){return m}});var r=t(7294);function i(e,n,t){return n in e?Object.defineProperty(e,n,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[n]=t,e}function o(e,n){var t=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);n&&(r=r.filter((function(n){return Object.getOwnPropertyDescriptor(e,n).enumerable}))),t.push.apply(t,r)}return t}function a(e){for(var n=1;n<arguments.length;n++){var t=null!=arguments[n]?arguments[n]:{};n%2?o(Object(t),!0).forEach((function(n){i(e,n,t[n])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(t)):o(Object(t)).forEach((function(n){Object.defineProperty(e,n,Object.getOwnPropertyDescriptor(t,n))}))}return e}function s(e,n){if(null==e)return{};var t,r,i=function(e,n){if(null==e)return{};var t,r,i={},o=Object.keys(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||(i[t]=e[t]);return i}(e,n);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)t=o[r],n.indexOf(t)>=0||Object.prototype.propertyIsEnumerable.call(e,t)&&(i[t]=e[t])}return i}var l=r.createContext({}),c=function(e){var n=r.useContext(l),t=n;return e&&(t="function"==typeof e?e(n):a(a({},n),e)),t},p=function(e){var n=c(e.components);return r.createElement(l.Provider,{value:n},e.children)},u={inlineCode:"code",wrapper:function(e){var n=e.children;return r.createElement(r.Fragment,{},n)}},d=r.forwardRef((function(e,n){var t=e.components,i=e.mdxType,o=e.originalType,l=e.parentName,p=s(e,["components","mdxType","originalType","parentName"]),d=c(t),m=i,g=d["".concat(l,".").concat(m)]||d[m]||u[m]||o;return t?r.createElement(g,a(a({ref:n},p),{},{components:t})):r.createElement(g,a({ref:n},p))}));function m(e,n){var t=arguments,i=n&&n.mdxType;if("string"==typeof e||i){var o=t.length,a=new Array(o);a[0]=d;var s={};for(var l in n)hasOwnProperty.call(n,l)&&(s[l]=n[l]);s.originalType=e,s.mdxType="string"==typeof e?e:i,a[1]=s;for(var c=2;c<o;c++)a[c]=t[c];return r.createElement.apply(null,a)}return r.createElement.apply(null,t)}d.displayName="MDXCreateElement"},6911:function(e,n,t){t.r(n),t.d(n,{assets:function(){return p},contentTitle:function(){return l},default:function(){return m},frontMatter:function(){return s},metadata:function(){return c},toc:function(){return u}});var r=t(7462),i=t(3366),o=(t(7294),t(3905)),a=["components"],s={},l=void 0,c={unversionedId:"genome_wide_association_studies/Principal components analysis/ld_pruning",id:"genome_wide_association_studies/Principal components analysis/ld_pruning",title:"ld_pruning",description:"Up to the table of contents - Back to the previous page - plink --vcf chr19-clean.vcf.gz --maf 0.01 --indep-pairwise 50 5 0.2 --out chr19-clean Forward to the next page",source:"@site/docs/genome_wide_association_studies/Principal components analysis/ld_pruning.md",sourceDirName:"genome_wide_association_studies/Principal components analysis",slug:"/genome_wide_association_studies/Principal components analysis/ld_pruning",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/ld_pruning",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/genome_wide_association_studies/Principal components analysis/ld_pruning.md",tags:[],version:"current",frontMatter:{},sidebar:"tutorialSidebar",previous:{title:"global_analysis",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/global_analysis"},next:{title:"overview",permalink:"/whg-training-resources/genome_wide_association_studies/Principal components analysis/overview"}},p={},u=[{value:"LD pruning of SNPs",id:"ld-pruning-of-snps",level:3},{value:"Relatedness pruning",id:"relatedness-pruning",level:4}],d={toc:u};function m(e){var n=e.components,t=(0,i.Z)(e,a);return(0,o.kt)("wrapper",(0,r.Z)({},d,t,{components:n,mdxType:"MDXLayout"}),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/Introduction"},"Up to the table of contents")," - ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/overview"},"Back to the previous page")," - plink --vcf chr19-clean.vcf.gz --maf 0.01 --indep-pairwise 50 5 0.2 --out chr19-clean ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/relatedness_pruning"},"Forward to the next page")),(0,o.kt)("h3",{id:"ld-pruning-of-snps"},"LD pruning of SNPs"),(0,o.kt)("p",null,"As described in the population genetics lecture this morning, genetic drift and other processes\nlead to linkage disequilibrium (LD) between SNPs along a chromosome. To ensure the PCs we compute\nrepresent genome-wide structure (not local LD) we'll first carry out LD pruning of our SNP set."),(0,o.kt)("p",null,"LD pruning removes correlated pairs of SNPs so that the remaining SNPs are roughly independent. (It\nalso helps to make subsequent computations quicker.) Run the following command to prune the dataset:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"$ plink --vcf chr19-clean.vcf.gz --maf 0.01 --indep-pairwise 50 5 0.2 --out chr19-clean \n")),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Note"),": commands like this should be typed out or pasted onto a single line in the terminal\nwindow (omitting the ",(0,o.kt)("inlineCode",{parentName:"p"},"$")," of course!). Then press ","<","enter",">"," to run the command. If all goes well\nyou'll see a bunch of output on the screen, starting with some information about the version of\nplink that is running."),(0,o.kt)("p",null,"The above command tells plink to load the file ",(0,o.kt)("inlineCode",{parentName:"p"},"chr19-clean.vcf.gz")," and to prune SNPs, leaving SNPs\nwith minor allele frequency (MAF) at least 1%, and with no pairs remaining with pairwise\nr",(0,o.kt)("sup",null,"2"),">0.2. (The other parameters, here 50 and 5, affect how the computation works in\nwindows across the genome. You can read about the behaviour here:\n",(0,o.kt)("a",{parentName:"p",href:"http://www.cog-genomics.org/plink2/ld"},"http://www.cog-genomics.org/plink2/ld"),")."),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Question"),". Look at the screen output from the above plink command.  How many variants were in the original dataset?  How many were removed because their frequency was below 1%?  How many variants were removed due to LD pruning?  How many variants remain?"),(0,o.kt)("p",null,"Type ",(0,o.kt)("inlineCode",{parentName:"p"},"ls")," or use the file manager to view the directory.  The command above produced a number of files that all begin with the ",(0,o.kt)("inlineCode",{parentName:"p"},"chr19-clean")," prefix.  For our purposes, the most important one is ",(0,o.kt)("inlineCode",{parentName:"p"},"chr19-clean.prune.in"),", as this lists the SNPs that remain after pruning.  Feel free to look at all these files using less or a text editor."),(0,o.kt)("h4",{id:"relatedness-pruning"},"Relatedness pruning"),(0,o.kt)("p",null,"When you're ready, ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/genome_wide_association_studies/Principal%20components%20analysis/relatedness_pruning"},"go to the next page to identify and remove close relationships"),"."))}m.isMDXComponent=!0}}]);