"use strict";(self.webpackChunkwhg_training_resources=self.webpackChunkwhg_training_resources||[]).push([[2066],{3905:function(e,t,n){n.d(t,{Zo:function(){return h},kt:function(){return g}});var a=n(7294);function r(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function o(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);t&&(a=a.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,a)}return n}function i(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?o(Object(n),!0).forEach((function(t){r(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):o(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,a,r=function(e,t){if(null==e)return{};var n,a,r={},o=Object.keys(e);for(a=0;a<o.length;a++)n=o[a],t.indexOf(n)>=0||(r[n]=e[n]);return r}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(a=0;a<o.length;a++)n=o[a],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(r[n]=e[n])}return r}var l=a.createContext({}),p=function(e){var t=a.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):i(i({},t),e)),n},h=function(e){var t=p(e.components);return a.createElement(l.Provider,{value:t},e.children)},c={inlineCode:"code",wrapper:function(e){var t=e.children;return a.createElement(a.Fragment,{},t)}},m=a.forwardRef((function(e,t){var n=e.components,r=e.mdxType,o=e.originalType,l=e.parentName,h=s(e,["components","mdxType","originalType","parentName"]),m=p(n),g=r,d=m["".concat(l,".").concat(g)]||m[g]||c[g]||o;return n?a.createElement(d,i(i({ref:t},h),{},{components:n})):a.createElement(d,i({ref:t},h))}));function g(e,t){var n=arguments,r=t&&t.mdxType;if("string"==typeof e||r){var o=n.length,i=new Array(o);i[0]=m;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"==typeof e?e:r,i[1]=s;for(var p=2;p<o;p++)i[p]=n[p];return a.createElement.apply(null,i)}return a.createElement.apply(null,n)}m.displayName="MDXCreateElement"},5107:function(e,t,n){n.r(t),n.d(t,{assets:function(){return h},contentTitle:function(){return l},default:function(){return g},frontMatter:function(){return s},metadata:function(){return p},toc:function(){return c}});var a=n(7462),r=n(3366),o=(n(7294),n(3905)),i=["components"],s={sidebar_position:2},l="What gene annotation data looks like",p={unversionedId:"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like",id:"programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like",title:"What gene annotation data looks like",description:"Up to table of contents / Back to the introduction / Go the next page",source:"@site/docs/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like.md",sourceDirName:"programming/programming_with_gene_annotations",slug:"/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like",draft:!1,editUrl:"https://github.com/whg-training/whg-training-resources/docs/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like.md",tags:[],version:"current",sidebarPosition:2,frontMatter:{sidebar_position:2},sidebar:"tutorialSidebar",previous:{title:"Introduction",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Introduction"},next:{title:"Processing GFF files",permalink:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code"}},h={},c=[{value:"How are gene annotations created?",id:"how-are-gene-annotations-created",level:3},{value:"What do these data records represent?",id:"what-do-these-data-records-represent",level:3},{value:"What types of records are there?",id:"what-types-of-records-are-there",level:3},{value:"Getting into coding",id:"getting-into-coding",level:2}],m={toc:c};function g(e){var t=e.components,s=(0,r.Z)(e,i);return(0,o.kt)("wrapper",(0,a.Z)({},m,s,{components:t,mdxType:"MDXLayout"}),(0,o.kt)("h1",{id:"what-gene-annotation-data-looks-like"},"What gene annotation data looks like"),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/"},"Up to table of contents")," / ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/Introduction"},"Back to the introduction")," / ",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code"},"Go the next page")),(0,o.kt)("p",null,"The data we'll process is in a format called 'General Feature Format' or ",(0,o.kt)("strong",{parentName:"p"},"GFF"),". You can ",(0,o.kt)("a",{parentName:"p",href:"https://m.ensembl.org/info/website/upload/gff3.html"},"read its specification here"),".  It is a commonly-used format for storing gene annotations."),(0,o.kt)("p",null,"Let's start by taking a look at some GFF data. To make it easy to take a look, I've pasted the top 1000 lines of\na human gene annotation file in the file ",(0,o.kt)("a",{target:"_blank",href:n(7360).Z},"encode.v38.annotation.head.gff"),". This\ncomes from ",(0,o.kt)("a",{parentName:"p",href:"https://www.gencodegenes.org"},"GENCODE"),", a project that curates human and mouse genes. (If you want,\nyou can go there and download the whole file instead.). And I've also placed the first 1000 lines of a\n",(0,o.kt)("em",{parentName:"p"},"P.falciparum")," (malaria) gene annotation file in\n",(0,o.kt)("a",{target:"_blank",href:n(9897).Z},"PlasmoDB-54_Pfalciparum3D7.head.gff"),". (This comes from\n",(0,o.kt)("a",{parentName:"p",href:"https://plasmodb.org"},"PlasmoDB"),".)"),(0,o.kt)("p",null,"Download these and have a look at them (for example using ",(0,o.kt)("inlineCode",{parentName:"p"},"less -S")," in your terminal).  The files look a bit\nlike this:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"##gff-version 3\n#description: evidence-based annotation of the human genome (GRCh38), version 38 (Ensembl 104)\n#provider: GENCODE\n#contact: gencode-help@ebi.ac.uk\n#format: gff3\n#date: 2021-03-12\n##sequence-region chr1 1 248956422\nchr1    HAVANA  gene    11869   14409   .   +   .   ID=ENSG00000223972.5;gene_id=ENSG00000223972.5;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;level=2;hgnc_id=HGNC:37102;havana_gene=OTTHUMG00000000961.2\nchr1    HAVANA  transcript  11869   14409   .   +   .   ID=ENST00000456328.2;Parent=ENSG00000223972.5;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\nchr1    HAVANA  exon    11869   12227   .   +   .   ID=exon:ENST00000456328.2:1;Parent=ENST00000456328.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;exon_number=1;exon_id=ENSE00002234944.1;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\nchr1    HAVANA  exon    12613   12721   .   +   .   ID=exon:ENST00000456328.2:2;Parent=ENST00000456328.2;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;exon_number=2;exon_id=ENSE00003582793.1;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\n...\n")),(0,o.kt)("p",null,"You'll see there is some stuff at the top starting with ",(0,o.kt)("inlineCode",{parentName:"p"},"#"),", that represent ",(0,o.kt)("strong",{parentName:"p"},"metadata")," - i.e. information about the file itself.  And then there are rows that maybe contain genes - but what do they all mean?"),(0,o.kt)("p",null,"Like many bioinformatics file formats, this file doesn't bother to tell you what the columns are - or\neven what they are called. You have to look at ",(0,o.kt)("a",{parentName:"p",href:"https://m.ensembl.org/info/website/upload/gff3.html"},"the spec"),"\nfor that."),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Challenge.")," Figure out what you think the above encodes."),(0,o.kt)("p",null,"Some things to try to figure out:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"Can you figure out what each row represents (for example, the ",(0,o.kt)("inlineCode",{parentName:"li"},"gene")," row above)?"),(0,o.kt)("li",{parentName:"ul"},"How many genes are in the file?  How many exons?"),(0,o.kt)("li",{parentName:"ul"},"What is a transcript anyway?  Does each gene have one transcript, or is it not that simple?"),(0,o.kt)("li",{parentName:"ul"},"The file clearly has seperate rows for genes, transcripts of genes (called 'transcript' or 'mRNA'), and exons.  How are they linked - can you tell?"),(0,o.kt)("li",{parentName:"ul"},"What else is in the 'attributes' column - anything that looks useful?")),(0,o.kt)("p",null,"Spend a few minutes trying to work out what you can about these questions (we'll come back to them below.)"),(0,o.kt)("h3",{id:"how-are-gene-annotations-created"},"How are gene annotations created?"),(0,o.kt)("p",null,"The Ensembl gene annotations are created using a particular pipeline that is described in detail ",(0,o.kt)("a",{parentName:"p",href:"https://www.ensembl.org/info/genome/genebuild/index.html"},"here"),"."),(0,o.kt)("p",null,"The NCBI is a second project that creates gene annotations (NCBI RefSeq).  Their pipeline for Eukaryotes is described ",(0,o.kt)("a",{parentName:"p",href:"https://www.ncbi.nlm.nih.gov/genome/annotation_euk/process/"},"here"),".  The process looks like this:"),(0,o.kt)("p",null,(0,o.kt)("img",{parentName:"p",src:"https://www.ncbi.nlm.nih.gov/core/assets/genome/images/Pipeline_RFAM.png",alt:"diagram of gene annotation pipeline"})),(0,o.kt)("p",null,"Simple!"),(0,o.kt)("p",null,"An important thing to notice is that these annotations are made from severe data sources - including\ncomputational predictions from the genome assemblies themselves, but also alignments of known RNA and protein\nsequences, and a great deal of automatic and for some genomes manual curation. They shouldn't be regarded as\nthe whole truth about genes, but as a current best guess - and for some genomes like humans they are fairly\ncomplete."),(0,o.kt)("h3",{id:"what-do-these-data-records-represent"},"What do these data records represent?"),(0,o.kt)("p",null,"Well, for example let's look at the second row in the extract above:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"chr1    HAVANA  transcript  11869   14409   .   +   .   ID=ENST00000456328.2;Parent=ENSG00000223972.5;gene_id=ENSG00000223972.5;transcript_id=ENST00000456328.2;gene_type=transcribed_unprocessed_pseudogene;gene_name=DDX11L1;transcript_type=processed_transcript;transcript_name=DDX11L1-202;level=2;transcript_support_level=1;hgnc_id=HGNC:37102;tag=basic;havana_gene=OTTHUMG00000000961.2;havana_transcript=OTTHUMT00000362751.1\n")),(0,o.kt)("p",null,"Comparing to ",(0,o.kt)("a",{parentName:"p",href:"https://m.ensembl.org/info/website/upload/gff3.html"},"the specification")," you'll see this indicates:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"the chromosome (chromosome 1),"),(0,o.kt)("li",{parentName:"ul"},'"HAVANA" which is the name of the program that generated the record'),(0,o.kt)("li",{parentName:"ul"},'"transcript" indicating that this record is for a single RNA transcript of a gene'),(0,o.kt)("li",{parentName:"ul"},"the start and end coordinates on chromosome 1 (relative the genome build specified in the metadata)"),(0,o.kt)("li",{parentName:"ul"},"unspecified ",(0,o.kt)("em",{parentName:"li"},"score")," and ",(0,o.kt)("em",{parentName:"li"},"phase")," values"),(0,o.kt)("li",{parentName:"ul"},"that the transcript has '+' strand.  (To understand this, you have to realise that the reference genome assembly directly encodes one strand of DNA, often called the forward strand - there is of course a complementary ",(0,o.kt)("em",{parentName:"li"},"reverse")," strand as well which contains the complementary bases.  Genes are always transcribed in the 5' -> 3' direction, but that could be either on the forward or reverse strand.  This one is transcribed from the forward strand."),(0,o.kt)("li",{parentName:"ul"},"and a bunch of additional data in the 'attributes' column, separated by semicolons.")),(0,o.kt)("p",null,"The attributes clearly contain a bunch of additional information.  Hvae a look at them now."),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"The ID and Parent"),". Two especially important attributes are the ",(0,o.kt)("inlineCode",{parentName:"p"},"ID")," and ",(0,o.kt)("inlineCode",{parentName:"p"},"Parent")," attributes. The ",(0,o.kt)("inlineCode",{parentName:"p"},"ID"),"\nattribute provides an identifier for this record (in this example, it is an\n","[Ensembl ID]","(",(0,o.kt)("a",{parentName:"p",href:"https://www.ebi.ac.uk/training/online/courses/ensembl-browsing-genomes/navigating-ensembl/investigating-a-ge"},"https://www.ebi.ac.uk/training/online/courses/ensembl-browsing-genomes/navigating-ensembl/investigating-a-ge"),"\nne/).)\nThe ",(0,o.kt)("inlineCode",{parentName:"p"},"Parent")," attribute also provides an identifier: it is the identifier of the parent record. For this\nrecord, the parent is 'ENSG00000223972.5', which means that *",(0,o.kt)("em",{parentName:"p"},"this is a transcript of the gene record with\n",(0,o.kt)("inlineCode",{parentName:"em"},"ID=ENSG00000223972.5")," (which occurs right above it in the file.). In this sense the file is "),"relational*: records link to other records that makes them form a hierarchy."),(0,o.kt)("h3",{id:"what-types-of-records-are-there"},"What types of records are there?"),(0,o.kt)("p",null,"The third column of the file above is called the ",(0,o.kt)("inlineCode",{parentName:"p"},"type"),". If you look in the spec you'll see that this contains\nvalues that are part of a controlled 'ontology', which looks ",(0,o.kt)("a",{parentName:"p",href:"http://www.sequenceontology.org/so_wiki/index.php/Category:SO:SOFA"},"a bit\nscary"),". But what's actually in this file? A\nquick way to find out is to use UNIX commands: use the ",(0,o.kt)("inlineCode",{parentName:"p"},"cut")," unix command to extract the ",(0,o.kt)("inlineCode",{parentName:"p"},"type")," column, use ",(0,o.kt)("inlineCode",{parentName:"p"},"grep")," to remove the metadata lines, and then the ",(0,o.kt)("inlineCode",{parentName:"p"},"sort -u")," to sort it and get a unique list:"),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre",className:"language-sh"},"$ cut -f3 gencode.v38.annotation.head.gff | grep -v '#' | sort -u\n$ cut -f3 PlasmoDB-54_Pfalciparum3D7.head.gff | grep -v '#' | sort -u\n")),(0,o.kt)("p",null,"(It would be much better to run this on the real files, not just the top 1000 lines I extracted above.  DO that by downloading the most up-to-date human annotation files now.  For humans this can be found ",(0,o.kt)("a",{parentName:"p",href:"https://www.gencodegenes.org/human/"},"on the GENCODE page")," - you want the 'Comprehensive gene annotation', for the reference chromosomes only, in ",(0,o.kt)("inlineCode",{parentName:"p"},"GFF3")," format.  For malaria parasites, the annotations can be found ",(0,o.kt)("a",{parentName:"p",href:"https://plasmodb.org/plasmo/app/downloads/Current_Release/Pfalciparum3D7/gff/data/"},"here"),"."),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Question:")," do the full files have any more types of record than these top-1000-lines files? What are they?"),(0,o.kt)("p",null,"IN any case What you'll see is that there are quite a few different record types! The file includes information\non genes, transcripts, exons, coding sequence (CDS), untranslated regions (UTR), translation start and stop\ncodons, and more. The ",(0,o.kt)("em",{parentName:"p"},"Plasmodium falciparum")," file is similar but slightly different, with a wider range of\ntypes. (Also, the ",(0,o.kt)("em",{parentName:"p"},"Pf")," file uses ",(0,o.kt)("inlineCode",{parentName:"p"},"type=mRNA")," to denote gene transcripts, while GENCODE uses ",(0,o.kt)("inlineCode",{parentName:"p"},"type=transcript")," -\na difference we'll have to allow for in code.)"),(0,o.kt)("p",null,"If you want to see how this information shouldbe interpreted, try searching for a gene on\n",(0,o.kt)("a",{parentName:"p",href:"http://www.ensembl.org/index.html"},"Ensembl")," or the ",(0,o.kt)("a",{parentName:"p",href:"https://genome.ucsc.edu"},"UCSC Genome Browser"),". E.g. for the gene in the listing above it takes you to\n",(0,o.kt)("a",{parentName:"p",href:"http://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000223972;r=1:11869-14409"},"this page on ENSG00000223972"),". The above transcript, ",(0,o.kt)("inlineCode",{parentName:"p"},"ENST00000456328.2"),", can be seen in the transcript table there. This already gives you a flavour of the complexity, because there are two transcripts on top of each other,\nwith different lengths and different numbers of exons."),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Question:")," Do the same for a protein-coding gene. Does it look any different? Which website do you like\nbetter for looking at genes, Ensembl or the ",(0,o.kt)("a",{parentName:"p",href:"https://genome.ucsc.edu"},"UCSC browser"),"? Spend some time exploring\nthese sites."),(0,o.kt)("h2",{id:"getting-into-coding"},"Getting into coding"),(0,o.kt)("p",null,"In this tutorial we're going to write some code to process these data files. When you're ready,\n",(0,o.kt)("a",{parentName:"p",href:"/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code"},"get started writing some code"),"."))}g.isMDXComponent=!0},9897:function(e,t,n){t.Z=n.p+"assets/files/PlasmoDB-54_Pfalciparum3D7.head-7bfaf2c78ae64f4c7801bd68cd183d06.gff"},7360:function(e,t,n){t.Z=n.p+"assets/files/gencode.v38.annotation.head-57247eb0cd85c862ec70fff216676f07.gff"}}]);