<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-next_generation_sequencing/Building an NGS processing pipeline/pipeline">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">pipeline | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/pipeline/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="pipeline | WHG training resources"><meta data-rh="true" name="description" content="Implementing a NGS data processing pipeline."><meta data-rh="true" property="og:description" content="Implementing a NGS data processing pipeline."><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/pipeline/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/pipeline/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/pipeline/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.a63558cc.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.81209970.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/next_generation_sequencing/Introduction to next-generation sequencing data analysis/">Introduction to next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/">Building an NGS data processing pipeline</a><button aria-label="Toggle the collapsible sidebar category &#x27;Building an NGS data processing pipeline&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Introduction/">Introduction</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Required software/">Required software</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Required_software/">Required_software</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/pipeline/">pipeline</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/">Variant calling, phasing and imputation</a><button aria-label="Toggle the collapsible sidebar category &#x27;Variant calling, phasing and imputation&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/next_generation_sequencing/IGV/">Visualising alignments</a><button aria-label="Toggle the collapsible sidebar category &#x27;Visualising alignments&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/whg-training-resources/programming/introduction_to_R/">programming</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/next_generation_sequencing/"><span itemprop="name">Next-generation sequencing</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/"><span itemprop="name">Building an NGS data processing pipeline</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">pipeline</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>pipeline</h1></header><h2 class="anchor anchorWithStickyNavbar_mojV" id="implementing-a-ngs-data-processing-pipeline">Implementing a NGS data processing pipeline.<a class="hash-link" href="#implementing-a-ngs-data-processing-pipeline" title="Direct link to heading">​</a></h2><p>If you&#x27;ve followed the <a href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Introduction/">introduction</a>, you should now have a bunch of software installed,
including the <code>snakemake</code> pipelining software, and you will have a set of 10 fastq files named in
the format <code>ERR[xxxxxx]_[1|2].fastq.gz</code>.  And you will also have downloaded the <em>P.falciparum</em> reference sequence, <code>Pf3D7_v3.fa.gz</code> - we recommend putting this in its own folder, say <code>data/reference/</code>.</p><p><strong>Challenge.</strong>  Write a snakemake pipeline that processes these reads.  Here are some points to consider.</p><ul><li>Your pipeline should start with the set of fastq read files named by accessions as described in the <a href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Introduction/">introduction</a>.  To keep things well-organised, it&#x27;s a good idea to keep these in a subdirectory, so they will look something like this:</li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR377582_1.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR377582_2.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR377591_1.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR377591_2.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR377629_1.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR377629_2.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR417621_1.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR417621_2.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR417627_1.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   data/reads/ERR417627_2.fastq.gz</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><ul><li><p>You will also need the reference sequence, placed in <code>data/reference/Pf3D7_v3.fa.gz</code>.</p></li><li><p>The pipelines will output a set of BAM files containing these reads aligned to the a reference sequence.  The reads should be coordinate sorted, duplicate read pairs should have been marked (or removed), and the reads should be indexed.  My advice is to put results in a seperate subdirectory, so they will look something like this:</p></li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0033-C.bam</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0033-C.bam.bai</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0041-C.bam</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0041-C.bam.bai</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0049-C.bam</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0049-C.bam.bai</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0056-C.bam</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0056-C.bam.bai</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0088-C.bam</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/aligned/QG0088-C.bam.bai</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><ul><li>The pipeline should also perform QC of the fastq files. We suggest using <code>fastqc</code> and <code>multiqc</code> for this.
So this will output some files that look like this:</li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">    results/qc/QG0033-C.fastqc.html</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/qc/QG0041-C.fastqc.html</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/qc/QG0049-C.fastqc.html</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/qc/QG0056-C.fastqc.html</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/qc/QG0088-C.fastqc.html</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/qc/multiqc_report.html</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>You should of course look at the output to look for anything odd!</p><ul><li>The pipeline will also output a <a href="http://genome.ucsc.edu/goldenPath/help/bedgraph.html" target="_blank" rel="noopener noreferrer"><code>bedgraph</code></a> file for each sample, reporting the coverage at each site in the genome.  (<code>bedtools genomecov -bg</code> is a good way to create this).  These will look like this:</li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">    results/coverage/QG0033-C.coverage.bedgraph</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/coverage/QG0041-C.coverage.bedgraph</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/coverage/QG0049-C.coverage.bedgraph</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/coverage/QG0056-C.coverage.bedgraph</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/coverage/QG0088-C.coverage.bedgraph</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><ul><li>And, if you implement the whole pipeline, it will also output a variant calls file (for this tutorial we suggest using the <a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer"><code>octopus</code> variant caller</a> for this).  These will be an indexed bgzipped <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf" target="_blank" rel="noopener noreferrer">vcf file</a>, which should look like this:</li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">    results/variant_calls/variant_calls.vcf.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    results/variant_calls/variant_calls.vcf.gz.tbi</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Here is a diagram of the overall pipeline:</p><p><img loading="lazy" alt="Diagram of pipeline" src="/whg-training-resources/assets/images/pipeline-f7b1538da80891cc35465d36fdd41f64.svg" width="1722" height="1971" class="img_E7b_"></p><p>You have to implement the green bits... good luck!</p><p>If you are running this as part of a <a href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/pipeline/www.well.ox.ac.uk/">WHG course</a>, we&#x27;ll discuss your pipeline and look at the outputs at the wrap-up session later in the week.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="tips-and-tricks">Tips and tricks<a class="hash-link" href="#tips-and-tricks" title="Direct link to heading">​</a></h3><p>Here is some guidance to help you write your pipeline.  Click the links to jump to the relevant section.</p><ul><li><a href="#But-I-want-to-run-the-yellow-bits-too">But I want to run the yellow bits too!</a></li><li><a href="#How-should-I-put-sample-information-in">How should I put sample information in?</a></li><li><a href="#How-should-I-organise-my-pipeline-files">How should I organise my pipeline files?</a></li><li><a href="#My-snakefiles-are-getting-too-big">My snakefiles are getting too big!</a></li><li><a href="#Keeping-a-fast-iteration-time-during-development">Keeping a fast iteration time during development</a>.</li><li><a href="#Dealing-with-intermediate-files">Dealing with intermediate files</a>.</li><li><a href="#Read-groups-what-now">Read groups what now?</a></li><li><a href="#Whats-in-the-fastq-header">What&#x27;s in the fastq header?</a></li><li><a href="#octopus-is-taking-too-long">Octopus is taking too long!</a></li><li><a href="#what-ploidy">What ploidy?</a></li><li><a href="#tools-that-use-temporary-directories">Tools that use temporary directories</a></li><li><a href="#tips-on-using-bwa-mem">Tips on using <code>bwa mem</code></a>.</li><li><a href="#tips-on-using-samtools">Tips on using <code>samtools</code></a>.</li></ul><h4 class="anchor anchorWithStickyNavbar_mojV" id="but-i-want-to-run-the-yellow-bits-too">But I want to run the yellow bits too!<a class="hash-link" href="#but-i-want-to-run-the-yellow-bits-too" title="Direct link to heading">​</a></h4><p>Be my guest! Running variant annotation in particular would be a good thing to do, as would looking at post-alignment QC metrics.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="how-should-i-put-sample-information-in">How should I put sample information in?<a class="hash-link" href="#how-should-i-put-sample-information-in" title="Direct link to heading">​</a></h4><p>I always put the information about the samples / the needed data in through a <strong>config file</strong>. This
is a file called (say) <code>config.json</code> that you pass in using the <code>--configfile</code> argument.  For example, for this project the <code>config.json</code> might look like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">{</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;reference&quot;: &quot;data/reference/Pf3D7_v3.fa.gz&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;fastq_filename_template&quot;: &quot;data/reads/subsampled/{ID}_{read}.fastq.gz&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;samples&quot;: [</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        { &quot;name&quot;: &quot;QG0033-C&quot;, &quot;ID&quot;: &quot;ERR377582&quot; },</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        { &quot;name&quot;: &quot;QG0041-C&quot;, &quot;ID&quot;: &quot;ERR377591&quot; },</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        { &quot;name&quot;: &quot;QG0049-C&quot;, &quot;ID&quot;: &quot;ERR417627&quot; },</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        { &quot;name&quot;: &quot;QG0056-C&quot;, &quot;ID&quot;: &quot;ERR417621&quot; },</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        { &quot;name&quot;: &quot;QG0088-C&quot;, &quot;ID&quot;: &quot;ERR377629&quot; }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>And I would run snakemake like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">snakemake -s pipelines/master.snakefile --configfile config.json</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The point of this is that it makes it easy to run the pipeline on different sets of data - you just swap out the config file for a different one.</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="how-should-i-organise-my-pipeline-files">How should I organise my pipeline files?<a class="hash-link" href="#how-should-i-organise-my-pipeline-files" title="Direct link to heading">​</a></h4><p>I stick to a specific hierarchy: data goes in the <code>data/</code> folder, analysis results go in the
<code>results</code> folder, and snakemake files and other scripts go in a dedicated <code>pipelines</code> folder.  So it might look something like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">toplevel/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  config.json</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  pipelines/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    master.snakemake</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  data/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    reads/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      ERR377582_1.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      ERR377582_2.fastq.gz</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      ...</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  results/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    qc/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    aligned/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ...</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>I then always run the pipeline from the top-level folder. All the paths used in the pipeline are
either <em>relative to the top-level folder</em>, or are specified in the config file.</p><p>The point of this arrangement is that the <code>pipelines</code> folder contains all of the code for the pipeline, but none of the data. This means it is easy to copy around, or to put it up on github, and so on.</p><p><strong>Note.</strong> The snakemake documentation suggests a <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html" target="_blank" rel="noopener noreferrer">similar, but slightly different layout</a>.</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="my-snakefiles-are-getting-too-big">My snakefiles are getting too big!<a class="hash-link" href="#my-snakefiles-are-getting-too-big" title="Direct link to heading">​</a></h4><p>To fix this, I often use the snakemake <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/modularization.html" target="_blank" rel="noopener noreferrer"><code>include</code>
feature</a> to split up the
file into components of related rules. For example, in our pipeline there are a bunch of rules for
read qc, some for alignment and post-processing, a bunch for variant calling, and a bunch for
computing coverage and so on. So I might have: <code>pipelines/ master.snakmake functions.snakemake
qc.snakemake alignment.snakemake variants.snakemake coverage.snakemake</code></p><p>My <code>master.snakemake</code> then looks like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">include: &quot;functions.snakemake&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">include: &quot;qc.snakemake&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">include: &quot;alignment.snakemake&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">include: &quot;variants.snakemake&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">include: &quot;coverage.snakemake&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>I like this because it keeps related things together, but the files become manageable in size.</p><p><strong>Note.</strong> You&#x27;ll notice I included a <code>functions.snakemake</code> above.  This is because you often need a few python functions to help with your pipeline - for example, mapping from sample names to fastq filenames and so on.  They tend to be re-used so it can be nice to group these into one file.</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="keeping-a-fast-iteration-time-during-development">Keeping a fast iteration time during development.<a class="hash-link" href="#keeping-a-fast-iteration-time-during-development" title="Direct link to heading">​</a></h4><p>When you&#x27;re developing a pipeline, you don&#x27;t want to wait two hours only to discover that it didn&#x27;t
work. A good idea would therefore be to develop your practical using smaller, sub-sampled version
of the datasets (whichever of the above raw data you use). For example, you could run:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">$ gunzip -c filename.fastq.gz | head -n 4000 | gzip -c &gt; filename.subsampled.fastq</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>to take the first few reads from each file.</p><p><strong>Question.</strong> The above command specifies a multiple of 4 lines.  Why?  How many reads does the above command extract?</p><p>If you set your pipeline up the way I suggest above then you can have a config file for the small
test dataset, and then once it is all working, rerun using the real config file specifying the full
dataset.</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="dealing-with-intermediate-files">Dealing with intermediate files<a class="hash-link" href="#dealing-with-intermediate-files" title="Direct link to heading">​</a></h4><p>The alignment steps in <a target="_blank" href="/whg-training-resources/assets/files/pipeline-f7b1538da80891cc35465d36fdd41f64.svg/">our pipeline</a> in particular are notorious for generating intermediate files.  Indeed:</p><ul><li>the alignment step outputs a SAM file...</li><li>which is converted to a BAM file...</li><li>which you then have to sort by position...</li><li>in which you then have to mark the duplicates...</li><li>which are then indexed.</li></ul><p>That&#x27;s at least 3 intermediate files along the way.  We don&#x27;t want to keep these, they were just needed during the pipeline.</p><p>I use a few ways to deal with this.  One way is just to use unix pipes to pipe command together within rules - as in</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">bwa mem reference.fa read1.fq read2.fq | samtools view -b -o aligned.bam</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>However, I find that too much of this makes the pipeline hard to debug (which command failed?  you can&#x27;t tell.)</p><p>Instead, I typically go for temp files and use the snakemake <code>temp()</code> function to tell snakemake files are temporary.  So I might write the alignment rule as:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">rule align_reads:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  input:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    fq1 = something,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    fq2 = something</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  output:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    sam = temp( &quot;results/alignment/{sample_name}.sam&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  shell: &quot;bwa mem ...&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>As you can see, I tend to also put temporary files into their own <code>tmp/</code> folder as well - this avoids cluttering up the results folder when jobs fail.</p><p>Second, rules can refer to other rule outputs, so the next step in the pipeline can be written:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">rule fix_matepair_information_and_convert_to_bam:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  input:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    sam = rules.align_reads.output.sam</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  output:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    bam = temp( &quot;results/alignment/tmp/{sample_name}.bam&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  shell: &quot;samtools fixmate -m {input.sam} {output.bam}&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>and so on down the pipeline.</p><p>Third - <code>snakemake</code> actually has a <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#piped-output" target="_blank" rel="noopener noreferrer">named pipe
output</a> feature, so
you can get the benefit of the UNIX pipe with the same syntax as above - just replace <code>temp()</code> with
<code>pipe()</code> and it should automatically work. (I&#x27;ve never actually used this feature but it&#x27;s a nice
idea for this step, because the SAM file output by <code>bwa</code> might be huge when applied to real data.)</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="read-groups-what-now">Read groups what now?<a class="hash-link" href="#read-groups-what-now" title="Direct link to heading">​</a></h4><p>Some programs require reads to have &#x27;read groups&#x27;. What are they and how do you get them in there?</p><p>BAM files can easily be post-processed and merged.  Read groups are a way to put information in that records the original sample and the sequencing run, so that downstream programs can distinguish these.  The read groups are encoded in the <code>@RG</code> header field of the BAM file (which you can see using <code>samtools view -h</code>), and in the <code>RG</code> tag for each alignment.  A good document on read groups is <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups" target="_blank" rel="noopener noreferrer">this one on the GATK website</a>.</p><p>In our pipeline these don&#x27;t seem that important (we have one alignment file per input fastq file
pair), but in other pipelines the same sample might have been sequenced many times and the results
merged. So for sensible downstream analysis it would be important to keep track of the originating
samples. In particular, variant callers like <code>octopus</code> require you to have read groups in the BAM
file.</p><p>The simplest way to put read groups into the BAM file is to have <code>bwa</code> put them in at the alignment
step.  For our experiment this can be done using the <code>-R</code> option like so:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">bwa mem -R &quot;@RG\tID:ERR377582\tSM:QG0033-C\tPL:ILLUMINA&quot; [other arguments]</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>You can put other stuff into a read group (see below), but the run ID, the sample name, and the platform are all that we need just now.</p><p>Of course you have to be able to generate this for each sample.  With the layout described above, I wrote the following code (which I put in <code>pipelines/functions.snakemake</code>) to do it:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">def find_sample_with_ID( ID ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    samples = config[&#x27;samples&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    sample = [ sample for sample in samples if sample[&#x27;ID&#x27;] == ID ]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    if  len( sample ) != 1:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        raise Exception( &quot;Wrong number of samples found with ID &#x27;%s&#x27; (%d, expected 1).&quot; % ( ID, len( sample )) )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return sample[0]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">def get_read_group_line( ID ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    sample = find_sample_with_ID( ID )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return &quot;@RG\\tID:{ID}\\tSM:{sample}\\tPL:ILLUMINA&quot;.format(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        ID = sample[&#x27;ID&#x27;],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        sample = sample[&#x27;name&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The alignment step can then be updated to use this function:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">rule align_reads:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  input:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    fq1 = something,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    fq2 = something</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  output:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    sam = temp( &quot;results/alignment/{ID}.sam&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  params:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    read_group_spec = lambda wildcards: get_read_group_line( wildcards.ID )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  shell: &quot;&quot;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    bwa mem -R {params.read_group_spec} ...</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  &quot;&quot;&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If you look at the resulting files, they have an <code>@RG</code> header record and <code>RG</code> tags for each read - octopus will then accept these files.</p><p>What else can go in a read group? As the <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups" target="_blank" rel="noopener noreferrer">GATK documentation
indicates</a> the read
group can also contain information about the sequencing flowcell, lane, and sample barcode, and an
identifier for the library itself. Unfortunately some of this information can be hard to come by
depending on where your reads come from. As we describe below, some of it can be obtained from the
read names in the fastq files. For the data in this practical, some parts such as the library
identifier can be found on the <a href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/pipeline/ERR377582/">ENA website</a>. But in general it&#x27;s a bit hard to put it
all together. (Luckily just the sample name and identifier are enough for our analysis.)</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="whats-in-the-fastq-header">What&#x27;s in the fastq header?<a class="hash-link" href="#whats-in-the-fastq-header" title="Direct link to heading">​</a></h4><p>If you look at the header / read name rows of a fastq file you&#x27;ll see they actually contain a bunch
of information - like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">@ERR377582.7615542 HS23_10792:2:2307:6524:31920#15/1</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This row tells us the sample ID (<code>ERR377582</code>) and the read identifier (<code>7615542</code>). And this is
followed by information identifying the instrument that generated the reads (<code>HS23_10792</code>), the
flowcell lane and tile number in the lane (<code>2:2307</code>), the coordinates of the
<a href="https://www.broadinstitute.org/files/shared/illuminavids/clusterGenSlides.pdf" target="_blank" rel="noopener noreferrer">cluster</a> within the
tile (<code>6524</code>, <code>31920</code>), a number identifying the index of the sample within a multiplexed set of
samples (i.e. all run at the same time; <code>#15</code>), and whether it&#x27;s read 1 or 2.</p><p>Some of this info can be put in the read group as well.</p><p><strong>Note.</strong> The format of this information is not standard across platforms, and it changes depending
on your data provider. Some other examples can be found <a href="https://en.wikipedia.org/wiki/FASTQ_format#Illumina_sequence_identifiers" target="_blank" rel="noopener noreferrer">on
wikipedia</a> or on the
<a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups" target="_blank" rel="noopener noreferrer">GATK read groups page</a>.</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="octopus-is-taking-too-long">Octopus is taking too long!<a class="hash-link" href="#octopus-is-taking-too-long" title="Direct link to heading">​</a></h4><p>The <a href="https://github.com/luntergroup/octopus" target="_blank" rel="noopener noreferrer">Octopus variant caller</a> can take a long time to do its work - hopefully reflecting that it is trying its best to make high-quality variant calls.  This might take too long to run on your laptop.  If so, here are some options for speeding it up:</p><ul><li><p>If you have a multi-core CPU, you can use more threads (<code>--threads</code> argument).</p></li><li><p>Restrict to a set of regions.  You can add the <code>--regions</code> option to tell Octopus to only work on specified regions.  For this tutorial, please include these regions: <code>--regions Pf3D7_02_v3:616190-656190 Pf3D7_02_v3:779288-859288 Pf3D7_11_v3:1023035-1081305</code>.  (This brought the calling down to about half and hour when I tried it.)</p></li><li><p>You could also try the Octopus &#x27;fast&#x27; or &#x27;very fast&#x27; modes - though I haven&#x27;t tried this.</p></li></ul><p>See <code>octopus --help</code> for a full list of options.</p><p>(In general this might be less of a problem for real work as you might run it a compute cluster.)</p><p>Another option is to try a different variant caller - <a href="https://gatk.broadinstitute.org/hc/en-us" target="_blank" rel="noopener noreferrer"><code>GATK HaplotypeCaller</code></a>, <a href="https://www.nature.com/articles/nbt.4235" target="_blank" rel="noopener noreferrer"><code>DeepVariant</code></a> or <a href="https://github.com/freebayes/freebayes" target="_blank" rel="noopener noreferrer"><code>freebayes</code></a> might be possibilities.  For a simple approach you could also use <a href="https://samtools.github.io/bcftools/bcftools.html" target="_blank" rel="noopener noreferrer"><code>bcftools mpileup</code> and <code>bcftools call</code></a> (this is likely to be substantially faster as it relies on the input alignments does not try to reconstruct local haplotypes near each variant.)</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="what-ploidy">What ploidy?<a class="hash-link" href="#what-ploidy" title="Direct link to heading">​</a></h4><p>Blood-stage malaria parasites are <a href="https://www.cdc.gov/malaria/about/biology/index.html" target="_blank" rel="noopener noreferrer">haploid</a>.  So set octopus <code>--organism-ploidy 1</code>.</p><p>On the other hand, <a href="https://doi.org/10.7554/eLife.40845.001" target="_blank" rel="noopener noreferrer">mixed infections are common</a>, and to handle this most projects actually treat samples as if they were diploid - they then treat heterozygote calls as &#x27;mixed&#x27; calls.  This is <em>ad hoc</em> but works ok.  So you could set <code>--organism-ploidy 2</code>.</p><p>For the purposes of this tutorial you could do either - or both so we can see the difference?</p><p><a href="#Tips-and-tricks">Go back to the tips and tricks</a>.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="tools-that-use-temporary-directories">Tools that use temporary directories<a class="hash-link" href="#tools-that-use-temporary-directories" title="Direct link to heading">​</a></h4><p>Some tools have that bad habit of leaving &#x27;stuff&#x27; in the directory you run them in. This is really
annoying for pipelines because you don&#x27;t want that - you want to put the temp stuff away somewhere
out of the way. <code>octopus</code> is one of these tools: if you run it you&#x27;ll see it outputs a directory
called <code>octopus-temp</code>.</p><p>Really the only way to deal with this is use the program help to find the option that renames the
temp dir - then send it somewhere different. For example:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">octopus [other options] --temp-directory-prefix results/variants/tmp/octopus/</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This will probably work here. (In general you may need to use snakemake wildcards etc. to name this temp directory so it doesn&#x27;t clash if the same rule runs multiple jobs in parallel.)s</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="tips-on-using-bwa-mem">Tips on using <code>bwa mem</code><a class="hash-link" href="#tips-on-using-bwa-mem" title="Direct link to heading">​</a></h4><p>Here are a few options you can use to <code>bwa mem</code> which you might want to consider using.</p><ul><li><p>Run <code>bwa mem</code> for a list of options.</p></li><li><p>The basic usage is <code>bwa mem -o output.sam [path to indexed fasta file] read1.fq.gz read2.fq.gz</code></p></li><li><p>The <code>-t</code> option tells <code>bwa</code> to use more than one thread.  (If doing this, make sure to also <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#threads" target="_blank" rel="noopener noreferrer">tell snakemake the number of threads it will use</a>.)</p></li><li><p>The <code>-R</code> option specifies that <code>bwa</code> should include a read group header line and read groups tags in the output - this is <a href="#read-groups-what-now">described above</a>.</p></li><li><p>By default, <code>bwa</code> will output only the best alignment for each read.  However, if you specify the <code>-a</code> option then <code>bwa</code> will output multiple possible alignments for each read, but all except one will be marked as &#x27;not a primary alignment&#x27; (using the <a href="https://broadinstitute.github.io/picard/explain-flags.html" target="_blank" rel="noopener noreferrer"><code>SAM flags</code></a>).  This can be useful in cases when you want to consider possible alternate alignments.</p></li></ul><p>There are also other aligners out there. <a href="https://github.com/lh3/minimap2" target="_blank" rel="noopener noreferrer"><code>minimap2</code></a> is another
good choice. Nevertheless the field typically currently uses <code>bwa mem</code> for Illumina short-read NGS
data.</p><h4 class="anchor anchorWithStickyNavbar_mojV" id="tips-on-using-samtools">Tips on using <code>samtools</code><a class="hash-link" href="#tips-on-using-samtools" title="Direct link to heading">​</a></h4><p>Here are some tips on using <code>samtools</code>:</p><ul><li><p>Run <code>samtools</code> for a list of commands, and <code>samtools [command]</code> for a list of options for each command.</p></li><li><p>Some commands, like <code>samtools markdup</code>, take the output filename as a seperate argument.  But others, such as <code>samtools view</code> or <code>samtools sort</code>, want you to use the <code>-o</code> option to specify the output file (otherwise they output to standard output).</p></li><li><p><code>multiqc</code> can read <code>samtools stats</code> output, useful for post-alignment QC.</p></li></ul><h2 class="anchor anchorWithStickyNavbar_mojV" id="good-luck">Good luck!<a class="hash-link" href="#good-luck" title="Direct link to heading">​</a></h2></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/next_generation_sequencing/Building an NGS processing pipeline/pipeline.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/next_generation_sequencing/Building an NGS processing pipeline/Required_software/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Required_software</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/next_generation_sequencing/variant_calling_and_imputation/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Variant calling, phasing and imputation</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#implementing-a-ngs-data-processing-pipeline" class="table-of-contents__link toc-highlight">Implementing a NGS data processing pipeline.</a><ul><li><a href="#tips-and-tricks" class="table-of-contents__link toc-highlight">Tips and tricks</a></li></ul></li><li><a href="#good-luck" class="table-of-contents__link toc-highlight">Good luck!</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.a63558cc.js"></script>
<script src="/whg-training-resources/assets/js/main.81209970.js"></script>
</body>
</html>