<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">Bayesian meta-analysis practical | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Bayesian meta-analysis practical | WHG training resources"><meta data-rh="true" name="description" content="In this practical we will investigate a simple method to investigate sharing of genetic effects between traits."><meta data-rh="true" property="og:description" content="In this practical we will investigate a simple method to investigate sharing of genetic effects between traits."><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.a63558cc.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.81209970.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/whg-training-resources/programming/introduction_to_R/">programming</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Hidden markov models/">README</a><button aria-label="Toggle the collapsible sidebar category &#x27;README&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/">GMS Statistical Modelling I - introduction to statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;GMS Statistical Modelling I - introduction to statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/notes/Probability solutions/">notes</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">practicals</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">a_tale_of_two_2x2_tables</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/">bayesian_meta_analysis</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/">Bayesian meta-analysis practical</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical/">linear_regression</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/">malaria_exposure</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison/">o_bld_grp</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/distributions warmup/">warmup</a></div></li></ul></li></ul></li></ul></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/"><span itemprop="name">Statistical modelling</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/Introduction/"><span itemprop="name">GMS Statistical Modelling I - introduction to statistical modelling</span></a><meta itemprop="position" content="2"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">practicals</span><meta itemprop="position" content="3"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">bayesian_meta_analysis</span><meta itemprop="position" content="4"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Bayesian meta-analysis practical</span><meta itemprop="position" content="5"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Bayesian meta-analysis practical</h1><p>In this practical we will investigate a simple method to investigate sharing of genetic effects between traits.
Specifically we will investigate 107 SNPs that have been associated with seven autoimmune-related conditions (from <a href="https://doi.org/10.1371/journal.pgen.1002254" target="_blank" rel="noopener noreferrer">https://doi.org/10.1371/journal.pgen.1002254</a>), and for each SNP ask: are these effects restricted to a single trait, shared across all traits, or somewhere in between?</p><p>To do this we will use a simple Bayesian method in which we consider all possible models of effect sharing across traits and
compute the evidence in the data for each model.  Like our other Bayesian analyses so far, we will do this by specifying prior parameters
generated by making reasonable guesses.</p><p>We use data from Cotsapas et al PLoS Genetics (2011)
Data is effect size estimates and standard errors for 107 SNPs and 7 autoimmune traits.  We&#x27;d like to know - to what extent are genetic effects shared across these traits?</p><p>To answer this we&#x27;ll use a Bayesian approach to meta-analysis across traits.  This considers all models of association with one trait, with two traits, and so on up to all traits.  It uses Bayes Factors to compute relative strengths of evidence.</p><p>Using Bayes factors allows us to make nice computations of posterior weights.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="loading-the-data">Loading the data<a class="hash-link" href="#loading-the-data" title="Direct link to heading">​</a></h2><p>Load the data and get the reported effect size estimates (betas), standard errors, and P-values as follows:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">cotsapas = load( &quot;practicals/cotsapas_data.RData&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames( cotsapas ) = cotsapas$SNP</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">betas = as.matrix( cotsapas[, grep( &quot;.beta&quot;, colnames( cotsapas ))] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ses = as.matrix( cotsapas[, grep( &quot;.SE&quot;, colnames( cotsapas ))] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Ps = as.matrix( cotsapas[, grep( &quot;[.]P&quot;, colnames( cotsapas ))] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">phenotypes = gsub( &quot;.beta&quot;, &quot;&quot;, grep( &quot;.beta&quot;, colnames( cotsapas ), value = T ), fixed = T )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="computing-bayes-factors">Computing Bayes factors<a class="hash-link" href="#computing-bayes-factors" title="Direct link to heading">​</a></h2><p>According to the lemma about multiplying multivariate normals (MVNs) - see the <!-- -->[Normal times normal
is normal]<!-- -->(../../notes/Normal times normal is normal.pdf) document, the product of an MVN prior and
an MVN likelihood is equal to an MVN multiplied by a constant. And the constant is itself
computable by an MVN evaluated at the effect size estimate.</p><p>We use the <code>mvtnorm</code> package to compute this for any given model, specificed as a 7x7 prior
covariance matrix, against the null model under which all true effects are zero.</p><p>(The implemention below additionally deals with missing beta estimates).</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">library( mvtnorm )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">compute.bf &lt;- function( betas, ses, prior ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    w = which( !is.na( betas ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dmvnorm(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        betas[w],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        sigma = prior[w,w] + diag( ses[w]^2 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ) / dmvnorm(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        betas[w],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        sigma = diag( ses[w]^2 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="choosing-priors-and-weights">Choosing priors and weights<a class="hash-link" href="#choosing-priors-and-weights" title="Direct link to heading">​</a></h2><p>We will evaluate every possible combination of trait associations - the <code>expand.grid()</code> function
can be used to enumerate these:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">priors = expand.grid(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    RA = c(0,1),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    PS = c(0,1),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    MS = c(0,1),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    SLE = c(0,1),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    CD = c(0,1),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    CeD = c(0,1),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    T1D = c(0,1)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames( priors ) = sapply( 1:nrow( priors ), function(i) { paste( priors[i,], collapse = &quot;&quot; )})</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">priors = priors[-1,] # remove the null model</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head(priors)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>For our initial analysis we will weight each model such that there is the same prior
that k phenotypes are associated, for each k = 1,... 7.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">prior.phenotype.counts = rowSums( priors )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">weights = matrix(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    sapply( prior.phenotype.counts, function( count ) { 1 / length( which( prior.phenotype.counts == count ))}),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    nrow = 1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">) / 7</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This prior weighting means there is more weight on (say) the single model of all traits, than there is on each of the 7 models of association with a single trait, or the (7 choose k) models of association with k traits (1 &lt; k &lt; 7).</p><p>This function is used to compute prior matrices for any given prior from the above table:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">compute.prior.matrix &lt;- function( row, rho = 0.9, sd = 0.2 ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    prior = matrix( 0, 7, 7 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    w = which( row == 1 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    prior[w,w] = rho * sd^2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    diag(prior)[w] = sd^2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return( prior )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The prior matrix has a nonzero variance in the diagonal entries corresponding to 1&#x27;s in the priors
table. And it assumes a correlation (by default 0.9) between the effects with nonzero variance.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="computing-bfs">Computing BFs<a class="hash-link" href="#computing-bfs" title="Direct link to heading">​</a></h2><p>We have to compute 127 BFs for each of 107 SNPs.  Let&#x27;s do it in a loop and store it in a matrix.</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">bfs = matrix( NA, nrow( cotsapas ), nrow( priors ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dimnames = list( cotsapas$SNP, rownames( priors ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sd = 0.2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for( i in 1:nrow( cotsapas )) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    beta_hat = betas[i,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    se = ses[i,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    for( j in 1:nrow(priors)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        prior = compute.prior( priors[j,] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        bfs[i,j] = compute.bf( beta_hat, se, prior )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The BFs represent evidence against the null.  Here, we aren&#x27;t interested in the null model (all SNPs have been chosen because they are associated).  But we are interested in the posterior probability of each model.  We can compute that by a) multiplying by the prior weights and b) normalising so that posteriors sum to 1:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">posteriors = matrix( NA, nrow( cotsapas ), nrow( priors ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dimnames = list( cotsapas$SNP, rownames( priors ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for( i in 1:nrow( posteriors ) ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    posteriors[i,] = bfs[i,] * weights</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    posteriors[i,] = posteriors[i,] / sum( posteriors[i,], na.rm = T )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>## Summarising the analysis</p><p>Now we can get some information out.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="summarising-by-count">Summarising by count<a class="hash-link" href="#summarising-by-count" title="Direct link to heading">​</a></h3><p>First, we could look at how many traits are associated in general.
To compute this we summarise by adding up the models with a given number of traits:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># Compute posterior on each number of effects</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">byCount = matrix(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    0,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    nrow = nrow( priors ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ncol = 7,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dimnames = list( rownames( priors ), 1:7 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for( n in 1:7 ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    byCount[ which( prior.phenotype.counts == n ),i] = 1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">count.posteriors = posteriors %*% byCount</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>For example we could find SNPs where there is evidence for more than one effect:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">which( sum( count.posteriors[,2:7] ) &gt; count.posteriors[,1] )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h3 class="anchor anchorWithStickyNavbar_mojV" id="summarising-by-phenotype">Summarising by phenotype<a class="hash-link" href="#summarising-by-phenotype" title="Direct link to heading">​</a></h3><p>We could also summarise by phenotype - e.g. to find SNPs consistent with an effect for a given phenotype:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">byPhenotype = matrix(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    0,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    nrow = nrow( priors ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ncol = 7,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dimnames = list( rownames( priors ), phenotypes )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for( i in 1:7 ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    byPhenotype[which( priors[,i] == 1 ),i] = 1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">phenotype.posteriors = posteriors %*% byPhenotype</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>E.g. to find the expected number of associations per phenotype:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">colSums( phenotype.posteriors )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>(Chrohn&#x27;s disease has the most)</p><h2></h2><p>How many SNPs have (say) &gt; twice, or greater than 10 times the posterior mass on models with at least 2 effects?</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">length( which( rowSums( count.posteriors[,2:7] ) &gt; 2 * count.posteriors[,1] ) )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">length( which( rowSums( count.posteriors[,2:7] ) &gt; 10 * count.posteriors[,1] ) )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Compare our results with those of Cotsapas et al paper.  They identified 47 SNPs with evidence of more than one association.  Their method, though, looked for statistical significance of &gt;1 effect and didn&#x27;t take into account effect size sharing.  Which of these is right?  Is either right?</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="interesting-findings">Interesting findings<a class="hash-link" href="#interesting-findings" title="Direct link to heading">​</a></h2><p>Some SNPs are identified as having effects on all traits - yet have very non-significant P-values:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">count.posteriors[&#x27;rs11755527&#x27;,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">phenotype.posteriors[&#x27;rs11755527&#x27;,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">betas[&#x27;rs11755527&#x27;,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ses[&#x27;rs11755527&#x27;,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Ps[&#x27;rs11755527&#x27;,]</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>What does this mean?  </p><p>Presumably, the -ve estimates for MS and CD imply that other -ve estimates appear likely under our model.  So our analysis concludes there are effects on all phenotypes.  (This is despite us assuming a +ve correlation between effects - this is not a strong condition in a MVN.)</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="caution">Caution<a class="hash-link" href="#caution" title="Direct link to heading">​</a></h2><p>This analysis highlights a typical aspect of a Bayesian analysis - it requires some thought.
This is both a burden on the bayesian analyst, but also a great advantage (c.f. the discussion in lectures of statistical and conceptual models).  In short you will have to do the thought anyway and this is a good place to do it.</p><p>Here, we have a particular prior assumption that each number of associations has equal weight.
Is that realistic?</p><p>We also have a prior assumption on the correlation between traits - is that realistic?</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="empirical-bayes">Empirical Bayes<a class="hash-link" href="#empirical-bayes" title="Direct link to heading">​</a></h2><p>We could now implement an &#x27;empirical Bayes&#x27; approach by learning new model weights using all the SNPs we&#x27;ve seen:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">empirical.weights = colSums( posteriors )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We would then recompute BFs using the empirical weights.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">A tale of two 2x2 tables</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">atp2b4_practical</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#loading-the-data" class="table-of-contents__link toc-highlight">Loading the data</a></li><li><a href="#computing-bayes-factors" class="table-of-contents__link toc-highlight">Computing Bayes factors</a></li><li><a href="#choosing-priors-and-weights" class="table-of-contents__link toc-highlight">Choosing priors and weights</a></li><li><a href="#computing-bfs" class="table-of-contents__link toc-highlight">Computing BFs</a><ul><li><a href="#summarising-by-count" class="table-of-contents__link toc-highlight">Summarising by count</a></li><li><a href="#summarising-by-phenotype" class="table-of-contents__link toc-highlight">Summarising by phenotype</a></li></ul></li><li><a href="#interesting-findings" class="table-of-contents__link toc-highlight">Interesting findings</a></li><li><a href="#caution" class="table-of-contents__link toc-highlight">Caution</a></li><li><a href="#empirical-bayes" class="table-of-contents__link toc-highlight">Empirical Bayes</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.a63558cc.js"></script>
<script src="/whg-training-resources/assets/js/main.81209970.js"></script>
</body>
</html>