<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">Model checking the malaria exposure finding | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Model checking the malaria exposure finding | WHG training resources"><meta data-rh="true" name="description" content="Consider the following table from :"><meta data-rh="true" property="og:description" content="Consider the following table from :"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.a63558cc.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.81209970.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/whg-training-resources/programming/introduction_to_R/">programming</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Hidden markov models/">README</a><button aria-label="Toggle the collapsible sidebar category &#x27;README&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/">GMS Statistical Modelling I - introduction to statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;GMS Statistical Modelling I - introduction to statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/notes/Probability solutions/">notes</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">practicals</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">a_tale_of_two_2x2_tables</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/">bayesian_meta_analysis</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical/">linear_regression</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/">malaria_exposure</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/">Model checking the malaria exposure finding</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/When asymptotics go wrong/">When asymptotics go wrong</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison/">o_bld_grp</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/distributions warmup/">warmup</a></div></li></ul></li></ul></li></ul></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/"><span itemprop="name">Statistical modelling</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/Introduction/"><span itemprop="name">GMS Statistical Modelling I - introduction to statistical modelling</span></a><meta itemprop="position" content="2"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">practicals</span><meta itemprop="position" content="3"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">malaria_exposure</span><meta itemprop="position" content="4"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Model checking the malaria exposure finding</span><meta itemprop="position" content="5"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Model checking the malaria exposure finding</h1><p>Consider the following table from <a href="https://science.sciencemag.org/content/348/6235/711" target="_blank" rel="noopener noreferrer">https://science.sciencemag.org/content/348/6235/711</a>:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">                                        rs60822373 G allele        rs60822373 C allele</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">unexposed populations (Europeans)                      1965        1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">malaria-exposed populations (Africans).                 707        17</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>It shows counts, taken from the 1000 Genomes Project data (<a href="https://www.internationalgenome.org" target="_blank" rel="noopener noreferrer">https://www.internationalgenome.org</a>), of the reference and non-reference allele of <code>rs60822373</code> in two groups of populations: European-ancestry individuals, and individuals from sub-Saharan Africa.</p><p>The implication of above paper, which was published in the journal Science, is that the rs60822373 &#x27;C&#x27; allele is likely under positive selection due to malaria.  The table above seems consistent with this, because the main burden of mortality due to malaria occurs in children in sub-Saharan African populations.  If rs60822373 were malaria-protective, individuals carrying it would have a better chance of surviving in these populations, so that the allele might have been pushed to high frequency by natural selection.</p><p>But how much evidence for natural selection does this table provide?</p><p>In this practical we&#x27;re going to use genome-wide genotype data to ask whether the table above really consistent with evidence for positive natural selection.</p><p><em>QN.</em> Before going further - look at the table and decide what you think about this.  (You can of course apply some statistics here - e.g. a chi-squared test (<code>chisq.test()</code>) or a Fisher&#x27;s exact test (<code>fisher.test()</code>) will give you some sense of the <em>statistical</em> strength of the data.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="modelling-the-2x2-table">Modelling the 2x2 table<a class="hash-link" href="#modelling-the-2x2-table" title="Direct link to heading">​</a></h2><p>Let&#x27;s load the table into R:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">exposure.table = matrix(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    c(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        1965, 1,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        707, 17</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ncol = 2,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    byrow = T,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dimnames = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        c( &quot;non-exposed&quot;, &quot;exposed&quot; ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        c( &quot;G&quot;, &quot;C&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Let&#x27;s model this table.</p><p>Individuals in the 1000 Genomes Project were sampled based on their ancestry.  Thus, the <em>row sums</em> of the above table were essentially determined by the study design and we can treat them as fixed.</p><p>This means we can model the rows as binomially distributed.  Let ϑ₁ and ϑ₂ be the true allele frequencies of the &#x27;C&#x27; allele in nonexposed and exposed populations.  Then we could think of the table as being generated by two binomial distributions:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           G    C        (total)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">unexposed  a    b        (N₁ = a+b)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  exposed  c    d        (N₂ = c+d)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  b ~ binomial( n = N₁, p = ϑ₁ )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  d ~ binomial( n = N₂, p = ϑ₂ )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This is great because we know how to implement that (using the code we wrote in class):</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># You need binomial.ll from the stats modelling session first!</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">table.ll &lt;- function(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  data,                                   # This is our table, as a 2x2 matrix</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  params = list( theta = c( 0.5, 0.5 ))   # Theta values for each row.</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # We assume binomial sampling in rows</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # And that the rows are independent</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # So multiply likelihoods (sum the log-likelihoods) over the rows</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return (</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        binomial.ll(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            x = data[1,2],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            params = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                n = rowSums( data )[1],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                p = params$theta[1]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        + binomial.ll(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            x = data[2,2],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            params = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                n = rowSums( data )[2],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                p = params$theta[2]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>Note:</em> coding-wise we&#x27;ve cheated a bit here - according to our description, really the row sums should form parameters rather than part of the data.  But this way is simpler to write the code. so let&#x27;s go with it.</p><p>Let&#x27;s look at the maximum likelihood model fit:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inspect.ll(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  table.ll,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  data = exposure.table,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  params = list( theta = c( 0.5, 0.5 ) )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>(The rows are independent here so really we are just plotting two binomial log-likelihoods).</p><p>It should be immediately obvious that our nice, gaussian, asymptotic behaviour described in the stats modelling session, is not really working for this table.
Especially the first row looks pretty skewed.  This is an important point which is described further in [<!-- -->[when asymptotics go wrong.md]<!-- -->].  However, there&#x27;s amore serious problem with this table that we&#x27;ll get to below.</p><h1>A better model via a log odds ratio</h1><p>A problem with the parameterisation above is that none of the parameters captures what we are really interested in - the difference in frequency between the two rows. </p><p>Another problem is that the parameters live in a slightly awkward space - they are in the unit interval <!-- -->[0,1]<!-- --> but are very close over the boundary.
This makes them kind of difficult to work with, both in code and mathematically.  (If you recall, the <code>binomial.ll()</code> function we wrote contains special code to make it give sensible answers if you accidentally try a value outside the interval).</p><p>A better parameterisation of the table for our problem works by making two changes as follows.</p><p>First, we replace the frequencies ϑ₁ and ϑ₂ with <strong>log odds</strong> parameters as follows:</p><p><img loading="lazy" src="https://render.githubusercontent.com/render/math?math=%5Ctextstyle+%5Cgamma_i+%3D+log+%5Cleft%28%5Cfrac%7B%5Ctheta_i%7D%7B1-%5Ctheta_i%7D%5Cright%29" alt="\gamma_i = log \left(\frac{\theta_i}{1-\theta_i}\right)" class="img_E7b_"></p><p>Why the log odds?  </p><p>The quintessential way of doing this is to reparameterise in terms of a <strong>baseline log odds</strong> (representing the frequency of C in the first row) and a <strong>log odds ratio</strong> (</p><p>To fix that we are going to reparameterise in terms of two other parameters - a <strong>baseline</strong> parameter that</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="when-asymptotics-goes-wrong">When asymptotics goes wrong<a class="hash-link" href="#when-asymptotics-goes-wrong" title="Direct link to heading">​</a></h2><p>Look at the top row of the plot.  This picture is typical for binomial likelhoods when the frequency is low or there&#x27;s not much data - the asymptotic approximation does not work well.  The likelihood becomes skewed, and the normal approximation starts to fail (here overstating the amount of uncertainty on the left, and understating it on the right.)</p><p>To show this more clearly - because this is just a binomial, we can also do an exact computation of the estimate and its confidence interval (this function also computes a P-value against the null model that the frequency is zero) using the <code>binom.test()</code> function, like this:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt; binom.test( n = 1966, x = 1, p = 0 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    Exact binomial test</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data:  1 and 1966</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">number of successes = 1, number of trials = 1966, p-value &lt; 2.2e-16</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">alternative hypothesis: true probability of success is not equal to 0</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">95 percent confidence interval:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"> 1.287774e-05 2.830707e-03</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sample estimates:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">probability of success </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           0.000508647 </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The estimate is <code>0.000508647</code> (which is 1/1966) and the 95% confidence interval is reported as <code>1.287774e-05</code> to <code>2.830707e-03</code>.  On the other hand the asymptotic method above computes the standard error as <code>0.0005093123</code> (see the function output) so we can compute asymptotic estimated confidence intervals as:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">theta = 1/1966</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">se = 0.0005093123</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sprintf( &quot;95%% CI = %.4f (%.4f - %.4f)&quot;, theta, theta - 1.96 * se, theta + 1.96 * se )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This reports <code>95% CI = 0.0005 (-0.0005 - 0.0015)</code>.  As the plot indicates, this asymptotic confidence interval overstates uncertainty on the left but understates it on the right.  Similarly, an asymptotic computation of a P-value (for example the Wald test P-value, which just looks at how far the estimate is in the tails of the approximating normal distribution) gets it somewhat wrong - here is a <strong>one-sided test</strong> against the null that ϑ₁ = 0:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># compute Wald test p-value using cumulative distribution function of Gaussian:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">pnorm( theta, mean = 0, sd = 0.0005093123, lower.tail = F ) </span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This reports the P-value as <code>0.15</code>.  But, since we&#x27;ve observed one C allele, it is <em>completely impossible</em> that the frequency is zero!  The true p-value is actually zero, because ϑ₁ = 0 <em>never</em> generates any C alleles at all.</p><p>On the other hand, when there&#x27;s lots of data and the frequency is not close to zero (e.g. the second row of the plot) things are fine.  The likelihood becomes a fairly well-rounded citizen (literally so - it becomes approximately normal, i.e. its log has approximately constant curvature).</p><h2></h2><p>We can compare this to our fit by converting back to standard errors like this:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">  theta1 = 0.000508647 </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  ci = c( 1.287774e-05, 2.830707e-03 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  se = theta1 - ci</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Thus the rows</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Essentially this means the row sums in the above table were determined beforehand.  A reasonable model for the data is therefore *that the rows are binomially distributed*.  If θ stands for</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>row1<br>
<!-- -->row2 </p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">We can therefore model the tbale like so:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">```R</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt; theTable</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">               G  C</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">non-exposed 1965  1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exposed      707 17</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The allele frequencies in the top and bottom row are very different - the C allele has almost zero frequency in the top row, but around
The odds ratio computed from this table is 47 - implying the frequencies of this variant are very different in European and African populations.
The Egan et al paper suggests this is evidence for malaria-driven natural selection of this allele.</p><p>How could we test that?  Well one way is </p><p>And a test of this table ( <code>chisq.test()</code> or <code>fisher.test()</code> shows that this is highly statistically significant.  That means:</p><ul><li><em>If</em> the data were truly sampled from a binomial distribution with given frequency in each row.</li><li>and <em>if</em> the true odds ratio was zero.</li><li><em>then</em> a table with such a large odds ratio would almost never occur.</li></ul><p>(<em>NB.</em> The assumptions of Fisher&#x27;s exact test and of the chi-squared test, which corresponds to &#x27;binomial sampling in rows&#x27;, are slightly different.
Fisher&#x27;s exact test is actually conditional on both row and column sums, which reduces things to one parameter.  This doesn&#x27;t matter much for most tables in practice because the row and column sums are not usually very informative about the odds ratio itself.)</p><p>Does this table provide evidence that the malaria-exposed population allele frequency is higher because of natural selection due to malaria?  (That&#x27;s the implicit message of this part of the original paper: <!-- -->[https://science.sciencemag.org/content/348/6235/711]<!-- -->(Egan et al Science 2015) ).</p><p>One of the advantages of genome-wide analysis is that we have a lot of data to work with - and this can be used for model checking. Suppose
we compare our finding with all other &#x27;similar&#x27; variants in the genome. They have all been
genotyped on the same set of samples, and as such they represent the same sampling biases (if any) and the same demography. However, due to recombination they also represent many independent draws from the genealogical history of the sample.</p><p>The file <code>allele_counts_with_exposed_count\=17.csv.gz</code> contains data for all the alleles in the 1000 Genomes populations studied, that have the same allele count of 17 in the exposed populations. Load it:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">X = read.csv(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;allele_counts_with_exposed_count=17.csv.gz&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    header = T,      # Tell R our file has a row of column names</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    as.is = T,       # Ask R to please not transmogrify any data</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    check.names = F  # Also ask R not to mess around with column names </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Take a look</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head(X)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Let&#x27;s first sanity check I&#x27;ve computed the data right.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">range( X$`exposed:B` )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>It&#x27;s often best to first collect numerical data into a matrix - this is a container of numbers that is layed out nicely in memory for computation.  (Unlike a data frame which holds columns of data of different types).  The <code>as.matrix()</code> function does this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data = as.matrix( X[, c( &quot;unexposed:A&quot;, &quot;unexposed:B&quot;, &quot;exposed:A&quot;, &quot;exposed:B&quot; )])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Take a look</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head( data )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Let&#x27;s compute the frequency of each variant:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">X$frequency = (data[,2] + data[,4]) / rowSums(data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># plot it</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hist(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    X$frequency,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    breaks = 50, # how many bars across the plot?</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;Allele frequency&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">abline( v = X$frequency[ which( X$rsid == &#x27;rs60822373&#x27; )], col = &#x27;red&#x27; )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>In this dataset we have conditioned on the exposed population frequency, so it&#x27;s proabably more sensible to look at the frequency in non-exposed populations:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">hist(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    X$`unexposed:B`,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    breaks = 50,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;log Odds ratio&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    main = &quot;Alleles with count = 17 in malaria-exposed populations&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># The value for our SNP is 1, so let&#x27;s replot clamped to 50 to focus better on that area</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hist(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pmin( X$`unexposed:B`, 50 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    breaks = 50,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;Count of allele in non-exposed pops&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    main = &quot;Alleles with count = 17 in malaria-exposed populations&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    #xlim = c( -0.5, 0.5 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    #xaxt = &#x27;n&#x27;  # Turn off the axis so we can draw our own</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Let&#x27;s put a line where rs60822373 is:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">abline(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    v = X$`unexposed:B`[ which( X$rsid == &#x27;rs60822373&#x27; )],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    col = &#x27;red&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">text(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    x = X$`unexposed:B`[ which( X$rsid == &#x27;rs60822373&#x27; )],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    y = 50000,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    adj = 0,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pos = 4, # draw text to right</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    label = &quot;rs60822373&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    col = &#x27;red&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Sure enough there are a bunch of alleles at higher counts in non-exposed populations.  But there is also a big spike at zero. </p><p>We could also mesaure differences directly using the log-odds ratio:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">X$logOR = log( ((data[,1]) * (data[,4])) / ( (data[,2]) * (data[,3])) )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Unfortunately if you do this you&#x27;ll find many are infinite (due to a 0 in the table):</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">length( which( X$logOR == Inf ))</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We will therefore compute the log OR by adding a dummy count of  1 to each cell.  This amounts to adding prior information.  This amounts to adding a prior.  (If you want to see what that prior looks like, plot the <code>table.ll.reparameterised()</code> function on a table full of 1&#x27;s.)</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">X$logOR = log( ((data[,1]+1) * (data[,4]+1)) / ( (data[,2]+1) * (data[,3]+1)) )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Let&#x27;s look at the distribution of the logOR:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># Compute log OR with additional prior count of 1 in each cell to avoid infinities.</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hist(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    X$logOR,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    breaks = 50,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;log Odds ratio&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    main = &quot;Alleles with count = 17 in malaria-exposed populations&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">abline(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    v = X$logOR[ which( X$rsid == &#x27;rs60822373&#x27; )],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    col = &#x27;red&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>(Because the 3rd and 4th cells of our table are constant here, this has exactly the same information as above but expressed as a log OR).</p><p>We originally computed a P-value which indicated that the table was very unlikely to arise by chance under the null.  But now we find that many other variants have at least as extreme counts! We can use our data to get an empirical P-value:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">empirical.P = length( which( X$`unexposed:B` &lt;= 1 )) / nrow(X)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The answer is <code>0.49</code>.  So about half the SNPs in the genome with this low count in malaria-exposed populations, have equally low frequencies in non-exposed populations.  They can&#x27;t all be under selection can they?</p><p>The null model says that all variation in the effect size (the log OR) is due to sampling variation true log OR = 0).  It should be distributed as a N(0, SE^2).  (all the standard errors will be similar here because of the choice of variant).  To test how well the null model fits, we can plot standard-error-adjusted effect sizes (&quot;Z-scores&quot;) against a gaussian distribution.  First compute the standard error:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># for table</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">#   a b</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">#   c d</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># the formula is sqrt( 1/a + 1/b + 1/c + 1/d )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">X$SE = sqrt( rowSums( 1/(data+1) ))</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>Warning</em>: this standard error is not quite correct due to our use of a prior additional count of 1 to each cell, which is fixed and not sampled.  This won&#x27;t matter for this exercise.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">region = c( -10, 10 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hist(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    X$logOR / X$SE,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    breaks = 100,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;x score&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlim = region,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ylim = c( 0, 0.5 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    main = &quot;Alleles matching O blood group frequency&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    freq = FALSE # plot y axis values as a density, for comparison</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x = seq( from = region[1], to = region[2], by = 0.01 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">points(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    x,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dnorm( x, mean = 0, sd = 1 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    type = &#x27;l&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    col = &#x27;red&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The null model, at least as expressed through the log OR and its standard error, is not even close to being true.</p><p>For completeness / comparison, a more sensitive plot that we don&#x27;t really need here, is a quantile-quantile plot, which plots ordered statistics against their expection.  We can do that here:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">zscores = sort( X$logOR / X$SE )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">expected = qnorm( (1:length(zscores))/(length(zscores)+1)) # Normal distribution quantiles</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    expected,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    zscores,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pch = 19 # nice round dots</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">abline( a = 0, b = 1, col = &#x27;red&#x27; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">grid()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If the expected and true distributions were similar the points would lie near the line.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="interpretation">Interpretation<a class="hash-link" href="#interpretation" title="Direct link to heading">​</a></h2><p>Even though this table and the one for O blood group looked superficially similar, they differ in a couple of ways.  </p><p>One is that the small counts in some cells of this table mean we are at the boundary of where traditional statistics based on asymptotic assumptions holds.  This was the problem with the infinite log ORs and means we should be careful about just rolling out standard tools.  (The original authors used &#x27;Fishers Exact Test&#x27; which does deal with this issue, although it introduces additional assumptions.)</p><p>Another more serious issue is that the assumed null model - that frequencies should be similar between the two groups - is not even close to being correct for most variants in the genome.  This is a serious issue which should cause us to reevaluate our assumptions.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/simulated_linear_model_practical/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">simulated_linear_model_practical</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/When asymptotics go wrong/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">When asymptotics go wrong</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#modelling-the-2x2-table" class="table-of-contents__link toc-highlight">Modelling the 2x2 table</a></li><li><a href="#when-asymptotics-goes-wrong" class="table-of-contents__link toc-highlight">When asymptotics goes wrong</a></li><li><a href="#interpretation" class="table-of-contents__link toc-highlight">Interpretation</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.a63558cc.js"></script>
<script src="/whg-training-resources/assets/js/main.81209970.js"></script>
</body>
</html>