<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-programming/programming_with_gene_annotations/Getting_started_writing_some_code">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">Processing GFF files | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Processing GFF files | WHG training resources"><meta data-rh="true" name="description" content="Up to table of contents / Back to the previous page / Go to the next page"><meta data-rh="true" property="og:description" content="Up to table of contents / Back to the previous page / Go to the next page"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.a63558cc.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.81209970.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/whg-training-resources/programming/introduction_to_R/">programming</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/">Introduction to R</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to R&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/introduction_to_command_line/">Introduction to the Command Line</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to the Command Line&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/">Programming with gene annotations</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming with gene annotations&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Introduction/">Introduction</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/">What gene annotation data looks like</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/">Processing GFF files</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/">Refactoring_makes_code_better</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/">Converting_gff_to_sqlite</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/">Gathering some first statistics</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them/">Memory_issues_and_how_to_solve_them</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/">Counting_genes_2</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/">Counting_genes_3</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths/">Getting_sequence_lengths</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/">How_much_of_the_genome_is_in_genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Scaling_up/">Scaling_up</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Visualisation/">Visualisation</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Closing_thoughts/">Closing_thoughts</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_organised_and_scaling_up/">Getting_organised_and_scaling_up</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/List_of_ensembl_files/">List_of_ensembl_files</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Solutions/">Solutions</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Where_next/">Where_next</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/parsing_gff3_the_low_level_way/">parsing_gff3_the_low_level_way</a></li></ul></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">programming</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/programming_with_gene_annotations/"><span itemprop="name">Programming with gene annotations</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Processing GFF files</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Processing GFF files</h1><p><a href="/whg-training-resources/programming/programming_with_gene_annotations/">Up to table of contents</a> / <a href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/">Back to the previous page</a> / <a href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/">Go to the next page</a></p><p>Your task, if you choose to accept it, is to write some code that processes a GFF file, makes sense of the
data, and uses it to gather some interesting statistics. And then to apply it to analyse genes from multiple
species. By doing this we can hopefully learn something about the genome biology of the world&#x27;s organisms.</p><p>The aim of this tutorial is that <strong>you</strong> to write the code that does this (although you can either do this on
your own or working together in a group). There will be lots of support, and this tutorial will guide you
through one way to do it. If you&#x27;re doing this as part of a WHG course, we&#x27;ll talk through the results and the
code itself in the discussion sessions.</p><div class="admonition admonition-tip alert alert--success"><div class="admonition-heading"><h5><span class="admonition-icon"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="16" viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>tip</h5></div><div class="admonition-content"><p>Remember we&#x27;re not writing this code for it&#x27;s own sake but to answer our questions like the ones in our introduction:</p><ul><li>How many genes are there?</li><li>How big are they?</li><li>How much of the genome is in genes?</li><li>How complex are genes - How many exons?  How many different transcripts?</li><li>How much of genes is actually protein-coding sequence - and how much is untranslated?</li><li>How much do these patterns differ across species?</li></ul></div></div><p>How to start writing this code?  Well there are a few ways:</p><p><strong>Do it yourself.</strong> It may well be that you already have a good idea how to go about this. If so,
feel free to dive straight in. You&#x27;re free to use any language or system you like for this -
standard options might be <a href="https://www.python.org" target="_blank" rel="noopener noreferrer">python</a> or <a href="https://cran.r-project.org" target="_blank" rel="noopener noreferrer">R</a>, but
you could also use <a href="https://julialang.org" target="_blank" rel="noopener noreferrer">julia</a>, or even
<a href="https://en.wikipedia.org/wiki/C%2B%2B" target="_blank" rel="noopener noreferrer">C++</a> or another compiled language.  Make sure you <a href="/whg-training-resources/prerequisites/">have these installed first</a>.</p><p><strong>Use a package.</strong> If you search, you will be able to find packages for your favourite programming language
that parse GFF3 for you - or perhaps a library that processes gene annotations at a higher level. Now, using
that would to some extent defeat the purpose of the exercise of this tutorial (which is about coding), but on
the other hand what we&#x27;re really interested in is genes rather than the coding itself. So if you want to take
that route and it gets you to better answers quicker, go ahead!</p><p><strong>Write everything from scratch.</strong> It&#x27;s also quite possible to do this task in (say) base python or R without
using any existing libraries. In fact that can be a good way to it because it gives you lots of control over
how it works, and (as we&#x27;ll see later) you might need control over things like performance and memory usage.</p><p>This tutorial will take a &#x27;middle&#x27; way. We will use python but will use the popular <a href="https://pandas.pydata.org" target="_blank" rel="noopener noreferrer"><code>pandas</code> dataframe
library</a> library to begin reading and manipulating the data. <code>pandas</code> is a natural
fit here because the GFF3 data is basically tabular in format (many rows x 9 named columns) and so it ought to
fit well in a dataframe. Unpacking the <code>attributes</code> column is a bit tricky of course - this will take most of
the work.</p><p>In the course of the tutorial we&#x27;ll develop a little python module to help us answer the questions.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="diving-straight-in---parsing-data">Diving straight in - parsing data<a class="hash-link" href="#diving-straight-in---parsing-data" title="Direct link to heading">​</a></h2><p>If you <a href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/">looked at the gene annotation data</a>, you&#x27;ll know
that it comes in rows of data that are tab-delimited, but that it is also relational (meaning that
the records refer to each other, via the <code>Parent</code> attribute). Moreover, since exons are associated
with transcripts, which are in turn associated with genes, we ultimately have to build some form of
hierarchical data structure to capture this.</p><p>That sounds complex, but we can break off a manageable first bit of the job by just focussing on
getting the data in (c.f. &quot;keep it simple&quot;). So let&#x27;s do the simplest thing possible and start by
writing a function to load the data.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">def parse_gff3_to_dataframe( data ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;&quot;&quot;Parse data in GFF3 format, and return a pandas dataframe&quot;&quot;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    result = (!! do some work here)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # more code here to add to result</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return result</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><strong>Note:</strong> If you are not familiar with python syntax, now would be a good time to refresh via any
of the available tutorials. The above code defines a function, and shows a documentation comment
(the <code>&quot;&quot;&quot;...&quot;&quot;&quot;</code> bit) and also has code comments (starting with <code>#</code>).</p><p>This code doesn&#x27;t work yet (you have to write the missing bits) but it already illustrates a few things that
might be helpful if you&#x27;re not used to writing code. First, the name is pretty clear about what this function
does (indeed it makes the documentation comment pretty useless at the moment). I spent quite a while deciding
on that name! Second, even before we&#x27;ve written it, we can see the function is going to follow a very simple
pattern: it creates a new thing (the result of the function, so it is called <code>result</code>) and returns it on the
last line. All the function has to do is build <code>result</code> - simple!</p><p>The other thing to see is that, even in this prototype state, this function is already reasonably testable.
Look, let&#x27;s test it with some fake data, like this for example:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">##gff-version 3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">#description: test data</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr1   me    gene    1    1000    .    +    .    ID=gene1;other_data=stuff</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr1   me    exon    10    900    .    +    .    ID=gene1.1;Parent=gene1</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Let&#x27;s encode that in python:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">test_data = &quot;&quot;&quot;##gff-version 3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">#description: test data</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr1\tme\tgene\t1\t1000\t.\t+\t.\tID=gene1;other_data=stuff</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr1\tme\texon\t10\t900\t.\t+\t.\tID=gene1.1;Parent=gene1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&quot;&quot;&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>(Where &#x27;\t&#x27; is python-speak for &#x27;a tab character&#x27;).</p><p>Here is a test:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">def test_parse_gff3_to_dataframe( data ):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    from io import StringIO</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    from math import isnan</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # 1. run our function:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    data = parse_gff3_to_dataframe( StringIO( data ))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # 2. test it:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert data[&#x27;seqid&#x27;][0] == &#x27;chr1&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert data[&#x27;strand&#x27;][0] == &#x27;+&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert data[&#x27;attributes&#x27;][0] == &#x27;ID=gene1;other_data=stuff&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert data[&#x27;start&#x27;][1] == 10 # start and end are integers</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert data[&#x27;end&#x27;][0] == 1000</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert isnan( data[&#x27;score&#x27;][1] ) # &quot;.&quot; indicates missing data in the GFF spec</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert data[&#x27;ID&#x27;][0] == &#x27;gene1&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert data[&#x27;ID&#x27;][1] == &#x27;gene1.1&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    assert data[&#x27;Parent&#x27;][1] == &#x27;gene1&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # etc.</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>(<strong>Note:</strong> The <code>StringIO</code> bit above is just there to handle the fact that we are passing in a string of data to our function. Normally we will pass in a file.  The <code>isnan()</code> function is there to test for not-a-number values.)</p><p>If you run this right now, you&#x27;ll get and error, something like:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt;&gt;&gt; test_parse_gff3_to_dataframe( test_data )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Traceback (most recent call last):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  File &quot;&lt;stdin&gt;&quot;, line 1, in &lt;module&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  File &quot;&lt;stdin&gt;&quot;, line 6, in test_parse_gff3_to_dataframe</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">NameError: name &#x27;parse_gff3_to_dataframe&#x27; is not defined</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The task is just to get this test to pass.</p><p>This leads us to a:</p><p><strong>Challenge:</strong> write <code>parse_gff3_to_dataframe()</code>. Can you write this so all the tests pass?  Or at least some of them?</p><p><strong>Hints:</strong> This challenge does have a few complexities to it.  Here are some points to think about that might be helpful:</p><ul><li><p>pandas has a <a href="https://pandas.pydata.org/docs/reference/api/pandas.read_table.html" target="_blank" rel="noopener noreferrer"><code>read_table</code> function</a> that is a good way to get the data in.  It has many arguments that control how the data is parsed.</p></li><li><p>The data itself doesn&#x27;t have column names in. But the test clearly requires them - you have to
add them in somehow.</p></li><li><p>To pass the tests you have to pay attention to missing data! Check the <a href="https://m.ensembl.org/info/website/upload/gff3.html" target="_blank" rel="noopener noreferrer">GFF
spec</a> for how these are represented. See
below for a couple of possible ways of dealing with it.</p></li><li><p>Also, some of the columns <a href="https://m.ensembl.org/info/website/upload/gff3.html" target="_blank" rel="noopener noreferrer">have different data
types</a> like integers and floats - to pass
the tests you also have to get the types right.</p></li><li><p>The last three rows of the test use the <code>ID</code> and <code>Parent</code> fields. If you <a href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/">examined the
data</a> you&#x27;ll know these aren&#x27;t seperate columns in GFF3, but instead
live inside the <code>attributes</code> column. So somehow you&#x27;ll have to pull these out into new columns.</p></li></ul><p><strong>More hints:</strong></p><ul><li><p>Be careful about whitespace. Python really cares about your indentation. In particular, you should decide at the
start if you are going to indent with spaces or tabs and stick with it (otherwise you will get all sorts of strange syntax
errors).</p></li><li><p>It is very worth looking around for an editor you like. On the Mac,
<a href="https://macromates.com" target="_blank" rel="noopener noreferrer">Textmate</a> or <a href="https://www.sublimetext.com" target="_blank" rel="noopener noreferrer">Sublime</a> or <a href="https://atom.io" target="_blank" rel="noopener noreferrer">github
Atom</a> might be good choices. On Windows
<a href="https://notepad-plus-plus.org" target="_blank" rel="noopener noreferrer">Notepad++</a> or <a href="https://atom.io" target="_blank" rel="noopener noreferrer">Atom</a> might work. Sublime and
Atom also work on linux. Of course there are also many others you can use.</p></li><li><p>It&#x27;s a good idea to write little helper functions to do bits of the task. (For example, it might be useful to write
a <code>parse_attributes()</code> function that parses a semi-colon-separated list of key=value pairs and return a dict, e.g. so that this:</p></li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt; parse_attributes( &#x27;ID=gene1;other_data=stuff&#x27; )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>would produce this object:
{
&quot;ID&quot;: &quot;gene1&quot;,
&quot;other_data&quot;: &quot;stuff&quot;
}</p><ul><li><p><code>.split()</code> can be used to split strings.  For example <code>&quot;Hello;world&quot;.split( &quot;;&quot; ) == [ &quot;Hello&quot;, &quot;world&quot; ]</code></p></li><li><p>If you want to apply a function to every row of a data frame - the <a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.apply.html" target="_blank" rel="noopener noreferrer">pandas <code>.apply()</code>
method</a> is the ticket.
For example, <code>int</code> is a function, so you could do:</p></li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">    df[&#x27;start&#x27;].apply( int )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><ul><li>How do handle missing data? One way is to use the python <a href="https://mail.python.org/pipermail/python-dev/2005-September/056846.html" target="_blank" rel="noopener noreferrer">syntax for conditional
expressions</a>:</li></ul><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># convert value to an int</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">value = None if value == &quot;.&quot; else int(value)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>But a better way for this task may be to exploit the <code>na_values</code> argument of <code>pandas.read_table</code>.</p><p><strong>Yet more hints:</strong> The <a target="_blank" href="/whg-training-resources/assets/files/gff_first_version-f9b8b73915b2712648ba982c8ad27505.py/"><code>solutions/part1/gff_first_version.py</code></a> file contains my
first go at a solution to this. As a comparison,
<a href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/solutions/part1/gff_python_version.py/"><code>solutions/part1/gff_python_version.py</code></a> also implements a similar pure python
version called <code>parse_gff3_to_list()</code>. (There are lots of other ways to write this of course - for example, gathering
the code into a class might be sensible - but I&#x27;ve gone with functions for simplicity.)</p><p>If my code still looks too complicated to you - that&#x27;s because it is. Having written the first version, I had a go<a href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/">refactored
the code</a> to produce a <a target="_blank" href="/whg-training-resources/assets/files/gff-9caed837b4ef044cc415f0a851eff819.py/">nicer version</a>.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="a-note-on-writing-tests">A note on writing tests<a class="hash-link" href="#a-note-on-writing-tests" title="Direct link to heading">​</a></h3><p>I&#x27;d hazard a guess that not many people writing scientific code think of writing their tests first, like we did above.
However, it is a very useful approach, because it forces you to think about how your code will be used before you spend
the effort of writing it. Also, writing tests first is one way to ensure you&#x27;re in the 1% of programmers whose code is
actually tested at all!</p><p>If you follow <a href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/">the section on refactoring</a>, you will also end up with code that is
easy to test at high and low levels. To illustrate this I&#x27;ve written a proper <a target="_blank" href="/whg-training-resources/assets/files/test_gff-526a50c301ebdf96eabe45cc8c29765c.py/">test for my version of<code>gff.py</code></a> using the python <a href="https://docs.python.org/3/library/unittest.html" target="_blank" rel="noopener noreferrer">unittest
module</a>. It includes the test above, but also tests individual
functions in the module.</p><p>(If you run this test using <code>python3 test_gff.py</code>, you&#x27;ll see that my <code>add_ID_and_Parent()</code> function actually has a bug
that kicks in if you give it strange <code>attributes</code> values.)</p><p>How much time and effort should you spend on testing? This depends on what you&#x27;re trying to achieve. If your script is
a one-off exploratory analysis, it might not matter that much. However, when you&#x27;re deep in a piece of scientific
research you don&#x27;t want to be worried about <a href="https://doi.org/10.1038/nm1107-1276b" target="_blank" rel="noopener noreferrer">off-by-one errors</a>. I suggest
writing at least some tests to make sure the code does what it says.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="using-the-code">Using the code<a class="hash-link" href="#using-the-code" title="Direct link to heading">​</a></h3><p>Do you have a working function <code>parse_gff3_to_dataframe()</code>? Then let&#x27;s <a href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/">turn this into a useful
program</a>.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/programming/programming_with_gene_annotations/Getting_started_writing_some_code.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">What gene annotation data looks like</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Refactoring_makes_code_better</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#diving-straight-in---parsing-data" class="table-of-contents__link toc-highlight">Diving straight in - parsing data</a><ul><li><a href="#a-note-on-writing-tests" class="table-of-contents__link toc-highlight">A note on writing tests</a></li><li><a href="#using-the-code" class="table-of-contents__link toc-highlight">Using the code</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.a63558cc.js"></script>
<script src="/whg-training-resources/assets/js/main.81209970.js"></script>
</body>
</html>