<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-programming/programming_with_gene_annotations/Converting_gff_to_sqlite">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">Converting_gff_to_sqlite | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Converting_gff_to_sqlite | WHG training resources"><meta data-rh="true" name="description" content="Up to table of contents"><meta data-rh="true" property="og:description" content="Up to table of contents"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.a63558cc.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.81209970.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/whg-training-resources/programming/introduction_to_R/">programming</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/">Introduction to R</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to R&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/introduction_to_command_line/">Introduction to the Command Line</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to the Command Line&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/">Programming with gene annotations</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming with gene annotations&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Introduction/">Introduction</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/What_gene_annotation_data_looks_like/">What gene annotation data looks like</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/">Processing GFF files</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/">Refactoring_makes_code_better</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/">Converting_gff_to_sqlite</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/">Gathering some first statistics</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Memory_issues_and_how_to_solve_them/">Memory_issues_and_how_to_solve_them</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/">Counting_genes_2</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_3/">Counting_genes_3</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_sequence_lengths/">Getting_sequence_lengths</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/How_much_of_the_genome_is_in_genes/">How_much_of_the_genome_is_in_genes</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Scaling_up/">Scaling_up</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Visualisation/">Visualisation</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Closing_thoughts/">Closing_thoughts</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_organised_and_scaling_up/">Getting_organised_and_scaling_up</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/List_of_ensembl_files/">List_of_ensembl_files</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Solutions/">Solutions</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/Where_next/">Where_next</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/parsing_gff3_the_low_level_way/">parsing_gff3_the_low_level_way</a></li></ul></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">programming</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/programming_with_gene_annotations/"><span itemprop="name">Programming with gene annotations</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Converting_gff_to_sqlite</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>Converting_gff_to_sqlite</h1></header><p><a href="/whg-training-resources/programming/programming_with_gene_annotations/">Up to table of contents</a></p><p><a href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/">Back to the previous page</a></p><p><a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/">Go to the next page</a></p><h2 class="anchor anchorWithStickyNavbar_mojV" id="writing-a-useful-conversion-program">Writing a useful conversion program.<a class="hash-link" href="#writing-a-useful-conversion-program" title="Direct link to heading">​</a></h2><p>In the <a href="/whg-training-resources/programming/programming_with_gene_annotations/Getting_started_writing_some_code/">first part of this tutorial</a> you will have written some code to parse a
GFF3 file and output a pandas dataframe.  I&#x27;ll assume your function is called <code>parse_gff3_to_dataframe()</code>.</p><p>Your function is already useful! To demonstrate this, let&#x27;s convert the GFF file into a different format - a <a href="https://www.sqlite.org" target="_blank" rel="noopener noreferrer">sqlite
database</a> database. <code>sqlite</code> is a very useful file format that works like a full database, but
doesn&#x27;t require a server or anything like that - it&#x27;s just a single file on the filesystem. But unlike a flat file, it
is relational, making it easy to link records together. (This is <a href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/[https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_sql.html/#compare-with-sql-join">similar to pandas
functionality</a> but doesn&#x27;t need to load all the data into memory.)</p><p>To get started, put your function in a file called <code>gff.py</code>. If you start python in the same
directory you should now be able to write:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">import gff</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data = gff.parse_gff3_into_dataframe(&#x27;PlasmoDB-54_Pfalciparum3D7.head.gff&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">print(data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Congratulations - you&#x27;ve written a python module!</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="writing-a-command-line-program">Writing a command-line program<a class="hash-link" href="#writing-a-command-line-program" title="Direct link to heading">​</a></h3><p>We will use your <code>gff</code> module to write a program called <code>gff_to_sqlite.py</code> which:</p><ul><li><p>uses the <a href="https://docs.python.org/3/library/argparse.html" target="_blank" rel="noopener noreferrer">argparse</a> library to process
command-line arguments for the input and output files.</p></li><li><p>loads the data using your function above.</p></li><li><p>stores the data in the output sqlite file.</p></li></ul><p>This turns out to be pretty easy. We&#x27;ll start by importing the relevant stuff, and writing a
function to parse the arguments. I find the <a href="https://docs.python.org/3/library/argparse.html" target="_blank" rel="noopener noreferrer">argparse
documentation</a> a bit confusing, so here is
a quick head start:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">import argparse, sqlite3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">import gff</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">def parse_arguments():</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    parser = argparse.ArgumentParser(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        description = &quot;&quot;&quot;Convert a GFF3 file to sqlite3 format.</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        The result will be a table with the GFF3 fields, and with ID and Parent fields in columns.</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        The resulting table will be indexed by the ID field for easy lookup.&quot;&quot;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    parser.add_argument(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        &#x27;--input&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        help =&#x27;The path of a GFF3-formatted file to work with&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        required = True</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # add other needed arguments here</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return parser.parse_args()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">args = parse_arguments()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If you put this in a file called <code>gff_to_sqlite.py</code> then you can run it from your shell like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">$ python3 gff_to_sqlite.py </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">usage: gff_to_sqlite.py [-h] --input INPUT</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gff_to_sqlite.py: error: the following arguments are required: --input</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>I&#x27;ll leave you to fill in the other needed arguments here. You will need an <code>--output</code> option to
specify the output file, and because sqlite files can contain several tables, maybe also a
<code>--table</code> option to specify the output table name. (If you specify, say, <code>default = &quot;gff_data&quot;</code> for
this, instead of <code>required=True</code>, the argument will have a default value.)</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="parsing-the-data">Parsing the data<a class="hash-link" href="#parsing-the-data" title="Direct link to heading">​</a></h3><p>This is easy right?  You&#x27;ve already written it:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data = gff.parse_gff_to_dataframe( args.input )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h3 class="anchor anchorWithStickyNavbar_mojV" id="outputting-to-sqlite">Outputting to sqlite<a class="hash-link" href="#outputting-to-sqlite" title="Direct link to heading">​</a></h3><p>This turns out to be easy too. Pandas dataframes have a <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_sql.html" target="_blank" rel="noopener noreferrer"><code>.to_sql()</code>
function</a>
that does this for you. To make this work, you open the database and then run it:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">db = sqlite3.connect( args.output )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data.to_sql( args.table, db, index = False )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Finally you can, if you like, add a SQL index on the ID (or other columns) so lookups are quick:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">db.execute( &quot;CREATE INDEX IF NOT EXISTS `%s_ID_index` ON `%s`( ID )&quot; % (args.table, args.table ))</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Job done!  Run it in your shell like this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">$ python gff_to_sqlite.py --input ../PlasmoDB-54_Pfalciparum3D7.head.gff --output genes.sqlite</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>To check ths worked, you can use the <a href="/whg-training-resources/programming/programming_with_gene_annotations/Converting_gff_to_sqlite/www.sqlite.org/">sqlite3 command-line client</a> to look at the
result:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">$ sqlite3 -header -column genes.sqlite &quot;SELECT COUNT(*) FROM gff_data WHERE type == &#x27;gene&#x27;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">$ sqlite3 -header -csv genes.sqlite &quot;SELECT * FROM gff_data WHERE type == &#x27;gene&#x27;&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">$ sqlite3 -line genes.sqlite &quot;SELECT * FROM gff_data WHERE attributes LIKE &#x27;%Name=ABO%&#x27;&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Or load it back into python:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt;&gt;&gt; import pandas, sqlite3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&gt;&gt;&gt; db = sqlite3.connect( &quot;genes.sqlite&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&gt;&gt;&gt; pandas.read_sql( &quot;SELECT Parent, COUNT(*) AS number_of_transcripts FROM gff WHERE type == &#x27;mRNA&#x27; GROUP BY Parent&quot;, db )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Or load it into R:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">library( RSQLite )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">db = dbConnect( dbDriver( &quot;SQLite&quot; ), &quot;genes.sqlite&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dbGetQuery( db, &quot;SELECT Parent, COUNT(*) AS number_of_transcripts FROM gff WHERE type == &#x27;mRNA&#x27; GROUP BY Parent&quot; )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>etc.</p><p>This combination of between-language interoperability and flexible querying are the main reasons to
write this script in the first place. There&#x27;s also another advantage which has to do with
controlling memory, that we&#x27;ll <a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_2/">come back to later</a>.</p><p><strong>Warning:</strong> Do check the output file size. For big organisms it can get quite big quite fast.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="job-not-done-yet">Job not done yet<a class="hash-link" href="#job-not-done-yet" title="Direct link to heading">​</a></h2><p>To make this genuinely useful there are a few things still to do.</p><ol><li><p>You don&#x27;t really want to leave your users hanging while your script processes stuff. It will
take not much of your time to add some <code>print()</code> statements to the script to tell the user what it
is doing, so you should do this. It will make you look good.</p></li><li><p>Your script is not so great at the moment if you want to process several files into the same
database - say several different organisms from
<a href="http://ftp.ensembl.org/pub/current_gff3/" target="_blank" rel="noopener noreferrer">Ensembl</a>. (Try running the above twice and you&#x27;ll see
what I mean.) For one thing, you don&#x27;t really know if the user wants to <em>append</em> the data to the
same table, or <em>overwrite</em> the table with the new one (this is a good candidate for another
command-line argument.). For another thing, if you did run it twice it would be difficult to figure
out which record came from which input file. Because of this I find it helps a lot to add an output
column that records an analysis name to distinguishes it from the other analyses. (This could
either be taken from another command-line argument, or taken from the input filename. I tend to
call this column <code>analysis</code>.)</p></li><li><p>The script can take a while (minutes) to process some of the larger files. It would be very worthwhile
to profile it to figure out if there are obvious slow bits that can be sped up. Profiling is beyond the scope of this
tutorial, but at some point you&#x27;re going to need it to make code fast. <a href="https://docs.python.org/3/library/profile.html" target="_blank" rel="noopener noreferrer">Here&#x27;s how you can profile code in
python</a>.)</p></li><li><p>How much memory does the script use?  (Try running it and watching <code>top</code> in another terminal).</p></li></ol><p>My version of this code implements some of the above - it is in
<a target="_blank" href="/whg-training-resources/assets/files/gff_to_sqlite_dataframe_version-b3cd86e00f4afc6c8d5dae626249f2dc.py/"><code>solutions/part1/gff_to_sqlite_dataframe_version.py</code></a>. There is also <a target="_blank" href="/whg-training-resources/assets/files/gff_to_sqlite_python_version-d6ed86511df6a752656ae7c3dca04e44.py/">a version that does not rely on pandas</a>.</p><p>I also tried profiling this code on a human gff file by running:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">python3 -m cProfile gff_to_sqlite.py [arguments...]</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If I read the profiling output right, it is
taking 60% of its time inside <code>parse_attributes()</code>. This means I could probably make it quite a bit
faster with some simple changes to <code>parse_gff3_to_dataframe()</code> - we&#x27;ll get back to this later.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="back-to-the-task">Back to the task<a class="hash-link" href="#back-to-the-task" title="Direct link to heading">​</a></h2><p>Now let&#x27;s get <a href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/">back to learning about genes</a>.</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/programming/programming_with_gene_annotations/Converting_gff_to_sqlite.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/programming/programming_with_gene_annotations/Refactoring_makes_code_better/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Refactoring_makes_code_better</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/programming/programming_with_gene_annotations/Counting_genes_1/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Gathering some first statistics</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#writing-a-useful-conversion-program" class="table-of-contents__link toc-highlight">Writing a useful conversion program.</a><ul><li><a href="#writing-a-command-line-program" class="table-of-contents__link toc-highlight">Writing a command-line program</a></li><li><a href="#parsing-the-data" class="table-of-contents__link toc-highlight">Parsing the data</a></li><li><a href="#outputting-to-sqlite" class="table-of-contents__link toc-highlight">Outputting to sqlite</a></li></ul></li><li><a href="#job-not-done-yet" class="table-of-contents__link toc-highlight">Job not done yet</a></li><li><a href="#back-to-the-task" class="table-of-contents__link toc-highlight">Back to the task</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.a63558cc.js"></script>
<script src="/whg-training-resources/assets/js/main.81209970.js"></script>
</body>
</html>