<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">A tale of two 2x2 tables | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="A tale of two 2x2 tables | WHG training resources"><meta data-rh="true" name="description" content="Welcome to (possibly) the world&#x27;s first choose your own statistical adventure!"><meta data-rh="true" property="og:description" content="Welcome to (possibly) the world&#x27;s first choose your own statistical adventure!"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.a63558cc.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.81209970.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/whg-training-resources/programming/introduction_to_R/">programming</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Hidden markov models/">README</a><button aria-label="Toggle the collapsible sidebar category &#x27;README&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/">GMS Statistical Modelling I - introduction to statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;GMS Statistical Modelling I - introduction to statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/notes/Probability solutions/">notes</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">practicals</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">a_tale_of_two_2x2_tables</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">Model checking the malaria exposure finding</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison/">O_bld_group_genome_wide_comparison</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/When asymptotics go wrong/">When asymptotics go wrong</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction/">A tale of two 2x2 tables</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/">bayesian_meta_analysis</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical/">linear_regression</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/">malaria_exposure</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison/">o_bld_grp</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/distributions warmup/">warmup</a></div></li></ul></li></ul></li></ul></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/"><span itemprop="name">Statistical modelling</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/Introduction/"><span itemprop="name">GMS Statistical Modelling I - introduction to statistical modelling</span></a><meta itemprop="position" content="2"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">practicals</span><meta itemprop="position" content="3"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">a_tale_of_two_2x2_tables</span><meta itemprop="position" content="4"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">A tale of two 2x2 tables</span><meta itemprop="position" content="5"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>A tale of two 2x2 tables</h1><p>Welcome to (possibly) the world&#x27;s first <em>choose your own statistical adventure</em>!</p><p>In this practical we are going to study the following two two-by-two tables, which come from two
published papers about genetic variants that affect malaria susceptibility:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">                       non-O      O                                          rs60822373   rs60822373</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                       blood gp.  blood gp.                                  G allele     C allele</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            controls:  3420       3233               unexposed populations:  1965         1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">severe malaria cases:  3925       2738         malaria-exposed populations:  707          17</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The left table contains data from <a href="https://doi.org/10.1038/s41467-019-13480-z" target="_blank" rel="noopener noreferrer">https://doi.org/10.1038/s41467-019-13480-z</a>, and shows that O blood group is at
lower frequency in severe malaria cases than in the general population, consistent with a protective effect of O blood
group.</p><p>The right table comes from <a href="https://science.sciencemag.org/content/348/6235/711" target="_blank" rel="noopener noreferrer">https://science.sciencemag.org/content/348/6235/711</a>, and shows that the rs60822373 &#x27;C&#x27;
alleles is at higher frequency in malaria-exposed population (here sub-Saharan Africans) than in non-exposed
populations (European-ancestry individuals). rs60822373 encodes the Cromer blood group, so this is consistent with a
protective effect of the Cromer blood group on malaria. (In this right table, the data actually comes from the <a href="http://www.nature.com/nature/journal/v491/n7422/full/nature11632.html" target="_blank" rel="noopener noreferrer">1000
Genomes Project phase I</a>).</p><p>If you run a statistical test on either table you will see that these are highly statistically significant differences
in frequencies between the two rows of each table, so these differences are not just due to sampling.</p><p>But <em>something is wrong with one of these tables</em>.  Your mission is to find out what!</p><p><em>QN</em>.  Maybe you can already see what is wrong?  Give it some thought before we proceed.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="modelling-2x2-tables">Modelling 2x2 tables<a class="hash-link" href="#modelling-2x2-tables" title="Direct link to heading">​</a></h2><p>To quantify the effects in both tables we need a model. In both tables, the data was collected based on the row labels
(i.e. case/control or population of collection) so the most straightforward model is of <em>binomial sampling in rows</em>.
In other words, for a table:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           X    Y        (total)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">unexposed  a    b        (N₁ = a+b)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  exposed  c    d        (N₂ = c+d)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We model the counts of <code>Y</code> as:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">  b ~ binomial( n = N₁, p = ϑ₁ )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  d ~ binomial( n = N₂, p = ϑ₂ )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>where ϑ₁ and ϑ₂ are frequencies in the two rows.</p><p><em>Exercise</em> You can implement this model easily using the code we have already written - see <a target="_blank" href="/whg-training-resources/assets/files/table.ll-a3f4177f29d917faced6ec349ec06988.R/">this file</a> for an implementation.</p><p>However, we&#x27;d really like to parameterise the model so that it contains an <em>effect size parameter</em> - a parameter that
reflects how different the two row frequencies are.  We will use the <em>log odds ratio</em> to do this.  In the above table, the observed log odds ratio is the quantity:</p><p><img loading="lazy" src="https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Ctext%7Blog%7D+%5Ctext%7BOR%7D%3D%5Clog%5Cleft%28%5Cfrac%7Ba+d%7D%7Bb+c%7D%5Cright%29" alt="\text{log} \text{OR}=\log\left(\frac{a d}{b c}\right)" class="img_E7b_"></p><p>There are a couple of reasons to focus on log odds ratio here.  One of them is mathematical convenience.  For a probability p, the odds is defined as p/(1-p).  The odds lies in <!-- -->[0,∞]<!-- --> and thus the log-odds lies in <!-- -->[-∞,∞]<!-- -->.  It is generally useful to have parameters living in an unbounded space.</p><p>Thus, transforming to log odds allows us to put our parameters on the real line.</p><p>Another way to think of this is through the inverse function, which maps log odds back to probability space.  This function is known as the <em>logistic</em> function and defined by:</p><p><img loading="lazy" src="https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Ctext%7Blogistic%7D%28x%29+%3D+%5Cfrac%7Be%5Ex%7D%7B1%2Be%5Ex%7D" alt="\text{logistic}(x) = \frac{e^x}{1+e^x}" class="img_E7b_"></p><p>Let&#x27;s plot it:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">logistic &lt;- function( x ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    exp(x) / ( 1 + exp(x) )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x = seq( from = -10, to = 10, by = 0.01 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot( x, logistic(x), type = &#x27;l&#x27;, bty = &#x27;n&#x27; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">grid()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="logistic function" src="/whg-training-resources/assets/images/logistic-bd5b92de76b0e2e0f9ef99fb2d86f201.png" width="379" height="243" class="img_E7b_"></p><p>You can see that the logistic function maps the real line (log-odds space, x axis) to the unit interval (y axis, probability space).  It is smooth and tails off to being essentially flat outside around <!-- -->[-10,10]<!-- -->.</p><p><em>Exercise</em> prove that <code>logistic()</code> is the inverse of the log odds (substitute one expression into the other and simplify). Alternatively, prove this to yourself by implementing both function in R.</p><p><em>Exercise</em> (advanced) what is the slope of the logistic function at x=0?  (Hint: you need to compute the derivative.)</p><p>Now that we are working in log-odds space (on the real line), this frees us up to express our log odds as a <em>baseline log odds</em> (applying to both rows of the table) plus an <em>additional log odds conferred by the second row</em>.  Let&#x27;s call these parameters μ and β.  We are thus writing:</p><p><img loading="lazy" src="https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Ctheta_1+%3D+%5Ctext%7Blogistic%7D%28%5Cmu%29%0A%5Cquad%5Ctext%7Band%7D%5Cquad%0A%5Ctheta_2+%3D+%5Ctext%7Blogistic%7D%28%5Cmu%2B%5Cbeta%29%0A" alt="\theta_1 = \text{logistic}(\mu)
\quad\text{and}\quad
\theta_2 = \text{logistic}(\mu+\beta)
" class="img_E7b_"></p><p>Or equivalently:</p><p><img loading="lazy" src="https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Ctext%7Blog%7D+%5Ctext%7Bodds%7D%28%5Ctheta_1%29%3D%5Cmu%0A%5Cquad%5Ctext%7Band%7D%5Cquad%0A%5Ctext%7Blog%7D+%5Ctext%7Bodds%7D%5Cleft%28%5Ctheta_2%5Cright%29+%3D+%5Cmu%2B%5Cbeta%0A" alt="\text{log} \text{odds}(\theta_1)=\mu
\quad\text{and}\quad
\text{log} \text{odds}\left(\theta_2\right) = \mu+\beta
" class="img_E7b_"></p><p>such that β is the log odds ratio.</p><p>How does this look in practice?  Lets implement it:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># Requires binomial.ll first!</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">table.ll &lt;- function( data, params = list( theta = c( 0.5, 0.5 )) ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # For a 2x2 table we assume binomial sampling in rows</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # So sum the lls over the rows</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return (</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        binomial.ll(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            x = data[1,2],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            params = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                n = rowSums( data )[1],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                p = params$theta[1]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        + binomial.ll(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            data[2,2],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            params = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                n = sum( data[2,] ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                p = params$theta[2]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">logistic &lt;- function( x ) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    exp(x) / ( 1 + exp(x) )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">reparameterised.table.ll &lt;- function( data, params = list( theta = 0.5, log.or = 0 )) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # params[1] is log-odds of baseline frequency</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    # params[2] is log odds ratio</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    theta = c(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        logistic( params$theta ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        logistic( params$theta + params$log.or )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    return( table.ll( data, params = list( theta = theta )))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>(These functions rely on <a href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/loglikelihoods/binomial.ll/">binomial.ll</a>).</p><p>Let&#x27;s load up the tables:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">tables = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    case_control = matrix(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        c( 3420, 3233, 3925, 2738 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        byrow = T, ncol = 2,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        dimnames = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            c( &quot;non-O&quot;, &quot;O&quot; ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            c( &quot;controls&quot;, &quot;cases&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    exposure = matrix(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        c( 1965, 1, 707, 17 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        byrow = T, ncol = 2,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        dimnames = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            c( &quot;G&quot;, &quot;C&quot; ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">            c( &quot;unexposed&quot;, &quot;exposed&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>And plot the likelihood (this uses <a href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/loglikelihoods/inspect.ll/">inspect.ll</a>).)</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">fit1 = inspect.ll( reparameterised.table.ll, data = tables[[1]], params = list( theta = 0.5, log.or = 0 ) )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fit2 = inspect.ll( reparameterised.table.ll, data = tables[[2]], params = list( theta = 0.5, log.or = 0 ) )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Look at both these plots.  For the first table:</p><p><img loading="lazy" alt="Table 1 likelihood" src="/whg-training-resources/assets/images/table_1_ll-72c889489d8f211c811b13d11a536dc8.png" width="800" height="600" class="img_E7b_"></p><p>The plot for the first table is about as good as this plot could look.  It estimates a log odds-ratio of -0.30, which is an odds ratio of ~0.74.  Also, the Gaussian fit to the likelihood is about as good as it could possibly be.  We could get a reasonable interval as:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">   &gt; sprintf(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">       &quot;%.2f ( %.2f - %.2f)&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">       exp( fit1$mle$log.or ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">       exp( fit1$mle$log.or - 1.96 * fit1$standard.errors[1] ), exp( fit1$mle$log.or + 1.96 * fit1$standard.errors[1] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">   [1] &quot;0.74 ( 0.70 - 0.77)&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>As discussed in class, you can interpret this in one of two ways. The bayesian interpretation is that our uncertainty
in the true odds ratio, given this particular data table, is well expressed by a small interval around 0.74. The above
interval is a 95% bayesian <em>credible interval</em> for the parameter - meaning an interval that contains 95% of the
posterior mass. (In principle this interpretation depends on assuming a flat prior, so that the posterior is
proportional to the likelihood. But because the data is so strong here, the prior won&#x27;t affect this much as long as the
prior is relatively spread out). The posterior is pretty much Gaussian, so the probability that the parameter was
actually zero, or something above zero, can be computed from the quantiles of the Gaussian distribution:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt; pnorm( 0, mean = fit1$mle$log.or, sd = fit1$standard.errors[2], lower.tail = F )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">[1] 1.711726e-18</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>meaning that (assuming the model holds) it&#x27;s vanishingly unlikely we have got the sign of this effect wrong.</p><p><em>Note</em> this is a bit daft because we shouldn&#x27;t be allowed a &#x27;flat prior&#x27; here (there&#x27;s no function that integrates to 1 over the whole real line).  And in any case a &#x27;flat prior&#x27; would place most of its mass outside any finite interval so this prior really claims that the effect is huge.  Adding a <em>proper prior</em> here is not hard - </p><p><em>Exercise</em> add a proper prior here (e.g. Gaussian with mean 0 and standard deviation sigma) - how does this change the above?</p><p>Alternatively, we discussed in lectures how this can be interpreted as a frequentist statement about the parameter. In
this setting we think of <em>replicates of the table from the same sampling process</em> and interpret the above interval as a
<em>confidence interval</em>. To state this specifically, it means:</p><ul><li>Imagine an infinite number of replicates of the data with some &#x27;true&#x27; parameters</li><li>Imagine for each such data table we used the above procedure to compute a confidence interval</li><li><em>Then</em> the true parameter would be in the computed interval 95% of the time.</li></ul><p>If this strikes you as a complicated interpretation - it is! My general advice is to interpret this interval as a
statement about our uncertainty in the parameter given the model - i.e. as a bayesian credible interval. In more
general situations with less informative data you should think carefully about what prior is appropriate. Where the
frequentist approach becomes useful is in <em>calibrating</em> or <em>model checking</em> our model against reality.  In this case it&#x27;s useful to know that <em>if the true log-odds were zero</em>, then the estimate would be distributed like</p><p><img loading="lazy" src="https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Chat%7B%5Cbeta%7D%7EN%280%2C%5Ctext%7Bse%7D%5E2%29" alt="\hat{\beta}~N(0,\text{se}^2)" class="img_E7b_"></p><p>so that a one-tailed P-value can be computed from quantiles of this normal distribution:</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt; pnorm( abs(fit1$mle$log.or), mean = 0, sd = fit1$standard.errors[2], lower.tail = F )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">[1] 1.711726e-18</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>Exercise</em> This number is of course exactly equal to the Bayesian probability of getting the sign wrong above - satisfy yourself why this is true.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="when-asymptotics-go-wrong">When asymptotics go wrong<a class="hash-link" href="#when-asymptotics-go-wrong" title="Direct link to heading">​</a></h2><p>On the other hand, look at the plot for the 2nd table:</p><p><img loading="lazy" alt="Table 2 likelihood" src="/whg-training-resources/assets/images/table_2_ll-e81c604db2012c4841dacdb6caee1644.png" width="800" height="600" class="img_E7b_"></p><p>Clearly something goes wrong with the asymptotics for this table.  The Gaussian approximation is not great.  Focussing on the log-odds ratio parameter, the approximation puts <em>too little weight</em> on small odds ratios, and it puts <em>too much weight</em> on larger odds ratios.  If we again compute a P-value</p><div class="language-R codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-R codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt; pnorm( abs(fit2$mle$log.or), mean = 0, sd = fit2$standard.errors[2], lower.tail = F )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">[1] 9.020974e-05</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>it is still statistically significant - but we should be cautious because the gaussian approximation is not great.</p><p>Here are two ways to deal with this.  First, we could use <em>Fisher&#x27;s exact test</em> intead.  Fisher&#x27;s exact test deals with this situation by additionally conditioning on the column sums of the table.  This is like saying we knew what the frequency of the variant was beforehand</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/When asymptotics go wrong/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">When asymptotics go wrong</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Bayesian meta-analysis practical</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#modelling-2x2-tables" class="table-of-contents__link toc-highlight">Modelling 2x2 tables</a></li><li><a href="#when-asymptotics-go-wrong" class="table-of-contents__link toc-highlight">When asymptotics go wrong</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.a63558cc.js"></script>
<script src="/whg-training-resources/assets/js/main.81209970.js"></script>
</body>
</html>