<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">O_bld_group_genome_wide_comparison | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="O_bld_group_genome_wide_comparison | WHG training resources"><meta data-rh="true" name="description" content="Calibration / model checking for the O blood group finding"><meta data-rh="true" property="og:description" content="Calibration / model checking for the O blood group finding"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.a63558cc.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.81209970.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/whg-training-resources/programming/introduction_to_R/">programming</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Hidden markov models/">README</a><button aria-label="Toggle the collapsible sidebar category &#x27;README&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/">GMS Statistical Modelling I - introduction to statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;GMS Statistical Modelling I - introduction to statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/notes/Probability solutions/">notes</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-3 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">practicals</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">a_tale_of_two_2x2_tables</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/">Model checking the malaria exposure finding</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison/">O_bld_group_genome_wide_comparison</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/When asymptotics go wrong/">When asymptotics go wrong</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-5 menu__list-item"><a class="menu__link" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/introduction/">A tale of two 2x2 tables</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/bayesian_meta_analysis/bayesian_meta_analysis_practical/">bayesian_meta_analysis</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/linear_regression/atp2b4/atp2b4_practical/">linear_regression</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/malaria_exposure/1000Genomes_genome_wide_comparison/">malaria_exposure</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/o_bld_grp/O_bld_group_genome_wide_comparison/">o_bld_grp</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-4 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/statistical_modelling/Introduction/practicals/warmup/distributions warmup/">warmup</a></div></li></ul></li></ul></li></ul></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/"><span itemprop="name">Statistical modelling</span></a><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/statistical_modelling/Introduction/"><span itemprop="name">GMS Statistical Modelling I - introduction to statistical modelling</span></a><meta itemprop="position" content="2"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">practicals</span><meta itemprop="position" content="3"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">a_tale_of_two_2x2_tables</span><meta itemprop="position" content="4"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">O_bld_group_genome_wide_comparison</span><meta itemprop="position" content="5"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>O_bld_group_genome_wide_comparison</h1></header><h2 class="anchor anchorWithStickyNavbar_mojV" id="calibration--model-checking-for-the-o-blood-group-finding">Calibration / model checking for the O blood group finding<a class="hash-link" href="#calibration--model-checking-for-the-o-blood-group-finding" title="Direct link to heading">​</a></h2><p>In class we analysed this table:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">                      non-O blood group    O blood group</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">controls              3420                 3233</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">severe malaria cases  3925                 2738</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>which can be created in R as:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data = matrix(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    c(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        3420, 3233,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        3925, 2738</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ncol = 2,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    byrow = T,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dimnames = list(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        c( &quot;controls&quot;, &quot;cases&quot; ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        c( &quot;non-O&quot;, &quot;O&quot; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The odds ratio computed from this table is 0.74 - suggesting O blood group is protective.  And a test of this table using <code>chisq.test()</code> or <code>fisher.test()</code> shows that this is highly statistically significant.  That means:</p><ul><li><em>If</em> the data were truly sampled from a binomial distribution with given frequency in each row</li><li>and <em>if</em> the true odds ratio was zero</li><li><em>then</em> a table with such a large odds ratio would almost never occur.</li></ul><p>These are approximations which don&#x27;t really hold.  Samples are not really independent (e.g. population structure is itself a reflection of distant relationships); frequencies might vary across populations; in a finite population it&#x27;s not really credible that the true effect size is really zero (though it might be very close).</p><p>We would therefore like to verify that there isn&#x27;t something going on we haven&#x27;t thought of.  I.e. is our null model really sensible?</p><p>One of the advantages of genome-wide analysis is that we have a lot of data to answer this with.
Suppose we compare our finding with all other &#x27;similar&#x27; variants in the genome.  They have all been
genotyped on the same set of samples.  So they represent the same sampling biases (if any) and the same demography.  However, due to recombination they also represent many independent draws from the genealogical history of the sample.</p><p>The file <code>recessive_counts_matching_O_frequency.csv.gz</code> contains data for all the alleles in our
study that match the frequency of the O blood group mutation.</p><p>Load it:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">X = read.csv(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    &quot;recessive_counts_matching_O_frequency.csv.gz&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    header = T,      # Tell R our file has a row of column names</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    as.is = T,       # Ask R to please not transmogrify any data</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    check.names = F  # Also ask R not to mess around with column names </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Let&#x27;s first sanity check I&#x27;ve computed frequencies right.  It&#x27;s often best to first collect numerical data into a matrix - this is a container of numbers that is layed out nicely in memory for computation.  (Unlike a data frame which holds columns of data of different types).  The <code>as.matrix()</code> function does this:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data = as.matrix( X[, c( &quot;controls:A&quot;, &quot;controls:B&quot;, &quot;cases:A&quot;, &quot;cases:B&quot; )])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># look at the data</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head( data )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">X$frequency = (data[,2] + data[,4]) / rowSums(data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># plot it</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hist(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    X$frequency,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    breaks = 5, # how many bars across the plot?</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;Recessive dosage frequency&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlim = c( 0.3, 0.7 ) </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">abline( v = X$frequency[ which( X$rsid == &#x27;rs8176719&#x27; )], col = &#x27;red&#x27; )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Compute the log-odds ratio for every variant:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">X$logOR = log( (data[,1] * data[,4]) / ( data[,2] * data[,3]) )</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Let&#x27;s look at the distribution of log odds ratios:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">hist(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    X$logOR,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    breaks = 50,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;log Odds ratio&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    main = &quot;Alleles matching O blood group frequency&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlim = c( -0.5, 0.5 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xaxt = &#x27;n&#x27;  # Turn off the axis so we can draw our own</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ticks = c( 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">axis(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    1,       # on the bottom edge</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    at = log(ticks),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    label = ticks</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Let&#x27;s put an arrow where O blood group allele is:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">arrows(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    x0 = X$logOR[ which( X$rsid == &#x27;rs8176719&#x27; )],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    x1 = X$logOR[ which( X$rsid == &#x27;rs8176719&#x27; )],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    y0 = 2000, y1 = 100,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    length = 0.1,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    col = &#x27;red&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">text(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    x = X$logOR[ which( X$rsid == &#x27;rs8176719&#x27; )], y = 2000,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pos = 3, # draw text above</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    label = &quot;O bld grp.&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    col = &#x27;red&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Compute the standard error of the log OR (using the standard formula):</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># for table</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">#   a b</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">#   c d</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># the formula is sqrt( 1/a + 1/b + 1/c + 1/d )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">X$SE = sqrt( rowSums( 1/data ))</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Is the null model at all plausible here?  The null model says that all variation in the effect size (the log OR) is due to sampling variation.  It should be distributed as a N(0, SE^2).  (all the stnadard errors will be similar here because of the choice of variant).  To test how well the null model fits, we can plot standard-error-adjusted effect sizes (&quot;Z-scores&quot;) against a gaussian distribution:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">region = c( -5, 5 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hist(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    X$logOR / X$SE,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    breaks = 100,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;x score&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlim = region,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ylim = c( 0, 0.5 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    main = &quot;Alleles matching O blood group frequency&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    freq = FALSE # plot y axis values as a density, for comparison</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x = seq( from = region[1], to = region[2], by = 0.01 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">points(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    x,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    dnorm( x, mean = 0, sd = 1 ),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    type = &#x27;l&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    col = &#x27;red&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>It&#x27;s not a bad fit!  There&#x27;s a small gap at the top and slightly fatter tail to the real data.
A more sensitive plot is a quantile-quantile plot, where we plot ordered statistics against their expection.  We can do that here:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">zscores = sort( X$logOR / X$SE )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">expected = qnorm( (1:length(zscores))/(length(zscores)+1)) # Normal distribution quantiles</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    expected,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    zscores,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pch = 19 # nice round dots</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">abline( a = 0, b = 1, col = &#x27;red&#x27; )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">grid()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The distribution definitely has fatter tails than expected.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="aside">Aside<a class="hash-link" href="#aside" title="Direct link to heading">​</a></h2><p>(A: I&#x27;ve the qq plot in terms of z-scores, because I like thinking in &#x27;effect size&#x27; space.  A more often seen qq-plot is based on a p-value as follows.  A traditional (Wald test) P-value compares
the z-score with the standard gaussian, as in our plot above:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">X$pvalue = pnorm( -abs(X$logOR), sd = X$SE ) * 2</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>Note</em>: the <code>abs()</code> and multiplying by two in the above are needed to capture both the left and right tail of the distribution.  We&#x27;d be interested in effects going in either direction.  (You could equally well use the squared z-score and <code>pchisq</code>.)  We can make a qq-plot by plotting p-values against ranked expected quantiles:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">pvalues = -log10( sort( X$pvalue ) )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">expected = -log10( (1:length(pvalues)  / (length(pvalues)+1)))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    expected,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlab = &quot;Expected log10 P-value&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pvalues,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ylab = &quot;Observed log10 P-value&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pch = 19, # nice round dots</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    xlim = c( 0, 20 ), # make it square</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    ylim = c( 0, 20 )  # make it square</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">abline( a = 0, b = 1, lty = 2 )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Let&#x27;s highlight O blood group on the plkot:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">w = which( pvalues == -log10( X$pvalue[ X$rsid == &#x27;rs8176719&#x27; ]))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">points(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    expected[w],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pvalues[w],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    col = &#x27;red&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">text(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    expected[w],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pvalues[w],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pos = 2,             # put it to left</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    adj = 1,             # right-justify</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    label = &quot;O bld grp&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Challenge: the kth point in this plot is -log10 of a kth order statistic.  Use this fact to add vertical error bars for each point?</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="summary">Summary<a class="hash-link" href="#summary" title="Direct link to heading">​</a></h2><p>We have shown</p><ol><li>That O blood has an estimated protective effect in these data.</li><li>That such a large estimated effect is unlikely <em>under the formal model assumptions</em> if the effect were really zero (hence the small P-value).</li><li>That other similar variants genome-wide do seem fairly compatible with the model of no effect.  This gives us confidence that our P-value (computed against the null) reflects something realistic. In particular, there isn&#x27;t some large systematic problem with our use of the P-value that would invalidate our analysis.</li><li>However, genome-wide variants don&#x27;t seem 100% compatible with the model assumptions.</li></ol><p>Note that:</p><ul><li>point 1 is purely about our statistical model and the data we have observed</li><li>point 2 is a comparison of our data with a <em>hypothetical infinite set of unobserved data</em> generated under the model assumptions.</li><li>point 3 is a comparison of our data with other real data generated in the same samples.</li><li>point 4 suggests there is still further calibration to perform.</li></ul><p>Of course we also know that the O blood group mutation (rs8176719) is a deletion of protein-coding sequence that alters red cell surface antigens, making us a priori likely to think this might be involved (compared to, say, a randomly chosen genetic variant).</p><p>What could we do re: point 4?  Are there actually many variants across the genome having nontrivial true effect on malaria susceptibility?  Or is there some systematic confounding going on?  (This seems not impossible, due to the diverse set of samples data was drawn from.)</p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/O_bld_group_genome_wide_comparison.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/1000Genomes_genome_wide_comparison/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Model checking the malaria exposure finding</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/statistical_modelling/Introduction/practicals/a_tale_of_two_2x2_tables/When asymptotics go wrong/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">When asymptotics go wrong</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#calibration--model-checking-for-the-o-blood-group-finding" class="table-of-contents__link toc-highlight">Calibration / model checking for the O blood group finding</a></li><li><a href="#aside" class="table-of-contents__link toc-highlight">Aside</a></li><li><a href="#summary" class="table-of-contents__link toc-highlight">Summary</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.a63558cc.js"></script>
<script src="/whg-training-resources/assets/js/main.81209970.js"></script>
</body>
</html>